{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "e16bdcee-7ae7-474f-ad5a-d2c97a4bf5bd",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/kvalem/.local/lib/python3.11/site-packages/numba/core/cpu.py:99: UserWarning: Numba extension module 'awkward.numba' failed to load due to 'ModuleNotFoundError(No module named 'importlib_metadata')'.\n",
      "  numba.core.entrypoints.init_all()\n"
     ]
    }
   ],
   "source": [
    "import warnings\n",
    "import os\n",
    "from pathlib import Path \n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import altair as alt\n",
    "import anndata\n",
    "from pydeseq2.dds import DeseqDataSet\n",
    "from pydeseq2.ds import DeseqStats\n",
    "import decoupler as dc\n",
    "import itertools\n",
    "from scipy import sparse\n",
    "#import lib.data_helpers as dh\n",
    "\n",
    "#import lib.scanpy_helpers as sh\n",
    "#import lib.pl.util as pu"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "22f47bd7-6b63-406c-90be-1976dde7e6c0",
   "metadata": {},
   "outputs": [],
   "source": [
    "path = \"/data/projects/2023/LCBiome/nsclc_gender_atlas_tmp/out/011_analysis_paired_remove_xy/pseudobulk/\"\n",
    "resDir_figures = \"/data/projects/2023/LCBiome/nsclc_gender_atlas_tmp/out/012_LUAD/03012025/figures/LUAD_DE/\"\n",
    "resDir_tables = \"/data/projects/2023/LCBiome/nsclc_gender_atlas_tmp/out/012_LUAD/03012025/tables/LUAD_DE/\"\n",
    "input_path = f\"{path}/paired_adata_clean.h5ad\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "b302a5bc-a55e-4546-8e09-df95182918ca",
   "metadata": {},
   "outputs": [],
   "source": [
    "#adata = sc.read_h5ad(input_path) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "96cdf2b8-6b2a-4e24-87b1-4459666b96eb",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = sc.read_h5ad(\"/data/projects/2023/LCBiome/nsclc_gender_atlas_tmp/out/012_LUAD/old/pseudobulk/paired_adata_clean_luad_all_cell_type.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "66bc75d2-19e1-43ff-8bd2-2d9a57ded167",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "standardized_cell_type\n",
      "macrophage           104271\n",
      "CD4+ T cell           94503\n",
      "CD8+ T cell           72603\n",
      "epithelial cell       21039\n",
      "malignant cell        16939\n",
      "B cell                16853\n",
      "regulatory T cell     16785\n",
      "dendritic cell        16702\n",
      "endothelial cell      16150\n",
      "plasma cell           15096\n",
      "neutrophil             6146\n",
      "fibroblast             4674\n",
      "Name: count, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "# Define the cell types to keep\n",
    "selected_cell_types = {\n",
    "    'macrophage',\n",
    "    'CD4+ T cell',\n",
    "    'CD8+ T cell',\n",
    "    'B cell',\n",
    "    'malignant cell',\n",
    "    'regulatory T cell',\n",
    "    'epithelial cell',\n",
    "    'neutrophil',\n",
    "    'dendritic cell',\n",
    "    'plasma cell',\n",
    "    'endothelial cell',\n",
    "    'fibroblast'\n",
    "}\n",
    "\n",
    "# Mapping for synonymous cell types in adata\n",
    "cell_type_mapping = {\n",
    "    'fibroblast of lung': 'fibroblast',\n",
    "    'vein endothelial cell': 'endothelial cell',\n",
    "    'capillary endothelial cell': 'endothelial cell',\n",
    "    'pulmonary artery endothelial cell': 'endothelial cell',\n",
    "    'endothelial cell of lymphatic vessel': 'endothelial cell',\n",
    "    'conventional dendritic cell': 'dendritic cell',\n",
    "    'CD1c-positive myeloid dendritic cell': 'dendritic cell',\n",
    "    'plasmacytoid dendritic cell': 'dendritic cell',\n",
    "    'bronchus fibroblast of lung': 'fibroblast'\n",
    "}\n",
    "\n",
    "# Standardize cell types based on mapping\n",
    "adata.obs['standardized_cell_type'] = adata.obs['cell_type'].replace(cell_type_mapping)\n",
    "\n",
    "# Filter adata to keep only selected cell types\n",
    "adata = adata[adata.obs['standardized_cell_type'].isin(selected_cell_types)].copy()\n",
    "\n",
    "# Verify result\n",
    "print(adata.obs['standardized_cell_type'].value_counts())\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d80dd4a4-e7e9-4b10-af92-17fa7b97f02c",
   "metadata": {},
   "source": [
    "## There are malignant cells in normal_adjacent samples\n",
    "Comming from 59 different patients and 14 datasets\n",
    "I cannot give an explanation for this so i remove this \"mislalbeled\" cells "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "456f79b3-729a-42e3-9a36-b248d431f319",
   "metadata": {},
   "outputs": [],
   "source": [
    "grouped = adata.obs.groupby(['standardized_cell_type', 'origin']).size().reset_index(name='cell_count')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "09f280ae-168e-4771-b732-822d128e42f6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>standardized_cell_type</th>\n",
       "      <th>origin</th>\n",
       "      <th>cell_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>epithelial cell</td>\n",
       "      <td>normal_adjacent</td>\n",
       "      <td>4957</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>epithelial cell</td>\n",
       "      <td>tumor_primary</td>\n",
       "      <td>16082</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>macrophage</td>\n",
       "      <td>normal_adjacent</td>\n",
       "      <td>65453</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>macrophage</td>\n",
       "      <td>tumor_primary</td>\n",
       "      <td>38818</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>B cell</td>\n",
       "      <td>normal_adjacent</td>\n",
       "      <td>2888</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>B cell</td>\n",
       "      <td>tumor_primary</td>\n",
       "      <td>13965</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>dendritic cell</td>\n",
       "      <td>normal_adjacent</td>\n",
       "      <td>6752</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>dendritic cell</td>\n",
       "      <td>tumor_primary</td>\n",
       "      <td>9950</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>CD4+ T cell</td>\n",
       "      <td>normal_adjacent</td>\n",
       "      <td>45654</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>CD4+ T cell</td>\n",
       "      <td>tumor_primary</td>\n",
       "      <td>48849</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>CD8+ T cell</td>\n",
       "      <td>normal_adjacent</td>\n",
       "      <td>31378</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>CD8+ T cell</td>\n",
       "      <td>tumor_primary</td>\n",
       "      <td>41225</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>neutrophil</td>\n",
       "      <td>normal_adjacent</td>\n",
       "      <td>4155</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>neutrophil</td>\n",
       "      <td>tumor_primary</td>\n",
       "      <td>1991</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>plasma cell</td>\n",
       "      <td>normal_adjacent</td>\n",
       "      <td>1951</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>plasma cell</td>\n",
       "      <td>tumor_primary</td>\n",
       "      <td>13145</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>regulatory T cell</td>\n",
       "      <td>normal_adjacent</td>\n",
       "      <td>2876</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>regulatory T cell</td>\n",
       "      <td>tumor_primary</td>\n",
       "      <td>13909</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>malignant cell</td>\n",
       "      <td>normal_adjacent</td>\n",
       "      <td>574</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>malignant cell</td>\n",
       "      <td>tumor_primary</td>\n",
       "      <td>16365</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>endothelial cell</td>\n",
       "      <td>normal_adjacent</td>\n",
       "      <td>9754</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>endothelial cell</td>\n",
       "      <td>tumor_primary</td>\n",
       "      <td>6396</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>fibroblast</td>\n",
       "      <td>normal_adjacent</td>\n",
       "      <td>2160</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>fibroblast</td>\n",
       "      <td>tumor_primary</td>\n",
       "      <td>2514</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   standardized_cell_type           origin  cell_count\n",
       "0         epithelial cell  normal_adjacent        4957\n",
       "1         epithelial cell    tumor_primary       16082\n",
       "2              macrophage  normal_adjacent       65453\n",
       "3              macrophage    tumor_primary       38818\n",
       "4                  B cell  normal_adjacent        2888\n",
       "5                  B cell    tumor_primary       13965\n",
       "6          dendritic cell  normal_adjacent        6752\n",
       "7          dendritic cell    tumor_primary        9950\n",
       "8             CD4+ T cell  normal_adjacent       45654\n",
       "9             CD4+ T cell    tumor_primary       48849\n",
       "10            CD8+ T cell  normal_adjacent       31378\n",
       "11            CD8+ T cell    tumor_primary       41225\n",
       "12             neutrophil  normal_adjacent        4155\n",
       "13             neutrophil    tumor_primary        1991\n",
       "14            plasma cell  normal_adjacent        1951\n",
       "15            plasma cell    tumor_primary       13145\n",
       "16      regulatory T cell  normal_adjacent        2876\n",
       "17      regulatory T cell    tumor_primary       13909\n",
       "18         malignant cell  normal_adjacent         574\n",
       "19         malignant cell    tumor_primary       16365\n",
       "20       endothelial cell  normal_adjacent        9754\n",
       "21       endothelial cell    tumor_primary        6396\n",
       "22             fibroblast  normal_adjacent        2160\n",
       "23             fibroblast    tumor_primary        2514"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "grouped"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "1237edea-08af-44e3-9b68-82aac54b7910",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create a Boolean mask for the cells that you want to keep\n",
    "mask = ~((adata.obs['origin'] == 'normal_adjacent') & (adata.obs['standardized_cell_type'] == 'malignant cell'))\n",
    "\n",
    "# Subset the adata object using the mask to exclude the specific cells\n",
    "adata = adata[mask].copy()\n",
    "\n",
    "# Now `adata_filtered` contains all cells except those with origin == 'normal_adjacent' and cell_type == 'malignant cell'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "d19a8270-bf0b-41a1-9dd0-a3d03ac3c54c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>standardized_cell_type</th>\n",
       "      <th>origin</th>\n",
       "      <th>cell_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>epithelial cell</td>\n",
       "      <td>normal_adjacent</td>\n",
       "      <td>4957</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>epithelial cell</td>\n",
       "      <td>tumor_primary</td>\n",
       "      <td>16082</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>macrophage</td>\n",
       "      <td>normal_adjacent</td>\n",
       "      <td>65453</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>macrophage</td>\n",
       "      <td>tumor_primary</td>\n",
       "      <td>38818</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>B cell</td>\n",
       "      <td>normal_adjacent</td>\n",
       "      <td>2888</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>B cell</td>\n",
       "      <td>tumor_primary</td>\n",
       "      <td>13965</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>dendritic cell</td>\n",
       "      <td>normal_adjacent</td>\n",
       "      <td>6752</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>dendritic cell</td>\n",
       "      <td>tumor_primary</td>\n",
       "      <td>9950</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>CD4+ T cell</td>\n",
       "      <td>normal_adjacent</td>\n",
       "      <td>45654</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>CD4+ T cell</td>\n",
       "      <td>tumor_primary</td>\n",
       "      <td>48849</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>CD8+ T cell</td>\n",
       "      <td>normal_adjacent</td>\n",
       "      <td>31378</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>CD8+ T cell</td>\n",
       "      <td>tumor_primary</td>\n",
       "      <td>41225</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>neutrophil</td>\n",
       "      <td>normal_adjacent</td>\n",
       "      <td>4155</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>neutrophil</td>\n",
       "      <td>tumor_primary</td>\n",
       "      <td>1991</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>plasma cell</td>\n",
       "      <td>normal_adjacent</td>\n",
       "      <td>1951</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>plasma cell</td>\n",
       "      <td>tumor_primary</td>\n",
       "      <td>13145</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>regulatory T cell</td>\n",
       "      <td>normal_adjacent</td>\n",
       "      <td>2876</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>regulatory T cell</td>\n",
       "      <td>tumor_primary</td>\n",
       "      <td>13909</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>malignant cell</td>\n",
       "      <td>normal_adjacent</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>malignant cell</td>\n",
       "      <td>tumor_primary</td>\n",
       "      <td>16365</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>endothelial cell</td>\n",
       "      <td>normal_adjacent</td>\n",
       "      <td>9754</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>endothelial cell</td>\n",
       "      <td>tumor_primary</td>\n",
       "      <td>6396</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>fibroblast</td>\n",
       "      <td>normal_adjacent</td>\n",
       "      <td>2160</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>fibroblast</td>\n",
       "      <td>tumor_primary</td>\n",
       "      <td>2514</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   standardized_cell_type           origin  cell_count\n",
       "0         epithelial cell  normal_adjacent        4957\n",
       "1         epithelial cell    tumor_primary       16082\n",
       "2              macrophage  normal_adjacent       65453\n",
       "3              macrophage    tumor_primary       38818\n",
       "4                  B cell  normal_adjacent        2888\n",
       "5                  B cell    tumor_primary       13965\n",
       "6          dendritic cell  normal_adjacent        6752\n",
       "7          dendritic cell    tumor_primary        9950\n",
       "8             CD4+ T cell  normal_adjacent       45654\n",
       "9             CD4+ T cell    tumor_primary       48849\n",
       "10            CD8+ T cell  normal_adjacent       31378\n",
       "11            CD8+ T cell    tumor_primary       41225\n",
       "12             neutrophil  normal_adjacent        4155\n",
       "13             neutrophil    tumor_primary        1991\n",
       "14            plasma cell  normal_adjacent        1951\n",
       "15            plasma cell    tumor_primary       13145\n",
       "16      regulatory T cell  normal_adjacent        2876\n",
       "17      regulatory T cell    tumor_primary       13909\n",
       "18         malignant cell  normal_adjacent           0\n",
       "19         malignant cell    tumor_primary       16365\n",
       "20       endothelial cell  normal_adjacent        9754\n",
       "21       endothelial cell    tumor_primary        6396\n",
       "22             fibroblast  normal_adjacent        2160\n",
       "23             fibroblast    tumor_primary        2514"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "grouped = adata.obs.groupby(['standardized_cell_type', 'origin']).size().reset_index(name='cell_count')\n",
    "grouped"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "b294c685-6846-4687-bb5e-f2d6259767e6",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_normal = adata[adata.obs[\"origin\"]==\"normal_adjacent\"]\n",
    "#adata_normal = adata_normal[~adata_normal.obs[\"cell_type\"].isin([\"malignant cell\"])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "e09eb487-4cd8-47c8-8a8f-763ea846365e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'B cell',\n",
       " 'CD1c-positive myeloid dendritic cell',\n",
       " 'CD4+ T cell',\n",
       " 'CD8+ T cell',\n",
       " 'bronchus fibroblast of lung',\n",
       " 'capillary endothelial cell',\n",
       " 'conventional dendritic cell',\n",
       " 'dendritic cell',\n",
       " 'endothelial cell of lymphatic vessel',\n",
       " 'epithelial cell',\n",
       " 'fibroblast of lung',\n",
       " 'macrophage',\n",
       " 'neutrophil',\n",
       " 'plasma cell',\n",
       " 'plasmacytoid dendritic cell',\n",
       " 'pulmonary artery endothelial cell',\n",
       " 'regulatory T cell',\n",
       " 'vein endothelial cell'}"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "set(adata_normal.obs.cell_type)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "328ddbdc-878e-4c07-85b4-5722f779b8a3",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_tumor = adata[adata.obs[\"origin\"]==\"tumor_primary\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "2d2130c4-eb94-47c5-8a7c-d10790e22005",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'B cell',\n",
       " 'CD1c-positive myeloid dendritic cell',\n",
       " 'CD4+ T cell',\n",
       " 'CD8+ T cell',\n",
       " 'bronchus fibroblast of lung',\n",
       " 'capillary endothelial cell',\n",
       " 'conventional dendritic cell',\n",
       " 'dendritic cell',\n",
       " 'endothelial cell of lymphatic vessel',\n",
       " 'epithelial cell',\n",
       " 'fibroblast of lung',\n",
       " 'macrophage',\n",
       " 'neutrophil',\n",
       " 'plasma cell',\n",
       " 'plasmacytoid dendritic cell',\n",
       " 'pulmonary artery endothelial cell',\n",
       " 'regulatory T cell',\n",
       " 'vein endothelial cell'}"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "set(adata_normal.obs.cell_type)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "1a71968f-bea9-4959-8f51-d88ea23a4ab5",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_normal = adata[adata.obs[\"origin\"]==\"normal_adjacent\"]\n",
    "#adata_normal = adata_normal[~adata_normal.obs[\"cell_type\"].isin([\"malignant cell\"])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "bf8fc463-3287-4c83-8695-856d6bef25a5",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_tumor = adata[adata.obs[\"origin\"]==\"tumor_primary\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "f9541bae-a756-46c4-8b9c-028bb4977437",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 401187 × 17811\n",
       "    obs: 'sample', 'uicc_stage', 'ever_smoker', 'age', 'donor_id', 'origin', 'dataset', 'ann_fine', 'cell_type_predicted', 'doublet_status', 'leiden', 'n_genes_by_counts', 'total_counts', 'total_counts_mito', 'pct_counts_mito', 'ann_coarse', 'cell_type_tumor', 'tumor_stage', 'EGFR_mutation', 'TP53_mutation', 'ALK_mutation', 'BRAF_mutation', 'ERBB2_mutation', 'KRAS_mutation', 'ROS_mutation', 'origin_fine', 'study', 'platform', 'cell_type_major', 'cell_type_neutro', 'cell_type_neutro_coarse', 'suspension_type', 'assay_ontology_term_id', 'cell_type_ontology_term_id', 'development_stage_ontology_term_id', 'disease_ontology_term_id', 'self_reported_ethnicity_ontology_term_id', 'is_primary_data', 'organism_ontology_term_id', 'sex_ontology_term_id', 'tissue_ontology_term_id', 'cell_type', 'assay', 'disease', 'organism', 'sex', 'tissue', 'self_reported_ethnicity', 'development_stage', 'standardized_cell_type'\n",
       "    var: 'is_highly_variable', 'mito', 'n_cells_by_counts', 'mean_counts', 'pct_dropout_by_counts', 'total_counts', 'feature_is_filtered', 'feature_name', 'feature_reference', 'feature_biotype'\n",
       "    uns: '_scvi', 'ann_fine_colors', 'batch_condition', 'cell_type_major_colors', 'cell_type_tumor_colors', 'dataset_colors', 'default_embedding', 'hvg', 'log1p', 'neighbors', 'origin_colors', 'platform_colors', 'schema_version', 'sex_colors', 'title', 'umap'\n",
       "    obsm: 'X_scANVI', 'X_scVI', 'X_umap'\n",
       "    layers: 'count', 'counts_length_scaled', 'log1p_norm'\n",
       "    obsp: 'connectivities', 'distances'"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "23eb009f-183e-4dff-8316-cffe2bc1091f",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_male = adata[adata.obs[\"sex\"]==\"male\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "4d1b5e85-2527-4d1c-99fa-610707886713",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_female = adata[adata.obs[\"sex\"]==\"female\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "93a631e2-f821-4c9d-ab5f-1c614e31555a",
   "metadata": {},
   "source": [
    "## Prepare for pseudobulk"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "3aab4e92-5686-4453-895f-d37d0c04d2a5",
   "metadata": {},
   "outputs": [],
   "source": [
    "contrasts = [\n",
    "        dict(var = \"origin\", condition = \"tumor_primary\", reference = \"normal_adjacent\"),\n",
    "    ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "b9fec3e6-3081-4f1b-a09b-f768bbac6449",
   "metadata": {
    "lines_to_next_cell": 0
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'var': 'origin',\n",
       "  'condition': 'tumor_primary',\n",
       "  'reference': 'normal_adjacent'}]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "contrasts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "e594b137-8749-4b47-8cf0-d3c4e5ea9d9b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'tumor_primary_vs_normal_adjacent'"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "contrasts[0][\"condition\"].replace(\" \", \"_\") + \"_vs_\" + contrasts[0][\"reference\"].replace(\" \", \"_\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "98d3567a-8711-4498-8330-b26adc6c1d7f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/data/projects/2023/LCBiome/nsclc_gender_atlas_tmp/out/012_LUAD/03012025/tables/LUAD_DE/'"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "resDir_tables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "08bd7241-10d2-4275-a5b9-f28c25be88e2",
   "metadata": {},
   "outputs": [],
   "source": [
    "for contrast in contrasts:\n",
    "    name = contrast[\"condition\"].replace(\" \", \"_\") + \"_vs_\" + contrast[\"reference\"].replace(\" \", \"_\")\n",
    "    contrast[\"name\"] = name\n",
    "    res_dir = Path(resDir_tables, name, \"tables\")\n",
    "    os.makedirs(resDir_tables, mode = 0o750, exist_ok = True)\n",
    "    contrast[\"res_dir\"] = resDir_tables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "ff2d6f1e-2481-4dae-a49d-fbb1426fd0b9",
   "metadata": {},
   "outputs": [],
   "source": [
    "cell_type_class = \"standardized_cell_type\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "7cbfe361-9a46-42e3-8b39-bcda74d64729",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/data/projects/2023/LCBiome/nsclc_gender_atlas_tmp/out/012_LUAD/03012025/tables/LUAD_DE/'"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "contrast[\"res_dir\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "bb2a9221-a889-41e7-a9c9-59d384cd68a0",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdata = dc.get_pseudobulk(adata,\n",
    "                          sample_col='sample',\n",
    "                          groups_col=cell_type_class,\n",
    "                          layer='count',\n",
    "                          mode='sum',\n",
    "                          min_cells=10,\n",
    "                          min_counts=1000\n",
    "                         )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "1fab92d4-0fbc-40b0-8e3e-ca276b4462f9",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdata.X = pdata.X.astype(int)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "26fca082-014b-4c1c-84b1-50a5f944db99",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdata.var_names = pdata.var[\"feature_name\"].astype(str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "946e30f3-ca0c-4527-959b-ebf62cb2678a",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdata_obs_df = pdata.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "aa4facd2-7fa1-43ca-bd3a-cd9abafc4af4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 1269 × 17811\n",
       "    obs: 'sample', 'standardized_cell_type', 'uicc_stage', 'ever_smoker', 'donor_id', 'origin', 'dataset', 'tumor_stage', 'study', 'platform', 'suspension_type', 'assay_ontology_term_id', 'development_stage_ontology_term_id', 'disease_ontology_term_id', 'self_reported_ethnicity_ontology_term_id', 'is_primary_data', 'organism_ontology_term_id', 'sex_ontology_term_id', 'tissue_ontology_term_id', 'assay', 'disease', 'organism', 'sex', 'tissue', 'self_reported_ethnicity', 'development_stage', 'psbulk_n_cells', 'psbulk_counts'\n",
       "    var: 'is_highly_variable', 'mito', 'n_cells_by_counts', 'mean_counts', 'pct_dropout_by_counts', 'total_counts', 'feature_is_filtered', 'feature_name', 'feature_reference', 'feature_biotype'\n",
       "    layers: 'psbulk_props'"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pdata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "682718d7-5233-425e-815b-b3ca570add58",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Group by patient and filter for patients who have samples in both 'tumor' and 'normal_adjacent' conditions\n",
    "patients_with_both_conditions = pdata_obs_df.groupby('donor_id')['origin'].apply(lambda x: set(x) >= {'tumor_primary', 'normal_adjacent'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "009b87fd-b915-4cb0-9e8b-ab8374af37b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get the patient IDs that meet the condition\n",
    "patients_with_both_conditions_ids = patients_with_both_conditions[patients_with_both_conditions].index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "bacd47b9-e19c-48c9-a5c7-525005477dbc",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdata_tumor_normal = pdata[pdata.obs['donor_id'].isin(patients_with_both_conditions_ids)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "e9c8e063-ffa2-44ed-a57a-1b40b5577119",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'standardized_cell_type'"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cell_type_class"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "2784f408-3b12-4f84-bf68-7e934b075386",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Run deseq2 on pseudobulk all cell types\n",
    "#cell_types = pdata_tumor_normal.obs[cell_type_class].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "1e7e7423-9fbb-43f8-94c1-b05a2e2185bc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'normal_adjacent'"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "contrast[\"reference\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "3bbbbd74-3ef2-486e-9666-9725c72cfffa",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'tumor_primary'"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "contrast[\"condition\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "4990e0c8-fdb6-483c-8dc1-675c9eaa378f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'tumor_primary_vs_normal_adjacent'"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "contrast[\"name\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "88c40d2e-b7a9-43d8-be5f-014136ce10f7",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_2960518/2235016355.py:1: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  pdata_tumor_normal.obs['sex'] = pdata_tumor_normal.obs['sex'].astype('category')\n"
     ]
    }
   ],
   "source": [
    "pdata_tumor_normal.obs['sex'] = pdata_tumor_normal.obs['sex'].astype('category')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "762683f5-ba14-4465-9606-a737f0fc4221",
   "metadata": {},
   "outputs": [],
   "source": [
    "cpus=16"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "7df7538a-42be-48bb-ae0a-f93107d67e44",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 1268 × 17811\n",
       "    obs: 'sample', 'standardized_cell_type', 'uicc_stage', 'ever_smoker', 'donor_id', 'origin', 'dataset', 'tumor_stage', 'study', 'platform', 'suspension_type', 'assay_ontology_term_id', 'development_stage_ontology_term_id', 'disease_ontology_term_id', 'self_reported_ethnicity_ontology_term_id', 'is_primary_data', 'organism_ontology_term_id', 'sex_ontology_term_id', 'tissue_ontology_term_id', 'assay', 'disease', 'organism', 'sex', 'tissue', 'self_reported_ethnicity', 'development_stage', 'psbulk_n_cells', 'psbulk_counts'\n",
       "    var: 'is_highly_variable', 'mito', 'n_cells_by_counts', 'mean_counts', 'pct_dropout_by_counts', 'total_counts', 'feature_is_filtered', 'feature_name', 'feature_reference', 'feature_biotype'\n",
       "    layers: 'psbulk_props'"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pdata_tumor_normal"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "22aa039f-b6c8-483a-a95d-30857f2b32da",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdata_tumor_normal.obs[\"standardized_cell_type\"]= pdata_tumor_normal.obs[\"standardized_cell_type\"].replace(['epithelial cell of lung','multi-ciliated epithelial cell',], 'epithelial cell')\n",
    "pdata_tumor_normal.obs[\"standardized_cell_type\"]= pdata_tumor_normal.obs[\"standardized_cell_type\"].replace(['alveolar macrophage'], 'macrophage')\n",
    "pdata_tumor_normal.obs[\"standardized_cell_type\"]= pdata_tumor_normal.obs[\"standardized_cell_type\"].replace(['CD4-positive, alpha-beta T cell'], 'CD4+ T cell')\n",
    "pdata_tumor_normal.obs[\"standardized_cell_type\"]= pdata_tumor_normal.obs[\"standardized_cell_type\"].replace(['CD8-positive, alpha-beta T cell'], 'CD8+ T cell')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "aab20e1e-8396-406c-96ba-e26c9d3786cb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "origin\n",
       "tumor_primary      708\n",
       "normal_adjacent    560\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pdata_tumor_normal.obs[\"origin\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "4c8e5253-21c7-457e-bd5b-e4112ee25b54",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdata_only_tumor = pdata_tumor_normal[pdata_tumor_normal.obs[\"origin\"]==\"tumor_primary\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "21e92241-4371-4d5e-8c9a-fee0e8623d18",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "View of AnnData object with n_obs × n_vars = 708 × 17811\n",
       "    obs: 'sample', 'standardized_cell_type', 'uicc_stage', 'ever_smoker', 'donor_id', 'origin', 'dataset', 'tumor_stage', 'study', 'platform', 'suspension_type', 'assay_ontology_term_id', 'development_stage_ontology_term_id', 'disease_ontology_term_id', 'self_reported_ethnicity_ontology_term_id', 'is_primary_data', 'organism_ontology_term_id', 'sex_ontology_term_id', 'tissue_ontology_term_id', 'assay', 'disease', 'organism', 'sex', 'tissue', 'self_reported_ethnicity', 'development_stage', 'psbulk_n_cells', 'psbulk_counts'\n",
       "    var: 'is_highly_variable', 'mito', 'n_cells_by_counts', 'mean_counts', 'pct_dropout_by_counts', 'total_counts', 'feature_is_filtered', 'feature_name', 'feature_reference', 'feature_biotype'\n",
       "    layers: 'psbulk_props'"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pdata_only_tumor"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "0f2146e6-70df-483c-ac40-996ea95e38b0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/data/projects/2023/LCBiome/nsclc_gender_atlas_tmp/out/012_LUAD/03012025/tables/LUAD_DE/'"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "resDir_tables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "11a0abfc-867f-4d25-ac53-38cd8f94ce0d",
   "metadata": {},
   "outputs": [],
   "source": [
    "cell_types = pdata_only_tumor.obs[cell_type_class].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "d834d08e-4aa3-4e67-a3de-f14cd7401f80",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdata_tumor_normal_male = pdata_tumor_normal[pdata_tumor_normal.obs[\"sex\"]==\"male\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "9305005a-583d-4fba-ba54-a3ff56ae3019",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdata_tumor_normal_female = pdata_tumor_normal[pdata_tumor_normal.obs[\"sex\"]==\"female\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "890a98e8-c950-4fb5-943d-3156cecd6ffd",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "cb52b035-26af-4d7c-b0b0-af55d080f0ff",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_2960518/680501246.py:1: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  pdata_tumor_normal_male.obs[\"origin\"] = pdata_tumor_normal_male.obs[\"origin\"].str.replace(\"_\", \"-\", regex=False)\n"
     ]
    }
   ],
   "source": [
    "pdata_tumor_normal_male.obs[\"origin\"] = pdata_tumor_normal_male.obs[\"origin\"].str.replace(\"_\", \"-\", regex=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "7558823a-62dc-4c68-8a20-a599079e1d77",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_2960518/307224016.py:1: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  pdata_tumor_normal_female.obs[\"origin\"] = pdata_tumor_normal_female.obs[\"origin\"].str.replace(\"_\", \"-\", regex=False)\n"
     ]
    }
   ],
   "source": [
    "pdata_tumor_normal_female.obs[\"origin\"] = pdata_tumor_normal_female.obs[\"origin\"].str.replace(\"_\", \"-\", regex=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "efdf3692-81ca-49e8-81c0-3d4203327fc4",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdata_tumor_normal_male = pdata_tumor_normal_male[~pdata_tumor_normal_male.obs[\"standardized_cell_type\"].isin([\"malignant cell\"])]\n",
    "pdata_tumor_normal_female= pdata_tumor_normal_female[~pdata_tumor_normal_female.obs[\"standardized_cell_type\"].isin([\"malignant cell\"])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "b00d38c3-7970-4609-9600-223ad9f601b9",
   "metadata": {},
   "outputs": [],
   "source": [
    "contrast[\"reference\"] = \"normal-adjacent\"\n",
    "contrast[\"condition\"] = \"tumor-primary\"\n",
    "contrast[\"name\"] = \"tumor-primary_vs_normal-adjacent\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "954b8ed9-3fa2-4a41-9851-192f49a6ad88",
   "metadata": {},
   "source": [
    "## Pydeseq tumor vs normal MALE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "id": "72136d8c-55f0-4b80-8513-91d05a142641",
   "metadata": {
    "lines_to_next_cell": 2,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "B cell\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Fitting size factors...\n",
      "... done in 0.03 seconds.\n",
      "\n",
      "Fitting dispersions...\n",
      "... done in 1.19 seconds.\n",
      "\n",
      "Fitting dispersion trend curve...\n",
      "... done in 0.35 seconds.\n",
      "\n",
      "Fitting MAP dispersions...\n",
      "... done in 1.43 seconds.\n",
      "\n",
      "Fitting LFCs...\n",
      "... done in 1.49 seconds.\n",
      "\n",
      "Calculating cook's distance...\n",
      "... done in 0.06 seconds.\n",
      "\n",
      "Replacing 59 outlier genes.\n",
      "\n",
      "Fitting dispersions...\n",
      "... done in 0.02 seconds.\n",
      "\n",
      "Fitting MAP dispersions...\n",
      "... done in 0.02 seconds.\n",
      "\n",
      "Fitting LFCs...\n",
      "... done in 0.02 seconds.\n",
      "\n",
      "Running Wald tests...\n",
      "... done in 8.70 seconds.\n",
      "\n",
      "Fitting MAP LFCs...\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Log2 fold change & Wald test p-value: origin tumor-primary vs normal-adjacent\n",
      "              baseMean  log2FoldChange     lfcSE      stat    pvalue      padj\n",
      "feature_name                                                                  \n",
      "A1BG          4.613264        0.000056  0.323332  0.000174  0.999861  0.999978\n",
      "A1BG-AS1      0.591671       -0.200691  0.570672 -0.351676  0.725081       NaN\n",
      "A2M           1.206051       -0.223805  0.620775 -0.360525  0.718455       NaN\n",
      "A2M-AS1       0.082406       -1.195479  1.465479 -0.815760  0.414638       NaN\n",
      "A2ML1         0.157295       -1.658056  1.488754 -1.113721  0.265399       NaN\n",
      "...                ...             ...       ...       ...       ...       ...\n",
      "ZXDC          0.928493       -0.355715  0.578193 -0.615219  0.538410       NaN\n",
      "ZYG11A        0.066189       -0.837893  1.683108 -0.497825  0.618607       NaN\n",
      "ZYG11B        1.394873        0.450700  0.493065  0.914079  0.360675       NaN\n",
      "ZYX           3.686372        0.780150  0.348659  2.237571  0.025249  0.698723\n",
      "ZZEF1         1.957142       -1.083770  0.426699 -2.539895  0.011089       NaN\n",
      "\n",
      "[17811 rows x 6 columns]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/kvalem/.conda/envs/crc_atlas/lib/python3.11/site-packages/pydeseq2/utils.py:1260: RuntimeWarning: overflow encountered in exp\n",
      "  counts - (counts + size) / (1 + size * np.exp(-xbeta - offset))\n",
      "/home/kvalem/.conda/envs/crc_atlas/lib/python3.11/site-packages/pydeseq2/utils.py:1260: RuntimeWarning: overflow encountered in exp\n",
      "  counts - (counts + size) / (1 + size * np.exp(-xbeta - offset))\n",
      "... done in 2.01 seconds.\n",
      "\n",
      "Fitting size factors...\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Shrunk log2 fold change & Wald test p-value: origin tumor-primary vs normal-adjacent\n",
      "              baseMean  log2FoldChange     lfcSE      stat    pvalue      padj\n",
      "feature_name                                                                  \n",
      "A1BG          4.613264        0.000005  0.052210  0.000174  0.999861  0.999978\n",
      "A1BG-AS1      0.591671        0.000091  0.063279 -0.351676  0.725081       NaN\n",
      "A2M           1.206051       -0.000674  0.053568 -0.360525  0.718455       NaN\n",
      "A2M-AS1       0.082406        0.000124  0.060909 -0.815760  0.414638       NaN\n",
      "A2ML1         0.157295       -0.000482  0.046209 -1.113721  0.265399       NaN\n",
      "...                ...             ...       ...       ...       ...       ...\n",
      "ZXDC          0.928493       -0.000956  0.053715 -0.615219  0.538410       NaN\n",
      "ZYG11A        0.066189        0.000512  0.077715 -0.497825  0.618607       NaN\n",
      "ZYG11B        1.394873       -0.005089  0.048999  0.914079  0.360675       NaN\n",
      "ZYX           3.686372        0.006723  0.061714  2.237571  0.025249  0.698723\n",
      "ZZEF1         1.957142       -0.006125  0.049189 -2.539895  0.011089       NaN\n",
      "\n",
      "[17811 rows x 6 columns]\n",
      "CD4+ T cell\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "... done in 0.04 seconds.\n",
      "\n",
      "Fitting dispersions...\n",
      "... done in 1.74 seconds.\n",
      "\n",
      "Fitting dispersion trend curve...\n",
      "... done in 0.39 seconds.\n",
      "\n",
      "Fitting MAP dispersions...\n",
      "... done in 1.69 seconds.\n",
      "\n",
      "Fitting LFCs...\n",
      "... done in 1.19 seconds.\n",
      "\n",
      "Calculating cook's distance...\n",
      "... done in 0.08 seconds.\n",
      "\n",
      "Replacing 883 outlier genes.\n",
      "\n",
      "Fitting dispersions...\n",
      "... done in 0.12 seconds.\n",
      "\n",
      "Fitting MAP dispersions...\n",
      "... done in 0.11 seconds.\n",
      "\n",
      "Fitting LFCs...\n",
      "... done in 0.09 seconds.\n",
      "\n",
      "Running Wald tests...\n",
      "... done in 8.69 seconds.\n",
      "\n",
      "Fitting MAP LFCs...\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Log2 fold change & Wald test p-value: origin tumor-primary vs normal-adjacent\n",
      "                baseMean  log2FoldChange     lfcSE      stat    pvalue  \\\n",
      "feature_name                                                             \n",
      "A1BG           68.409417       -0.403161  0.363153 -1.110170  0.266926   \n",
      "A1BG-AS1       12.580317        0.623197  0.342646  1.818775  0.068946   \n",
      "A2M            12.078685        0.277251  0.423670  0.654403  0.512852   \n",
      "A2M-AS1        11.573832       -0.346347  0.355080 -0.975406  0.329359   \n",
      "A2ML1           0.199986       -0.256593  1.561691 -0.164304  0.869491   \n",
      "...                  ...             ...       ...       ...       ...   \n",
      "ZXDC           13.965527        0.137765  0.302803  0.454965  0.649134   \n",
      "ZYG11A          0.226355       -0.205677  1.871029 -0.109927  0.912467   \n",
      "ZYG11B         22.581971       -0.044293  0.285489 -0.155148  0.876704   \n",
      "ZYX           131.585482        0.032505  0.185704  0.175034  0.861053   \n",
      "ZZEF1          39.032896       -0.074626  0.207538 -0.359578  0.719163   \n",
      "\n",
      "                  padj  \n",
      "feature_name            \n",
      "A1BG          0.716684  \n",
      "A1BG-AS1      0.438869  \n",
      "A2M           0.862854  \n",
      "A2M-AS1       0.774291  \n",
      "A2ML1              NaN  \n",
      "...                ...  \n",
      "ZXDC          0.907375  \n",
      "ZYG11A             NaN  \n",
      "ZYG11B        0.970695  \n",
      "ZYX           0.966623  \n",
      "ZZEF1         0.930312  \n",
      "\n",
      "[17811 rows x 6 columns]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/kvalem/.conda/envs/crc_atlas/lib/python3.11/site-packages/pydeseq2/utils.py:1260: RuntimeWarning: overflow encountered in exp\n",
      "  counts - (counts + size) / (1 + size * np.exp(-xbeta - offset))\n",
      "/home/kvalem/.conda/envs/crc_atlas/lib/python3.11/site-packages/pydeseq2/utils.py:1260: RuntimeWarning: overflow encountered in exp\n",
      "  counts - (counts + size) / (1 + size * np.exp(-xbeta - offset))\n",
      "/home/kvalem/.conda/envs/crc_atlas/lib/python3.11/site-packages/pydeseq2/utils.py:1260: RuntimeWarning: overflow encountered in exp\n",
      "  counts - (counts + size) / (1 + size * np.exp(-xbeta - offset))\n",
      "/home/kvalem/.conda/envs/crc_atlas/lib/python3.11/site-packages/pydeseq2/utils.py:1260: RuntimeWarning: overflow encountered in exp\n",
      "  counts - (counts + size) / (1 + size * np.exp(-xbeta - offset))\n",
      "/home/kvalem/.conda/envs/crc_atlas/lib/python3.11/site-packages/pydeseq2/utils.py:1260: RuntimeWarning: overflow encountered in exp\n",
      "  counts - (counts + size) / (1 + size * np.exp(-xbeta - offset))\n",
      "/home/kvalem/.conda/envs/crc_atlas/lib/python3.11/site-packages/pydeseq2/utils.py:1260: RuntimeWarning: overflow encountered in exp\n",
      "  counts - (counts + size) / (1 + size * np.exp(-xbeta - offset))\n",
      "... done in 3.08 seconds.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Shrunk log2 fold change & Wald test p-value: origin tumor-primary vs normal-adjacent\n",
      "                baseMean  log2FoldChange     lfcSE      stat    pvalue  \\\n",
      "feature_name                                                             \n",
      "A1BG           68.409417   -3.090000e-06  0.001431 -1.110170  0.266926   \n",
      "A1BG-AS1       12.580317    5.328158e-06  0.001753  1.818775  0.068946   \n",
      "A2M            12.078685    1.613500e-06  0.001617  0.654403  0.512852   \n",
      "A2M-AS1        11.573832   -2.851820e-06  0.001456 -0.975406  0.329359   \n",
      "A2ML1           0.199986   -3.554808e-08  0.001541 -0.164304  0.869491   \n",
      "...                  ...             ...       ...       ...       ...   \n",
      "ZXDC           13.965527    1.712778e-06  0.001600  0.454965  0.649134   \n",
      "ZYG11A          0.226355   -8.508677e-08  0.001524 -0.109927  0.912467   \n",
      "ZYG11B         22.581971   -5.581784e-07  0.001532 -0.155148  0.876704   \n",
      "ZYX           131.585482    1.051806e-06  0.001554  0.175034  0.861053   \n",
      "ZZEF1          39.032896   -1.870712e-06  0.001523 -0.359578  0.719163   \n",
      "\n",
      "                  padj  \n",
      "feature_name            \n",
      "A1BG          0.716684  \n",
      "A1BG-AS1      0.438869  \n",
      "A2M           0.862854  \n",
      "A2M-AS1       0.774291  \n",
      "A2ML1              NaN  \n",
      "...                ...  \n",
      "ZXDC          0.907375  \n",
      "ZYG11A             NaN  \n",
      "ZYG11B        0.970695  \n",
      "ZYX           0.966623  \n",
      "ZZEF1         0.930312  \n",
      "\n",
      "[17811 rows x 6 columns]\n",
      "CD8+ T cell\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Fitting size factors...\n",
      "... done in 0.04 seconds.\n",
      "\n",
      "Fitting dispersions...\n",
      "... done in 1.80 seconds.\n",
      "\n",
      "Fitting dispersion trend curve...\n",
      "... done in 0.39 seconds.\n",
      "\n",
      "Fitting MAP dispersions...\n",
      "... done in 1.71 seconds.\n",
      "\n",
      "Fitting LFCs...\n",
      "... done in 1.55 seconds.\n",
      "\n",
      "Calculating cook's distance...\n",
      "... done in 0.08 seconds.\n",
      "\n",
      "Replacing 972 outlier genes.\n",
      "\n",
      "Fitting dispersions...\n",
      "... done in 0.13 seconds.\n",
      "\n",
      "Fitting MAP dispersions...\n",
      "... done in 0.13 seconds.\n",
      "\n",
      "Fitting LFCs...\n",
      "... done in 0.12 seconds.\n",
      "\n",
      "Running Wald tests...\n",
      "... done in 8.71 seconds.\n",
      "\n",
      "Fitting MAP LFCs...\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Log2 fold change & Wald test p-value: origin tumor-primary vs normal-adjacent\n",
      "                baseMean  log2FoldChange     lfcSE      stat    pvalue  \\\n",
      "feature_name                                                             \n",
      "A1BG           52.908189        0.200887  0.378468  0.530789  0.595565   \n",
      "A1BG-AS1        9.445406       -0.103584  0.375120 -0.276136  0.782444   \n",
      "A2M            14.829458       -0.553485  0.379607 -1.458047  0.144828   \n",
      "A2M-AS1        32.573034       -1.071010  0.356349 -3.005506  0.002651   \n",
      "A2ML1           0.261897        0.117141  1.952931  0.059982  0.952170   \n",
      "...                  ...             ...       ...       ...       ...   \n",
      "ZXDC           12.725702       -0.378616  0.356195 -1.062946  0.287806   \n",
      "ZYG11A          0.283978       -0.662599  1.907608 -0.347345  0.728332   \n",
      "ZYG11B         14.679519       -0.459053  0.311875 -1.471915  0.141044   \n",
      "ZYX           136.861614        0.115195  0.172723  0.666935  0.504813   \n",
      "ZZEF1          35.284278       -0.236552  0.213814 -1.106347  0.268577   \n",
      "\n",
      "                  padj  \n",
      "feature_name            \n",
      "A1BG          0.931940  \n",
      "A1BG-AS1      0.969758  \n",
      "A2M           0.624598  \n",
      "A2M-AS1       0.060087  \n",
      "A2ML1              NaN  \n",
      "...                ...  \n",
      "ZXDC          0.798647  \n",
      "ZYG11A             NaN  \n",
      "ZYG11B        0.617775  \n",
      "ZYX           0.912210  \n",
      "ZZEF1         0.784351  \n",
      "\n",
      "[17811 rows x 6 columns]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/kvalem/.conda/envs/crc_atlas/lib/python3.11/site-packages/pydeseq2/utils.py:1260: RuntimeWarning: overflow encountered in exp\n",
      "  counts - (counts + size) / (1 + size * np.exp(-xbeta - offset))\n",
      "... done in 3.27 seconds.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Shrunk log2 fold change & Wald test p-value: origin tumor-primary vs normal-adjacent\n",
      "                baseMean  log2FoldChange     lfcSE      stat    pvalue  \\\n",
      "feature_name                                                             \n",
      "A1BG           52.908189    1.480970e-06  0.001538  0.530789  0.595565   \n",
      "A1BG-AS1        9.445406   -3.433613e-07  0.001460 -0.276136  0.782444   \n",
      "A2M            14.829458   -3.958412e-06  0.001349 -1.458047  0.144828   \n",
      "A2M-AS1        32.573034   -9.029687e-06  0.001271 -3.005506  0.002651   \n",
      "A2ML1           0.261897   -3.224398e-07  0.001260  0.059982  0.952170   \n",
      "...                  ...             ...       ...       ...       ...   \n",
      "ZXDC           12.725702   -2.981506e-06  0.001403 -1.062946  0.287806   \n",
      "ZYG11A          0.283978   -1.865567e-07  0.001289 -0.347345  0.728332   \n",
      "ZYG11B         14.679519   -4.997782e-06  0.001382 -1.471915  0.141044   \n",
      "ZYX           136.861614    4.419663e-06  0.001517  0.666935  0.504813   \n",
      "ZZEF1          35.284278   -5.392318e-06  0.001433 -1.106347  0.268577   \n",
      "\n",
      "                  padj  \n",
      "feature_name            \n",
      "A1BG          0.931940  \n",
      "A1BG-AS1      0.969758  \n",
      "A2M           0.624598  \n",
      "A2M-AS1       0.060087  \n",
      "A2ML1              NaN  \n",
      "...                ...  \n",
      "ZXDC          0.798647  \n",
      "ZYG11A             NaN  \n",
      "ZYG11B        0.617775  \n",
      "ZYX           0.912210  \n",
      "ZZEF1         0.784351  \n",
      "\n",
      "[17811 rows x 6 columns]\n",
      "dendritic cell\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Fitting size factors...\n",
      "... done in 0.03 seconds.\n",
      "\n",
      "Fitting dispersions...\n",
      "... done in 1.36 seconds.\n",
      "\n",
      "Fitting dispersion trend curve...\n",
      "... done in 0.39 seconds.\n",
      "\n",
      "Fitting MAP dispersions...\n",
      "... done in 1.63 seconds.\n",
      "\n",
      "Fitting LFCs...\n",
      "... done in 1.36 seconds.\n",
      "\n",
      "Calculating cook's distance...\n",
      "... done in 0.06 seconds.\n",
      "\n",
      "Replacing 59 outlier genes.\n",
      "\n",
      "Fitting dispersions...\n",
      "... done in 0.02 seconds.\n",
      "\n",
      "Fitting MAP dispersions...\n",
      "... done in 0.02 seconds.\n",
      "\n",
      "Fitting LFCs...\n",
      "... done in 0.02 seconds.\n",
      "\n",
      "Running Wald tests...\n",
      "... done in 8.69 seconds.\n",
      "\n",
      "Fitting MAP LFCs...\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Log2 fold change & Wald test p-value: origin tumor-primary vs normal-adjacent\n",
      "               baseMean  log2FoldChange     lfcSE      stat    pvalue  \\\n",
      "feature_name                                                            \n",
      "A1BG          27.346383       -0.187979  0.249624 -0.753049  0.451421   \n",
      "A1BG-AS1       2.723454       -0.021324  0.289433 -0.073676  0.941268   \n",
      "A2M           36.400901       -0.455576  0.233276 -1.952943  0.050826   \n",
      "A2M-AS1        0.704191        0.145368  0.665253  0.218515  0.827028   \n",
      "A2ML1          0.103433       -0.189091  1.902912 -0.099369  0.920845   \n",
      "...                 ...             ...       ...       ...       ...   \n",
      "ZXDC           3.328346       -0.094078  0.293786 -0.320225  0.748798   \n",
      "ZYG11A         0.103470        0.374354  2.423044  0.154498  0.877217   \n",
      "ZYG11B         5.576791        0.038424  0.239230  0.160615  0.872397   \n",
      "ZYX           75.852552        0.240159  0.096117  2.498604  0.012468   \n",
      "ZZEF1          6.935962       -0.288888  0.200627 -1.439930  0.149887   \n",
      "\n",
      "                  padj  \n",
      "feature_name            \n",
      "A1BG          0.865467  \n",
      "A1BG-AS1      0.988038  \n",
      "A2M           0.405537  \n",
      "A2M-AS1       0.969994  \n",
      "A2ML1              NaN  \n",
      "...                ...  \n",
      "ZXDC          0.952243  \n",
      "ZYG11A             NaN  \n",
      "ZYG11B        0.977547  \n",
      "ZYX           0.214069  \n",
      "ZZEF1         0.624394  \n",
      "\n",
      "[17811 rows x 6 columns]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/kvalem/.conda/envs/crc_atlas/lib/python3.11/site-packages/pydeseq2/utils.py:1260: RuntimeWarning: overflow encountered in exp\n",
      "  counts - (counts + size) / (1 + size * np.exp(-xbeta - offset))\n",
      "/home/kvalem/.conda/envs/crc_atlas/lib/python3.11/site-packages/pydeseq2/utils.py:1260: RuntimeWarning: overflow encountered in exp\n",
      "  counts - (counts + size) / (1 + size * np.exp(-xbeta - offset))\n",
      "/home/kvalem/.conda/envs/crc_atlas/lib/python3.11/site-packages/pydeseq2/utils.py:1260: RuntimeWarning: overflow encountered in exp\n",
      "  counts - (counts + size) / (1 + size * np.exp(-xbeta - offset))\n",
      "/home/kvalem/.conda/envs/crc_atlas/lib/python3.11/site-packages/pydeseq2/utils.py:1260: RuntimeWarning: overflow encountered in exp\n",
      "  counts - (counts + size) / (1 + size * np.exp(-xbeta - offset))\n",
      "/home/kvalem/.conda/envs/crc_atlas/lib/python3.11/site-packages/pydeseq2/utils.py:1260: RuntimeWarning: overflow encountered in exp\n",
      "  counts - (counts + size) / (1 + size * np.exp(-xbeta - offset))\n",
      "/home/kvalem/.conda/envs/crc_atlas/lib/python3.11/site-packages/pydeseq2/utils.py:1260: RuntimeWarning: overflow encountered in exp\n",
      "  counts - (counts + size) / (1 + size * np.exp(-xbeta - offset))\n",
      "/home/kvalem/.conda/envs/crc_atlas/lib/python3.11/site-packages/pydeseq2/utils.py:1260: RuntimeWarning: overflow encountered in exp\n",
      "  counts - (counts + size) / (1 + size * np.exp(-xbeta - offset))\n",
      "/home/kvalem/.conda/envs/crc_atlas/lib/python3.11/site-packages/pydeseq2/utils.py:1260: RuntimeWarning: overflow encountered in exp\n",
      "  counts - (counts + size) / (1 + size * np.exp(-xbeta - offset))\n",
      "/home/kvalem/.conda/envs/crc_atlas/lib/python3.11/site-packages/pydeseq2/utils.py:1260: RuntimeWarning: overflow encountered in exp\n",
      "  counts - (counts + size) / (1 + size * np.exp(-xbeta - offset))\n",
      "/home/kvalem/.conda/envs/crc_atlas/lib/python3.11/site-packages/pydeseq2/utils.py:1260: RuntimeWarning: overflow encountered in exp\n",
      "  counts - (counts + size) / (1 + size * np.exp(-xbeta - offset))\n",
      "... done in 3.30 seconds.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Shrunk log2 fold change & Wald test p-value: origin tumor-primary vs normal-adjacent\n",
      "               baseMean  log2FoldChange     lfcSE      stat    pvalue  \\\n",
      "feature_name                                                            \n",
      "A1BG          27.346383   -3.148722e-06  0.001433 -0.753049  0.451421   \n",
      "A1BG-AS1       2.723454   -2.306607e-07  0.001493 -0.073676  0.941268   \n",
      "A2M           36.400901   -8.781803e-06  0.001381 -1.952943  0.050826   \n",
      "A2M-AS1        0.704191    1.827489e-07  0.001549  0.218515  0.827028   \n",
      "A2ML1          0.103433   -3.227013e-07  0.001283 -0.099369  0.920845   \n",
      "...                 ...             ...       ...       ...       ...   \n",
      "ZXDC           3.328346   -1.171606e-06  0.001486 -0.320225  0.748798   \n",
      "ZYG11A         0.103470    3.866156e-07  0.002458  0.154498  0.877217   \n",
      "ZYG11B         5.576791    7.035676e-07  0.001498  0.160615  0.872397   \n",
      "ZYX           75.852552    2.682768e-05  0.001530  2.498604  0.012468   \n",
      "ZZEF1          6.935962   -7.480573e-06  0.001460 -1.439930  0.149887   \n",
      "\n",
      "                  padj  \n",
      "feature_name            \n",
      "A1BG          0.865467  \n",
      "A1BG-AS1      0.988038  \n",
      "A2M           0.405537  \n",
      "A2M-AS1       0.969994  \n",
      "A2ML1              NaN  \n",
      "...                ...  \n",
      "ZXDC          0.952243  \n",
      "ZYG11A             NaN  \n",
      "ZYG11B        0.977547  \n",
      "ZYX           0.214069  \n",
      "ZZEF1         0.624394  \n",
      "\n",
      "[17811 rows x 6 columns]\n",
      "endothelial cell\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Fitting size factors...\n",
      "... done in 0.02 seconds.\n",
      "\n",
      "Fitting dispersions...\n",
      "... done in 1.23 seconds.\n",
      "\n",
      "Fitting dispersion trend curve...\n",
      "... done in 0.37 seconds.\n",
      "\n",
      "Fitting MAP dispersions...\n",
      "... done in 1.43 seconds.\n",
      "\n",
      "Fitting LFCs...\n",
      "... done in 1.13 seconds.\n",
      "\n",
      "Calculating cook's distance...\n",
      "... done in 0.04 seconds.\n",
      "\n",
      "Replacing 94 outlier genes.\n",
      "\n",
      "Fitting dispersions...\n",
      "... done in 0.03 seconds.\n",
      "\n",
      "Fitting MAP dispersions...\n",
      "... done in 0.03 seconds.\n",
      "\n",
      "Fitting LFCs...\n",
      "... done in 0.03 seconds.\n",
      "\n",
      "Running Wald tests...\n",
      "... done in 9.17 seconds.\n",
      "\n",
      "Fitting MAP LFCs...\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Log2 fold change & Wald test p-value: origin tumor-primary vs normal-adjacent\n",
      "                baseMean  log2FoldChange     lfcSE      stat    pvalue  \\\n",
      "feature_name                                                             \n",
      "A1BG            1.507280       -0.010133  0.444279 -0.022808  0.981803   \n",
      "A1BG-AS1        0.249101        0.329657  0.870426  0.378730  0.704888   \n",
      "A2M           152.621226        1.062725  0.220783  4.813427  0.000001   \n",
      "A2M-AS1         0.446190        0.545755  0.822543  0.663498  0.507012   \n",
      "A2ML1           0.026018       -0.157982  3.116414 -0.050693  0.959570   \n",
      "...                  ...             ...       ...       ...       ...   \n",
      "ZXDC            1.064180        0.649751  0.464808  1.397891  0.162146   \n",
      "ZYG11A          0.056380        0.372103  3.103386  0.119902  0.904561   \n",
      "ZYG11B          2.856751        0.249009  0.312447  0.796964  0.425472   \n",
      "ZYX            11.161912       -0.118473  0.269639 -0.439376  0.660389   \n",
      "ZZEF1           2.342306        0.271431  0.354526  0.765615  0.443906   \n",
      "\n",
      "                  padj  \n",
      "feature_name            \n",
      "A1BG          0.994760  \n",
      "A1BG-AS1           NaN  \n",
      "A2M           0.000142  \n",
      "A2M-AS1            NaN  \n",
      "A2ML1              NaN  \n",
      "...                ...  \n",
      "ZXDC          0.523405  \n",
      "ZYG11A             NaN  \n",
      "ZYG11B        0.767087  \n",
      "ZYX           0.887347  \n",
      "ZZEF1         0.780258  \n",
      "\n",
      "[17811 rows x 6 columns]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/kvalem/.conda/envs/crc_atlas/lib/python3.11/site-packages/pydeseq2/utils.py:1260: RuntimeWarning: overflow encountered in exp\n",
      "  counts - (counts + size) / (1 + size * np.exp(-xbeta - offset))\n",
      "/home/kvalem/.conda/envs/crc_atlas/lib/python3.11/site-packages/pydeseq2/utils.py:1260: RuntimeWarning: overflow encountered in exp\n",
      "  counts - (counts + size) / (1 + size * np.exp(-xbeta - offset))\n",
      "/home/kvalem/.conda/envs/crc_atlas/lib/python3.11/site-packages/pydeseq2/utils.py:1260: RuntimeWarning: overflow encountered in exp\n",
      "  counts - (counts + size) / (1 + size * np.exp(-xbeta - offset))\n",
      "/home/kvalem/.conda/envs/crc_atlas/lib/python3.11/site-packages/pydeseq2/utils.py:1260: RuntimeWarning: overflow encountered in exp\n",
      "  counts - (counts + size) / (1 + size * np.exp(-xbeta - offset))\n",
      "/home/kvalem/.conda/envs/crc_atlas/lib/python3.11/site-packages/pydeseq2/utils.py:1260: RuntimeWarning: overflow encountered in exp\n",
      "  counts - (counts + size) / (1 + size * np.exp(-xbeta - offset))\n",
      "... done in 2.88 seconds.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Shrunk log2 fold change & Wald test p-value: origin tumor-primary vs normal-adjacent\n",
      "                baseMean  log2FoldChange     lfcSE      stat    pvalue  \\\n",
      "feature_name                                                             \n",
      "A1BG            1.507280   -1.923844e-08  0.001366 -0.022808  0.981803   \n",
      "A1BG-AS1        0.249101    7.402598e-07  0.001393  0.378730  0.704888   \n",
      "A2M           152.621226    9.613744e-01  0.225848  4.813427  0.000001   \n",
      "A2M-AS1         0.446190    1.136828e-06  0.001453  0.663498  0.507012   \n",
      "A2ML1           0.026018   -1.667452e-07  0.001155 -0.050693  0.959570   \n",
      "...                  ...             ...       ...       ...       ...   \n",
      "ZXDC            1.064180    3.368096e-06  0.001402  1.397891  0.162146   \n",
      "ZYG11A          0.056380    2.507463e-07  0.002485  0.119902  0.904561   \n",
      "ZYG11B          2.856751    2.644848e-06  0.001389  0.796964  0.425472   \n",
      "ZYX            11.161912   -1.628779e-06  0.001367 -0.439376  0.660389   \n",
      "ZZEF1           2.342306    2.185815e-06  0.001390  0.765615  0.443906   \n",
      "\n",
      "                  padj  \n",
      "feature_name            \n",
      "A1BG          0.994760  \n",
      "A1BG-AS1           NaN  \n",
      "A2M           0.000142  \n",
      "A2M-AS1            NaN  \n",
      "A2ML1              NaN  \n",
      "...                ...  \n",
      "ZXDC          0.523405  \n",
      "ZYG11A             NaN  \n",
      "ZYG11B        0.767087  \n",
      "ZYX           0.887347  \n",
      "ZZEF1         0.780258  \n",
      "\n",
      "[17811 rows x 6 columns]\n",
      "epithelial cell\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Fitting size factors...\n",
      "... done in 0.02 seconds.\n",
      "\n",
      "Fitting dispersions...\n",
      "... done in 1.35 seconds.\n",
      "\n",
      "Fitting dispersion trend curve...\n",
      "... done in 0.39 seconds.\n",
      "\n",
      "Fitting MAP dispersions...\n",
      "... done in 1.49 seconds.\n",
      "\n",
      "Fitting LFCs...\n",
      "... done in 1.53 seconds.\n",
      "\n",
      "Calculating cook's distance...\n",
      "... done in 0.05 seconds.\n",
      "\n",
      "Replacing 317 outlier genes.\n",
      "\n",
      "Fitting dispersions...\n",
      "... done in 0.10 seconds.\n",
      "\n",
      "Fitting MAP dispersions...\n",
      "... done in 0.09 seconds.\n",
      "\n",
      "Fitting LFCs...\n",
      "... done in 0.10 seconds.\n",
      "\n",
      "Running Wald tests...\n",
      "... done in 8.88 seconds.\n",
      "\n",
      "Fitting MAP LFCs...\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Log2 fold change & Wald test p-value: origin tumor-primary vs normal-adjacent\n",
      "              baseMean  log2FoldChange     lfcSE      stat    pvalue      padj\n",
      "feature_name                                                                  \n",
      "A1BG          9.640422       -0.084555  0.506829 -0.166832  0.867502  0.931543\n",
      "A1BG-AS1      2.222426       -0.811751  0.570795 -1.422142  0.154985  0.337826\n",
      "A2M           2.187057        0.683752  0.561851  1.216964  0.223618  0.429649\n",
      "A2M-AS1       0.366523        0.159750  1.005975  0.158801  0.873826       NaN\n",
      "A2ML1         0.034975        0.105248  3.107701  0.033867  0.972983       NaN\n",
      "...                ...             ...       ...       ...       ...       ...\n",
      "ZXDC          3.827585       -0.355821  0.301693 -1.179414  0.238234  0.446787\n",
      "ZYG11A        0.142235        0.434760  1.760640  0.246933  0.804960       NaN\n",
      "ZYG11B        7.112276        0.099663  0.270380  0.368605  0.712422  0.841209\n",
      "ZYX           5.964752        0.788411  0.361182  2.182865  0.029046  0.106924\n",
      "ZZEF1         3.524595       -0.545854  0.321310 -1.698837  0.089350  0.233934\n",
      "\n",
      "[17811 rows x 6 columns]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "... done in 1.75 seconds.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Shrunk log2 fold change & Wald test p-value: origin tumor-primary vs normal-adjacent\n",
      "              baseMean  log2FoldChange     lfcSE      stat    pvalue      padj\n",
      "feature_name                                                                  \n",
      "A1BG          9.640422       -0.032273  0.381509 -0.166832  0.867502  0.931543\n",
      "A1BG-AS1      2.222426       -0.300044  0.467179 -1.422142  0.154985  0.337826\n",
      "A2M           2.187057        0.267037  0.469288  1.216964  0.223618  0.429649\n",
      "A2M-AS1       0.366523        0.019218  0.497158  0.158801  0.873826       NaN\n",
      "A2ML1         0.034975       -0.028946  0.468061  0.033867  0.972983       NaN\n",
      "...                ...             ...       ...       ...       ...       ...\n",
      "ZXDC          3.827585       -0.243915  0.279220 -1.179414  0.238234  0.446787\n",
      "ZYG11A        0.142235        0.013593  0.590206  0.246933  0.804960       NaN\n",
      "ZYG11B        7.112276        0.069289  0.248130  0.368605  0.712422  0.841209\n",
      "ZYX           5.964752        0.559688  0.362183  2.182865  0.029046  0.106924\n",
      "ZZEF1         3.524595       -0.380495  0.308373 -1.698837  0.089350  0.233934\n",
      "\n",
      "[17811 rows x 6 columns]\n",
      "fibroblast\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Fitting size factors...\n",
      "... done in 0.02 seconds.\n",
      "\n",
      "Fitting dispersions...\n",
      "... done in 1.21 seconds.\n",
      "\n",
      "Fitting dispersion trend curve...\n",
      "... done in 0.37 seconds.\n",
      "\n",
      "Fitting MAP dispersions...\n",
      "... done in 1.51 seconds.\n",
      "\n",
      "Fitting LFCs...\n",
      "... done in 1.48 seconds.\n",
      "\n",
      "Calculating cook's distance...\n",
      "... done in 0.03 seconds.\n",
      "\n",
      "Replacing 129 outlier genes.\n",
      "\n",
      "Fitting dispersions...\n",
      "... done in 0.03 seconds.\n",
      "\n",
      "Fitting MAP dispersions...\n",
      "... done in 0.03 seconds.\n",
      "\n",
      "Fitting LFCs...\n",
      "... done in 0.04 seconds.\n",
      "\n",
      "Running Wald tests...\n",
      "... done in 8.63 seconds.\n",
      "\n",
      "Fitting MAP LFCs...\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Log2 fold change & Wald test p-value: origin tumor-primary vs normal-adjacent\n",
      "                baseMean  log2FoldChange     lfcSE      stat    pvalue  \\\n",
      "feature_name                                                             \n",
      "A1BG            6.599187        1.049944  0.505666  2.076356  0.037861   \n",
      "A1BG-AS1        0.456214        0.669571  0.837132  0.799839  0.423804   \n",
      "A2M           497.023760       -0.200188  0.392714 -0.509755  0.610223   \n",
      "A2M-AS1         0.405584       -0.270720  0.764998 -0.353883  0.723427   \n",
      "A2ML1           0.002863       -0.382078  3.204995 -0.119213  0.905106   \n",
      "...                  ...             ...       ...       ...       ...   \n",
      "ZXDC            1.776901       -0.167808  0.553695 -0.303070  0.761837   \n",
      "ZYG11A          0.010638       -0.620068  3.191379 -0.194295  0.845945   \n",
      "ZYG11B          5.866106       -0.055193  0.358517 -0.153949  0.877650   \n",
      "ZYX            37.964425        0.020617  0.263309  0.078299  0.937590   \n",
      "ZZEF1           2.391990        0.619780  0.459674  1.348301  0.177561   \n",
      "\n",
      "                  padj  \n",
      "feature_name            \n",
      "A1BG          0.209989  \n",
      "A1BG-AS1           NaN  \n",
      "A2M           0.846202  \n",
      "A2M-AS1            NaN  \n",
      "A2ML1              NaN  \n",
      "...                ...  \n",
      "ZXDC          0.920513  \n",
      "ZYG11A             NaN  \n",
      "ZYG11B        0.964336  \n",
      "ZYX           0.979594  \n",
      "ZZEF1         0.499176  \n",
      "\n",
      "[17811 rows x 6 columns]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/kvalem/.conda/envs/crc_atlas/lib/python3.11/site-packages/pydeseq2/utils.py:1260: RuntimeWarning: overflow encountered in exp\n",
      "  counts - (counts + size) / (1 + size * np.exp(-xbeta - offset))\n",
      "/home/kvalem/.conda/envs/crc_atlas/lib/python3.11/site-packages/pydeseq2/utils.py:1260: RuntimeWarning: overflow encountered in exp\n",
      "  counts - (counts + size) / (1 + size * np.exp(-xbeta - offset))\n",
      "/home/kvalem/.conda/envs/crc_atlas/lib/python3.11/site-packages/pydeseq2/utils.py:1260: RuntimeWarning: overflow encountered in exp\n",
      "  counts - (counts + size) / (1 + size * np.exp(-xbeta - offset))\n",
      "/home/kvalem/.conda/envs/crc_atlas/lib/python3.11/site-packages/pydeseq2/utils.py:1260: RuntimeWarning: overflow encountered in exp\n",
      "  counts - (counts + size) / (1 + size * np.exp(-xbeta - offset))\n",
      "... done in 2.95 seconds.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Shrunk log2 fold change & Wald test p-value: origin tumor-primary vs normal-adjacent\n",
      "                baseMean  log2FoldChange     lfcSE      stat    pvalue  \\\n",
      "feature_name                                                             \n",
      "A1BG            6.599187    4.218071e-06  0.001870  2.076356  0.037861   \n",
      "A1BG-AS1        0.456214    1.481234e-06  0.001874  0.799839  0.423804   \n",
      "A2M           497.023760   -1.782393e-06  0.001507 -0.509755  0.610223   \n",
      "A2M-AS1         0.405584   -3.363530e-07  0.001550 -0.353883  0.723427   \n",
      "A2ML1           0.002863    6.766228e-08  0.002404 -0.119213  0.905106   \n",
      "...                  ...             ...       ...       ...       ...   \n",
      "ZXDC            1.776901   -4.230252e-07  0.001566 -0.303070  0.761837   \n",
      "ZYG11A          0.010638   -1.701478e-07  0.001192 -0.194295  0.845945   \n",
      "ZYG11B          5.866106   -4.316333e-07  0.001582 -0.153949  0.877650   \n",
      "ZYX            37.964425    3.254124e-07  0.001589  0.078299  0.937590   \n",
      "ZZEF1           2.391990    3.207554e-06  0.001671  1.348301  0.177561   \n",
      "\n",
      "                  padj  \n",
      "feature_name            \n",
      "A1BG          0.209989  \n",
      "A1BG-AS1           NaN  \n",
      "A2M           0.846202  \n",
      "A2M-AS1            NaN  \n",
      "A2ML1              NaN  \n",
      "...                ...  \n",
      "ZXDC          0.920513  \n",
      "ZYG11A             NaN  \n",
      "ZYG11B        0.964336  \n",
      "ZYX           0.979594  \n",
      "ZZEF1         0.499176  \n",
      "\n",
      "[17811 rows x 6 columns]\n",
      "macrophage\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Fitting size factors...\n",
      "... done in 0.03 seconds.\n",
      "\n",
      "Fitting dispersions...\n",
      "... done in 1.53 seconds.\n",
      "\n",
      "Fitting dispersion trend curve...\n",
      "/home/kvalem/.conda/envs/crc_atlas/lib/python3.11/site-packages/pydeseq2/dds.py:716: UserWarning: The dispersion trend curve fitting did not converge. Switching to a mean-based dispersion trend.\n",
      "  self._fit_parametric_dispersion_trend(vst)\n",
      "... done in 0.35 seconds.\n",
      "\n",
      "Fitting MAP dispersions...\n",
      "... done in 1.89 seconds.\n",
      "\n",
      "Fitting LFCs...\n",
      "... done in 1.18 seconds.\n",
      "\n",
      "Calculating cook's distance...\n",
      "... done in 0.07 seconds.\n",
      "\n",
      "Replacing 116 outlier genes.\n",
      "\n",
      "Fitting dispersions...\n",
      "... done in 0.04 seconds.\n",
      "\n",
      "Fitting MAP dispersions...\n",
      "... done in 0.03 seconds.\n",
      "\n",
      "Fitting LFCs...\n",
      "... done in 0.03 seconds.\n",
      "\n",
      "Running Wald tests...\n",
      "... done in 8.95 seconds.\n",
      "\n",
      "Fitting MAP LFCs...\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Log2 fold change & Wald test p-value: origin tumor-primary vs normal-adjacent\n",
      "                baseMean  log2FoldChange     lfcSE      stat    pvalue  \\\n",
      "feature_name                                                             \n",
      "A1BG           85.516680        0.353116  0.267730  1.318924  0.187194   \n",
      "A1BG-AS1       15.799651        0.184641  0.149826  1.232368  0.217812   \n",
      "A2M           552.829138        1.522115  0.321093  4.740417  0.000002   \n",
      "A2M-AS1         5.048961        0.212698  0.367639  0.578550  0.562893   \n",
      "A2ML1           0.159131        0.310714  0.851395  0.364947  0.715151   \n",
      "...                  ...             ...       ...       ...       ...   \n",
      "ZXDC           21.092824       -0.220321  0.171323 -1.285999  0.198443   \n",
      "ZYG11A          0.127500        0.500826  0.946272  0.529262  0.596624   \n",
      "ZYG11B         84.127390       -0.360350  0.116007 -3.106279  0.001895   \n",
      "ZYX           450.161470        0.048292  0.134613  0.358744  0.719786   \n",
      "ZZEF1          47.326305       -0.042285  0.110813 -0.381595  0.702762   \n",
      "\n",
      "                  padj  \n",
      "feature_name            \n",
      "A1BG          0.355304  \n",
      "A1BG-AS1      0.395544  \n",
      "A2M           0.000031  \n",
      "A2M-AS1       0.726014  \n",
      "A2ML1              NaN  \n",
      "...                ...  \n",
      "ZXDC          0.370071  \n",
      "ZYG11A             NaN  \n",
      "ZYG11B        0.009996  \n",
      "ZYX           0.838034  \n",
      "ZZEF1         0.827546  \n",
      "\n",
      "[17811 rows x 6 columns]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "... done in 1.84 seconds.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Shrunk log2 fold change & Wald test p-value: origin tumor-primary vs normal-adjacent\n",
      "                baseMean  log2FoldChange     lfcSE      stat    pvalue  \\\n",
      "feature_name                                                             \n",
      "A1BG           85.516680        0.259216  0.254724  1.318924  0.187194   \n",
      "A1BG-AS1       15.799651        0.162138  0.146756  1.232368  0.217812   \n",
      "A2M           552.829138        1.415559  0.328612  4.740417  0.000002   \n",
      "A2M-AS1         5.048961        0.106252  0.315818  0.578550  0.562893   \n",
      "A2ML1           0.159131       -0.080275  0.469801  0.364947  0.715151   \n",
      "...                  ...             ...       ...       ...       ...   \n",
      "ZXDC           21.092824       -0.191127  0.164089 -1.285999  0.198443   \n",
      "ZYG11A          0.127500        0.010762  0.472897  0.529262  0.596624   \n",
      "ZYG11B         84.127390       -0.339615  0.114463 -3.106279  0.001895   \n",
      "ZYX           450.161470        0.051654  0.130568  0.358744  0.719786   \n",
      "ZZEF1          47.326305       -0.039404  0.108910 -0.381595  0.702762   \n",
      "\n",
      "                  padj  \n",
      "feature_name            \n",
      "A1BG          0.355304  \n",
      "A1BG-AS1      0.395544  \n",
      "A2M           0.000031  \n",
      "A2M-AS1       0.726014  \n",
      "A2ML1              NaN  \n",
      "...                ...  \n",
      "ZXDC          0.370071  \n",
      "ZYG11A             NaN  \n",
      "ZYG11B        0.009996  \n",
      "ZYX           0.838034  \n",
      "ZZEF1         0.827546  \n",
      "\n",
      "[17811 rows x 6 columns]\n",
      "malignant cell\n",
      "Not running DEseq for: malignant cell : only present in: set()\n",
      "neutrophil\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/kvalem/.local/lib/python3.11/site-packages/anndata/_core/aligned_df.py:67: ImplicitModificationWarning: Transforming to str index.\n",
      "  warnings.warn(\"Transforming to str index.\", ImplicitModificationWarning)\n",
      "Fitting size factors...\n",
      "... done in 0.01 seconds.\n",
      "\n",
      "Fitting dispersions...\n",
      "... done in 0.92 seconds.\n",
      "\n",
      "Fitting dispersion trend curve...\n",
      "... done in 0.32 seconds.\n",
      "\n",
      "Fitting MAP dispersions...\n",
      "... done in 0.96 seconds.\n",
      "\n",
      "Fitting LFCs...\n",
      "... done in 1.11 seconds.\n",
      "\n",
      "Calculating cook's distance...\n",
      "... done in 0.01 seconds.\n",
      "\n",
      "Replacing 0 outlier genes.\n",
      "\n",
      "Running Wald tests...\n",
      "... done in 8.77 seconds.\n",
      "\n",
      "Fitting MAP LFCs...\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Log2 fold change & Wald test p-value: origin tumor-primary vs normal-adjacent\n",
      "                baseMean  log2FoldChange     lfcSE      stat    pvalue  \\\n",
      "feature_name                                                             \n",
      "A1BG            1.219286       -1.271715  2.330923 -0.545584  0.585352   \n",
      "A1BG-AS1        0.000000             NaN       NaN       NaN       NaN   \n",
      "A2M             4.403454        0.032516  1.572091  0.020683  0.983498   \n",
      "A2M-AS1         0.394540        0.647985  2.970378  0.218149  0.827313   \n",
      "A2ML1           0.000000             NaN       NaN       NaN       NaN   \n",
      "...                  ...             ...       ...       ...       ...   \n",
      "ZXDC            7.328170       -0.284063  0.693867 -0.409392  0.682252   \n",
      "ZYG11A          0.181016        2.184656  3.825929  0.571013  0.567991   \n",
      "ZYG11B         17.418452       -0.466853  0.546436 -0.854360  0.392905   \n",
      "ZYX           228.603724       -0.275976  0.352594 -0.782704  0.433801   \n",
      "ZZEF1          11.721639       -0.534696  0.675023 -0.792115  0.428293   \n",
      "\n",
      "                  padj  \n",
      "feature_name            \n",
      "A1BG          0.999748  \n",
      "A1BG-AS1           NaN  \n",
      "A2M           0.999748  \n",
      "A2M-AS1       0.999748  \n",
      "A2ML1              NaN  \n",
      "...                ...  \n",
      "ZXDC          0.999748  \n",
      "ZYG11A        0.999748  \n",
      "ZYG11B        0.999748  \n",
      "ZYX           0.999748  \n",
      "ZZEF1         0.999748  \n",
      "\n",
      "[17811 rows x 6 columns]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/kvalem/.conda/envs/crc_atlas/lib/python3.11/site-packages/pydeseq2/utils.py:1260: RuntimeWarning: overflow encountered in exp\n",
      "  counts - (counts + size) / (1 + size * np.exp(-xbeta - offset))\n",
      "/home/kvalem/.conda/envs/crc_atlas/lib/python3.11/site-packages/pydeseq2/utils.py:1260: RuntimeWarning: overflow encountered in exp\n",
      "  counts - (counts + size) / (1 + size * np.exp(-xbeta - offset))\n",
      "/home/kvalem/.conda/envs/crc_atlas/lib/python3.11/site-packages/pydeseq2/utils.py:1260: RuntimeWarning: overflow encountered in exp\n",
      "  counts - (counts + size) / (1 + size * np.exp(-xbeta - offset))\n",
      "... done in 2.87 seconds.\n",
      "\n",
      "Fitting size factors...\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Shrunk log2 fold change & Wald test p-value: origin tumor-primary vs normal-adjacent\n",
      "                baseMean  log2FoldChange     lfcSE      stat    pvalue  \\\n",
      "feature_name                                                             \n",
      "A1BG            1.219286   -2.184719e-07  0.001225 -0.545584  0.585352   \n",
      "A1BG-AS1        0.000000             NaN       NaN       NaN       NaN   \n",
      "A2M             4.403454    1.536174e-08  0.001393  0.020683  0.983498   \n",
      "A2M-AS1         0.394540    1.373664e-08  0.001455  0.218149  0.827313   \n",
      "A2ML1           0.000000             NaN       NaN       NaN       NaN   \n",
      "...                  ...             ...       ...       ...       ...   \n",
      "ZXDC            7.328170   -6.326384e-07  0.001344 -0.409392  0.682252   \n",
      "ZYG11A          0.181016    2.100066e-07  0.002789  0.571013  0.567991   \n",
      "ZYG11B         17.418452   -1.604203e-06  0.001316 -0.854360  0.392905   \n",
      "ZYX           228.603724   -2.295424e-06  0.001372 -0.782704  0.433801   \n",
      "ZZEF1          11.721639   -1.229991e-06  0.001313 -0.792115  0.428293   \n",
      "\n",
      "                  padj  \n",
      "feature_name            \n",
      "A1BG          0.999748  \n",
      "A1BG-AS1           NaN  \n",
      "A2M           0.999748  \n",
      "A2M-AS1       0.999748  \n",
      "A2ML1              NaN  \n",
      "...                ...  \n",
      "ZXDC          0.999748  \n",
      "ZYG11A        0.999748  \n",
      "ZYG11B        0.999748  \n",
      "ZYX           0.999748  \n",
      "ZZEF1         0.999748  \n",
      "\n",
      "[17811 rows x 6 columns]\n",
      "plasma cell\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "... done in 0.02 seconds.\n",
      "\n",
      "Fitting dispersions...\n",
      "... done in 1.24 seconds.\n",
      "\n",
      "Fitting dispersion trend curve...\n",
      "... done in 0.38 seconds.\n",
      "\n",
      "Fitting MAP dispersions...\n",
      "... done in 1.41 seconds.\n",
      "\n",
      "Fitting LFCs...\n",
      "... done in 1.30 seconds.\n",
      "\n",
      "Calculating cook's distance...\n",
      "... done in 0.05 seconds.\n",
      "\n",
      "Replacing 104 outlier genes.\n",
      "\n",
      "Fitting dispersions...\n",
      "... done in 0.03 seconds.\n",
      "\n",
      "Fitting MAP dispersions...\n",
      "... done in 0.03 seconds.\n",
      "\n",
      "Fitting LFCs...\n",
      "... done in 0.04 seconds.\n",
      "\n",
      "Running Wald tests...\n",
      "... done in 8.85 seconds.\n",
      "\n",
      "Fitting MAP LFCs...\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Log2 fold change & Wald test p-value: origin tumor-primary vs normal-adjacent\n",
      "               baseMean  log2FoldChange     lfcSE      stat    pvalue  \\\n",
      "feature_name                                                            \n",
      "A1BG          20.520586       -0.359202  0.336171 -1.068509  0.285291   \n",
      "A1BG-AS1       1.088761       -0.399286  0.651671 -0.612711  0.540068   \n",
      "A2M            3.439281        0.355859  0.678731  0.524300  0.600070   \n",
      "A2M-AS1        0.140953       -1.599917  1.999993 -0.799961  0.423733   \n",
      "A2ML1          0.062925       -1.613224  2.929489 -0.550685  0.581850   \n",
      "...                 ...             ...       ...       ...       ...   \n",
      "ZXDC           0.549366       -0.570483  0.833440 -0.684492  0.493664   \n",
      "ZYG11A         0.347641       -1.367677  0.925531 -1.477722  0.139482   \n",
      "ZYG11B         2.327975       -0.417588  0.556814 -0.749959  0.453279   \n",
      "ZYX            3.123633       -0.819743  0.402255 -2.037868  0.041563   \n",
      "ZZEF1          2.614562        0.031174  0.454054  0.068657  0.945263   \n",
      "\n",
      "                  padj  \n",
      "feature_name            \n",
      "A1BG          0.847459  \n",
      "A1BG-AS1           NaN  \n",
      "A2M           0.944497  \n",
      "A2M-AS1            NaN  \n",
      "A2ML1              NaN  \n",
      "...                ...  \n",
      "ZXDC               NaN  \n",
      "ZYG11A             NaN  \n",
      "ZYG11B        0.909793  \n",
      "ZYX           0.500504  \n",
      "ZZEF1         0.998045  \n",
      "\n",
      "[17811 rows x 6 columns]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/kvalem/.conda/envs/crc_atlas/lib/python3.11/site-packages/pydeseq2/utils.py:1260: RuntimeWarning: overflow encountered in exp\n",
      "  counts - (counts + size) / (1 + size * np.exp(-xbeta - offset))\n",
      "/home/kvalem/.conda/envs/crc_atlas/lib/python3.11/site-packages/pydeseq2/utils.py:1260: RuntimeWarning: overflow encountered in exp\n",
      "  counts - (counts + size) / (1 + size * np.exp(-xbeta - offset))\n",
      "/home/kvalem/.conda/envs/crc_atlas/lib/python3.11/site-packages/pydeseq2/utils.py:1260: RuntimeWarning: overflow encountered in exp\n",
      "  counts - (counts + size) / (1 + size * np.exp(-xbeta - offset))\n",
      "/home/kvalem/.conda/envs/crc_atlas/lib/python3.11/site-packages/pydeseq2/utils.py:1260: RuntimeWarning: overflow encountered in exp\n",
      "  counts - (counts + size) / (1 + size * np.exp(-xbeta - offset))\n",
      "/home/kvalem/.conda/envs/crc_atlas/lib/python3.11/site-packages/pydeseq2/utils.py:1260: RuntimeWarning: overflow encountered in exp\n",
      "  counts - (counts + size) / (1 + size * np.exp(-xbeta - offset))\n",
      "/home/kvalem/.conda/envs/crc_atlas/lib/python3.11/site-packages/pydeseq2/utils.py:1260: RuntimeWarning: overflow encountered in exp\n",
      "  counts - (counts + size) / (1 + size * np.exp(-xbeta - offset))\n",
      "/home/kvalem/.conda/envs/crc_atlas/lib/python3.11/site-packages/pydeseq2/utils.py:1260: RuntimeWarning: overflow encountered in exp\n",
      "  counts - (counts + size) / (1 + size * np.exp(-xbeta - offset))\n",
      "... done in 3.21 seconds.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Shrunk log2 fold change & Wald test p-value: origin tumor-primary vs normal-adjacent\n",
      "               baseMean  log2FoldChange     lfcSE      stat    pvalue  \\\n",
      "feature_name                                                            \n",
      "A1BG          20.520586   -3.330070e-06  0.001719 -1.068509  0.285291   \n",
      "A1BG-AS1       1.088761   -7.542754e-07  0.002410 -0.612711  0.540068   \n",
      "A2M            3.439281    7.998650e-07  0.002042  0.524300  0.600070   \n",
      "A2M-AS1        0.140953   -6.126836e-07  0.001516 -0.799961  0.423733   \n",
      "A2ML1          0.062925   -3.529262e-07  0.001321 -0.550685  0.581850   \n",
      "...                 ...             ...       ...       ...       ...   \n",
      "ZXDC           0.549366   -6.343671e-07  0.002370 -0.684492  0.493664   \n",
      "ZYG11A         0.347641   -1.513020e-06  0.002137 -1.477722  0.139482   \n",
      "ZYG11B         2.327975   -1.359097e-06  0.002098 -0.749959  0.453279   \n",
      "ZYX            3.123633   -4.781257e-06  0.002172 -2.037868  0.041563   \n",
      "ZZEF1          2.614562    1.738228e-07  0.002460  0.068657  0.945263   \n",
      "\n",
      "                  padj  \n",
      "feature_name            \n",
      "A1BG          0.847459  \n",
      "A1BG-AS1           NaN  \n",
      "A2M           0.944497  \n",
      "A2M-AS1            NaN  \n",
      "A2ML1              NaN  \n",
      "...                ...  \n",
      "ZXDC               NaN  \n",
      "ZYG11A             NaN  \n",
      "ZYG11B        0.909793  \n",
      "ZYX           0.500504  \n",
      "ZZEF1         0.998045  \n",
      "\n",
      "[17811 rows x 6 columns]\n",
      "regulatory T cell\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Fitting size factors...\n",
      "... done in 0.03 seconds.\n",
      "\n",
      "Fitting dispersions...\n",
      "... done in 1.53 seconds.\n",
      "\n",
      "Fitting dispersion trend curve...\n",
      "... done in 0.36 seconds.\n",
      "\n",
      "Fitting MAP dispersions...\n",
      "... done in 1.43 seconds.\n",
      "\n",
      "Fitting LFCs...\n",
      "... done in 1.54 seconds.\n",
      "\n",
      "Calculating cook's distance...\n",
      "... done in 0.06 seconds.\n",
      "\n",
      "Replacing 1377 outlier genes.\n",
      "\n",
      "Fitting dispersions...\n",
      "... done in 0.17 seconds.\n",
      "\n",
      "Fitting MAP dispersions...\n",
      "... done in 0.15 seconds.\n",
      "\n",
      "Fitting LFCs...\n",
      "... done in 0.14 seconds.\n",
      "\n",
      "Running Wald tests...\n",
      "... done in 8.75 seconds.\n",
      "\n",
      "Fitting MAP LFCs...\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Log2 fold change & Wald test p-value: origin tumor-primary vs normal-adjacent\n",
      "               baseMean  log2FoldChange     lfcSE      stat    pvalue  \\\n",
      "feature_name                                                            \n",
      "A1BG           7.033111       -0.142249  0.394206 -0.360849  0.718212   \n",
      "A1BG-AS1       1.485339        0.159139  0.623456  0.255254  0.798527   \n",
      "A2M            1.273227        0.290248  0.728302  0.398527  0.690242   \n",
      "A2M-AS1        0.173113       -1.321383  1.405063 -0.940444  0.346990   \n",
      "A2ML1          0.176934       -2.254269  2.922899 -0.771244  0.440562   \n",
      "...                 ...             ...       ...       ...       ...   \n",
      "ZXDC           1.019239        0.850460  0.792190  1.073555  0.283022   \n",
      "ZYG11A         0.014163       -2.067385  3.053920 -0.676961  0.498431   \n",
      "ZYG11B         3.356246       -0.176350  0.442509 -0.398524  0.690244   \n",
      "ZYX           14.740712       -0.403523  0.279140 -1.445596  0.148290   \n",
      "ZZEF1          3.895481        0.452235  0.394207  1.147203  0.251298   \n",
      "\n",
      "                  padj  \n",
      "feature_name            \n",
      "A1BG          0.963157  \n",
      "A1BG-AS1           NaN  \n",
      "A2M                NaN  \n",
      "A2M-AS1            NaN  \n",
      "A2ML1              NaN  \n",
      "...                ...  \n",
      "ZXDC               NaN  \n",
      "ZYG11A             NaN  \n",
      "ZYG11B        0.958322  \n",
      "ZYX           0.708795  \n",
      "ZZEF1         0.803712  \n",
      "\n",
      "[17811 rows x 6 columns]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/kvalem/.conda/envs/crc_atlas/lib/python3.11/site-packages/pydeseq2/utils.py:1260: RuntimeWarning: overflow encountered in exp\n",
      "  counts - (counts + size) / (1 + size * np.exp(-xbeta - offset))\n",
      "/home/kvalem/.conda/envs/crc_atlas/lib/python3.11/site-packages/pydeseq2/utils.py:1260: RuntimeWarning: overflow encountered in exp\n",
      "  counts - (counts + size) / (1 + size * np.exp(-xbeta - offset))\n",
      "/home/kvalem/.conda/envs/crc_atlas/lib/python3.11/site-packages/pydeseq2/utils.py:1260: RuntimeWarning: overflow encountered in exp\n",
      "  counts - (counts + size) / (1 + size * np.exp(-xbeta - offset))\n",
      "/home/kvalem/.conda/envs/crc_atlas/lib/python3.11/site-packages/pydeseq2/utils.py:1260: RuntimeWarning: overflow encountered in exp\n",
      "  counts - (counts + size) / (1 + size * np.exp(-xbeta - offset))\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Shrunk log2 fold change & Wald test p-value: origin tumor-primary vs normal-adjacent\n",
      "               baseMean  log2FoldChange     lfcSE      stat    pvalue  \\\n",
      "feature_name                                                            \n",
      "A1BG           7.033111   -9.155528e-07  0.001640 -0.360849  0.718212   \n",
      "A1BG-AS1       1.485339    6.611014e-07  0.001920  0.255254  0.798527   \n",
      "A2M            1.273227    8.387755e-07  0.001885  0.398527  0.690242   \n",
      "A2M-AS1        0.173113    1.044228e-07  0.002154 -0.940444  0.346990   \n",
      "A2ML1          0.176934   -7.613284e-08  0.001425 -0.771244  0.440562   \n",
      "...                 ...             ...       ...       ...       ...   \n",
      "ZXDC           1.019239    2.913140e-06  0.002257  1.073555  0.283022   \n",
      "ZYG11A         0.014163    1.193448e-07  0.002688 -0.676961  0.498431   \n",
      "ZYG11B         3.356246   -9.622742e-07  0.001682 -0.398524  0.690244   \n",
      "ZYX           14.740712   -4.786789e-06  0.001580 -1.445596  0.148290   \n",
      "ZZEF1          3.895481    3.102382e-06  0.001964  1.147203  0.251298   \n",
      "\n",
      "                  padj  \n",
      "feature_name            \n",
      "A1BG          0.963157  \n",
      "A1BG-AS1           NaN  \n",
      "A2M                NaN  \n",
      "A2M-AS1            NaN  \n",
      "A2ML1              NaN  \n",
      "...                ...  \n",
      "ZXDC               NaN  \n",
      "ZYG11A             NaN  \n",
      "ZYG11B        0.958322  \n",
      "ZYX           0.708795  \n",
      "ZZEF1         0.803712  \n",
      "\n",
      "[17811 rows x 6 columns]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "... done in 3.28 seconds.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "import pdb\n",
    "for contrast in contrasts:\n",
    "    de_res = {}\n",
    "\n",
    "    for ct in cell_types:\n",
    "        print(ct)\n",
    "        \n",
    "        pb_ct = pdata_tumor_normal_male[pdata_tumor_normal_male.obs[cell_type_class] == ct].copy()\n",
    "\n",
    "     \n",
    "        #pdb.set_trace()\n",
    "        if len(set(pb_ct.obs[\"origin\"]).intersection([contrast[\"reference\"], contrast[\"condition\"]])) < 1:\n",
    "             print(\n",
    "                 \"Not running DEseq for: \"\n",
    "                  + ct\n",
    "                  + \" : only present in: \"\n",
    "                  + str(set(pb_ct.obs[\"sex\"]).intersection([contrast[\"reference\"], contrast[\"condition\"]]))\n",
    "             )\n",
    "             continue\n",
    "        #pdb.set_trace()\n",
    "                     \n",
    "        dds = DeseqDataSet(\n",
    "           adata=pb_ct,\n",
    "           design_factors=[contrast[\"var\"]],\n",
    "           ref_level=[contrast[\"var\"], 'normal-adjacent'],\n",
    "           refit_cooks=True,\n",
    "           n_cpus=cpus,\n",
    "           )\n",
    "             \n",
    "        \n",
    "\n",
    "        # Compute LFCs\n",
    "        dds.deseq2()\n",
    "\n",
    " \n",
    "\n",
    "        \n",
    "\n",
    "        # Extract contrast\n",
    "        stat_res = DeseqStats(\n",
    "              dds,\n",
    "              contrast=[contrast[\"var\"],'tumor-primary', 'normal-adjacent'],\n",
    "              #n_cpus=cpus,\n",
    "          )\n",
    "   \n",
    "    \n",
    "\n",
    "        # Compute Wald test\n",
    "        stat_res.summary()\n",
    "\n",
    "        # Shrink LFCs\n",
    "\n",
    "        coeff = contrast[\"var\"] + \"_\" + contrast[\"name\"]\n",
    "        stat_res.lfc_shrink(coeff=coeff)\n",
    "\n",
    "        # Register cell type results\n",
    "        de_res[ct] = stat_res.results_df\n",
    "        de_res[ct][\"standardized_cell_type\"] = ct\n",
    "        de_res[ct][\"feature_name\"] = stat_res.results_df.index.values\n",
    "\n",
    "        de_res[ct].drop(columns=[\"feature_name\"]).to_csv(\n",
    "                Path(\n",
    "                    contrast[\"res_dir\"],\n",
    "                    ct.replace(\" \", \"_\") + \"_\" + contrast[\"name\"] + \"_deseq.tsv\",\n",
    "                ),\n",
    "                sep=\"\\t\",\n",
    "            )\n",
    "\n",
    "        # Register results for current contrast\n",
    "        contrast[\"de_res\"] = de_res\n",
    "        contrast[\"de_res_all\"] = pd.concat([df.assign(cell_type=ct) for ct, df in de_res.items()])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "id": "3dc71515-7ccf-4e41-907b-060df5511162",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>baseMean</th>\n",
       "      <th>log2FoldChange</th>\n",
       "      <th>lfcSE</th>\n",
       "      <th>stat</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>padj</th>\n",
       "      <th>standardized_cell_type</th>\n",
       "      <th>feature_name</th>\n",
       "      <th>cell_type</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>feature_name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A1BG</th>\n",
       "      <td>4.613264</td>\n",
       "      <td>5.484595e-06</td>\n",
       "      <td>0.052210</td>\n",
       "      <td>0.000174</td>\n",
       "      <td>0.999861</td>\n",
       "      <td>0.999978</td>\n",
       "      <td>B cell</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>B cell</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1BG-AS1</th>\n",
       "      <td>0.591671</td>\n",
       "      <td>9.067974e-05</td>\n",
       "      <td>0.063279</td>\n",
       "      <td>-0.351676</td>\n",
       "      <td>0.725081</td>\n",
       "      <td>NaN</td>\n",
       "      <td>B cell</td>\n",
       "      <td>A1BG-AS1</td>\n",
       "      <td>B cell</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2M</th>\n",
       "      <td>1.206051</td>\n",
       "      <td>-6.738284e-04</td>\n",
       "      <td>0.053568</td>\n",
       "      <td>-0.360525</td>\n",
       "      <td>0.718455</td>\n",
       "      <td>NaN</td>\n",
       "      <td>B cell</td>\n",
       "      <td>A2M</td>\n",
       "      <td>B cell</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2M-AS1</th>\n",
       "      <td>0.082406</td>\n",
       "      <td>1.239167e-04</td>\n",
       "      <td>0.060909</td>\n",
       "      <td>-0.815760</td>\n",
       "      <td>0.414638</td>\n",
       "      <td>NaN</td>\n",
       "      <td>B cell</td>\n",
       "      <td>A2M-AS1</td>\n",
       "      <td>B cell</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2ML1</th>\n",
       "      <td>0.157295</td>\n",
       "      <td>-4.821732e-04</td>\n",
       "      <td>0.046209</td>\n",
       "      <td>-1.113721</td>\n",
       "      <td>0.265399</td>\n",
       "      <td>NaN</td>\n",
       "      <td>B cell</td>\n",
       "      <td>A2ML1</td>\n",
       "      <td>B cell</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZXDC</th>\n",
       "      <td>1.019239</td>\n",
       "      <td>2.913140e-06</td>\n",
       "      <td>0.002257</td>\n",
       "      <td>1.073555</td>\n",
       "      <td>0.283022</td>\n",
       "      <td>NaN</td>\n",
       "      <td>regulatory T cell</td>\n",
       "      <td>ZXDC</td>\n",
       "      <td>regulatory T cell</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZYG11A</th>\n",
       "      <td>0.014163</td>\n",
       "      <td>1.193448e-07</td>\n",
       "      <td>0.002688</td>\n",
       "      <td>-0.676961</td>\n",
       "      <td>0.498431</td>\n",
       "      <td>NaN</td>\n",
       "      <td>regulatory T cell</td>\n",
       "      <td>ZYG11A</td>\n",
       "      <td>regulatory T cell</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZYG11B</th>\n",
       "      <td>3.356246</td>\n",
       "      <td>-9.622742e-07</td>\n",
       "      <td>0.001682</td>\n",
       "      <td>-0.398524</td>\n",
       "      <td>0.690244</td>\n",
       "      <td>0.958322</td>\n",
       "      <td>regulatory T cell</td>\n",
       "      <td>ZYG11B</td>\n",
       "      <td>regulatory T cell</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZYX</th>\n",
       "      <td>14.740712</td>\n",
       "      <td>-4.786789e-06</td>\n",
       "      <td>0.001580</td>\n",
       "      <td>-1.445596</td>\n",
       "      <td>0.148290</td>\n",
       "      <td>0.708795</td>\n",
       "      <td>regulatory T cell</td>\n",
       "      <td>ZYX</td>\n",
       "      <td>regulatory T cell</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZZEF1</th>\n",
       "      <td>3.895481</td>\n",
       "      <td>3.102382e-06</td>\n",
       "      <td>0.001964</td>\n",
       "      <td>1.147203</td>\n",
       "      <td>0.251298</td>\n",
       "      <td>0.803712</td>\n",
       "      <td>regulatory T cell</td>\n",
       "      <td>ZZEF1</td>\n",
       "      <td>regulatory T cell</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>195921 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               baseMean  log2FoldChange     lfcSE      stat    pvalue  \\\n",
       "feature_name                                                            \n",
       "A1BG           4.613264    5.484595e-06  0.052210  0.000174  0.999861   \n",
       "A1BG-AS1       0.591671    9.067974e-05  0.063279 -0.351676  0.725081   \n",
       "A2M            1.206051   -6.738284e-04  0.053568 -0.360525  0.718455   \n",
       "A2M-AS1        0.082406    1.239167e-04  0.060909 -0.815760  0.414638   \n",
       "A2ML1          0.157295   -4.821732e-04  0.046209 -1.113721  0.265399   \n",
       "...                 ...             ...       ...       ...       ...   \n",
       "ZXDC           1.019239    2.913140e-06  0.002257  1.073555  0.283022   \n",
       "ZYG11A         0.014163    1.193448e-07  0.002688 -0.676961  0.498431   \n",
       "ZYG11B         3.356246   -9.622742e-07  0.001682 -0.398524  0.690244   \n",
       "ZYX           14.740712   -4.786789e-06  0.001580 -1.445596  0.148290   \n",
       "ZZEF1          3.895481    3.102382e-06  0.001964  1.147203  0.251298   \n",
       "\n",
       "                  padj standardized_cell_type feature_name          cell_type  \n",
       "feature_name                                                                   \n",
       "A1BG          0.999978                 B cell         A1BG             B cell  \n",
       "A1BG-AS1           NaN                 B cell     A1BG-AS1             B cell  \n",
       "A2M                NaN                 B cell          A2M             B cell  \n",
       "A2M-AS1            NaN                 B cell      A2M-AS1             B cell  \n",
       "A2ML1              NaN                 B cell        A2ML1             B cell  \n",
       "...                ...                    ...          ...                ...  \n",
       "ZXDC               NaN      regulatory T cell         ZXDC  regulatory T cell  \n",
       "ZYG11A             NaN      regulatory T cell       ZYG11A  regulatory T cell  \n",
       "ZYG11B        0.958322      regulatory T cell       ZYG11B  regulatory T cell  \n",
       "ZYX           0.708795      regulatory T cell          ZYX  regulatory T cell  \n",
       "ZZEF1         0.803712      regulatory T cell        ZZEF1  regulatory T cell  \n",
       "\n",
       "[195921 rows x 9 columns]"
      ]
     },
     "execution_count": 116,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "contrast[\"de_res_all\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "id": "65c5a5f2-ed05-4a1c-903a-9dad96afd3bf",
   "metadata": {},
   "outputs": [],
   "source": [
    "resDir = resDir_tables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "id": "aef73d54-5ec7-4f36-ac03-c895012debc4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Saved full B cell data to /data/projects/2023/LCBiome/nsclc_gender_atlas_tmp/out/012_LUAD/03012025/tables/LUAD_DE/Bcell_deg.csv\n",
      "Saved significant B cell data to /data/projects/2023/LCBiome/nsclc_gender_atlas_tmp/out/012_LUAD/03012025/tables/LUAD_DE/Bcell_sig_deg.csv\n",
      "Saved full CD4+ T cell data to /data/projects/2023/LCBiome/nsclc_gender_atlas_tmp/out/012_LUAD/03012025/tables/LUAD_DE/CD4+Tcell_deg.csv\n",
      "Saved significant CD4+ T cell data to /data/projects/2023/LCBiome/nsclc_gender_atlas_tmp/out/012_LUAD/03012025/tables/LUAD_DE/CD4+Tcell_sig_deg.csv\n",
      "Saved full CD8+ T cell data to /data/projects/2023/LCBiome/nsclc_gender_atlas_tmp/out/012_LUAD/03012025/tables/LUAD_DE/CD8+Tcell_deg.csv\n",
      "Saved significant CD8+ T cell data to /data/projects/2023/LCBiome/nsclc_gender_atlas_tmp/out/012_LUAD/03012025/tables/LUAD_DE/CD8+Tcell_sig_deg.csv\n",
      "Saved full dendritic cell data to /data/projects/2023/LCBiome/nsclc_gender_atlas_tmp/out/012_LUAD/03012025/tables/LUAD_DE/dendriticcell_deg.csv\n",
      "Saved significant dendritic cell data to /data/projects/2023/LCBiome/nsclc_gender_atlas_tmp/out/012_LUAD/03012025/tables/LUAD_DE/dendriticcell_sig_deg.csv\n",
      "Saved full endothelial cell data to /data/projects/2023/LCBiome/nsclc_gender_atlas_tmp/out/012_LUAD/03012025/tables/LUAD_DE/endothelialcell_deg.csv\n",
      "Saved significant endothelial cell data to /data/projects/2023/LCBiome/nsclc_gender_atlas_tmp/out/012_LUAD/03012025/tables/LUAD_DE/endothelialcell_sig_deg.csv\n",
      "Saved full epithelial cell data to /data/projects/2023/LCBiome/nsclc_gender_atlas_tmp/out/012_LUAD/03012025/tables/LUAD_DE/epithelialcell_deg.csv\n",
      "Saved significant epithelial cell data to /data/projects/2023/LCBiome/nsclc_gender_atlas_tmp/out/012_LUAD/03012025/tables/LUAD_DE/epithelialcell_sig_deg.csv\n",
      "Saved full fibroblast data to /data/projects/2023/LCBiome/nsclc_gender_atlas_tmp/out/012_LUAD/03012025/tables/LUAD_DE/fibroblast_deg.csv\n",
      "Saved significant fibroblast data to /data/projects/2023/LCBiome/nsclc_gender_atlas_tmp/out/012_LUAD/03012025/tables/LUAD_DE/fibroblast_sig_deg.csv\n",
      "Saved full macrophage data to /data/projects/2023/LCBiome/nsclc_gender_atlas_tmp/out/012_LUAD/03012025/tables/LUAD_DE/macrophage_deg.csv\n",
      "Saved significant macrophage data to /data/projects/2023/LCBiome/nsclc_gender_atlas_tmp/out/012_LUAD/03012025/tables/LUAD_DE/macrophage_sig_deg.csv\n",
      "Saved full neutrophil data to /data/projects/2023/LCBiome/nsclc_gender_atlas_tmp/out/012_LUAD/03012025/tables/LUAD_DE/neutrophil_deg.csv\n",
      "Saved significant neutrophil data to /data/projects/2023/LCBiome/nsclc_gender_atlas_tmp/out/012_LUAD/03012025/tables/LUAD_DE/neutrophil_sig_deg.csv\n",
      "Saved full plasma cell data to /data/projects/2023/LCBiome/nsclc_gender_atlas_tmp/out/012_LUAD/03012025/tables/LUAD_DE/plasmacell_deg.csv\n",
      "Saved significant plasma cell data to /data/projects/2023/LCBiome/nsclc_gender_atlas_tmp/out/012_LUAD/03012025/tables/LUAD_DE/plasmacell_sig_deg.csv\n",
      "Saved full regulatory T cell data to /data/projects/2023/LCBiome/nsclc_gender_atlas_tmp/out/012_LUAD/03012025/tables/LUAD_DE/regulatoryTcell_deg.csv\n",
      "Saved significant regulatory T cell data to /data/projects/2023/LCBiome/nsclc_gender_atlas_tmp/out/012_LUAD/03012025/tables/LUAD_DE/regulatoryTcell_sig_deg.csv\n"
     ]
    }
   ],
   "source": [
    "cell_types = contrast[\"de_res_all\"][\"cell_type\"].unique()\n",
    "\n",
    "# Loop through each cell type and save its data to CSV\n",
    "for cell_type in cell_types:\n",
    "    # Filter the DataFrame for the current cell type\n",
    "    ct_all_deg = contrast[\"de_res_all\"][contrast[\"de_res_all\"][\"cell_type\"] == cell_type]\n",
    "    \n",
    "    # Remove spaces from cell_type for the filename\n",
    "    cell_type_filename = cell_type.replace(\" \", \"\")\n",
    "    \n",
    "    # Save the full data for the cell type\n",
    "    filename_deg = f\"{cell_type_filename}_deg.csv\"\n",
    "    filepath_deg = os.path.join(resDir, filename_deg)\n",
    "    ct_all_deg.to_csv(filepath_deg, index=False)\n",
    "    print(f\"Saved full {cell_type} data to {filepath_deg}\")\n",
    "    \n",
    "    # Apply additional filtering for significant results\n",
    "    filtered_df = ct_all_deg[(ct_all_deg['padj'] < 0.1) & (ct_all_deg['log2FoldChange'].abs() > 1)]\n",
    "    \n",
    "    # Save the filtered significant data for the cell type\n",
    "    filename_sig_deg = f\"{cell_type_filename}_sig_deg.csv\"\n",
    "    filepath_sig_deg = os.path.join(resDir, filename_sig_deg)\n",
    "    filtered_df.to_csv(filepath_sig_deg, index=False)\n",
    "    print(f\"Saved significant {cell_type} data to {filepath_sig_deg}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "id": "97bcba56-bdca-4332-918c-34362f11c8d9",
   "metadata": {
    "lines_to_next_cell": 2
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PDF with all volcano plots saved as 'volcano_plots_all_cell_types.pdf'\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<Figure size 800x400 with 0 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "<Figure size 800x400 with 0 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "<Figure size 800x400 with 0 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "<Figure size 800x400 with 0 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "<Figure size 800x400 with 0 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "<Figure size 800x400 with 0 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "<Figure size 800x400 with 0 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "<Figure size 800x400 with 0 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "<Figure size 800x400 with 0 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "<Figure size 800x400 with 0 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "<Figure size 800x400 with 0 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "from matplotlib.backends.backend_pdf import PdfPages\n",
    "\n",
    "# Assuming `contrast` and `dc` are already defined and set up in your environment\n",
    "cell_types = contrast[\"de_res_all\"][\"cell_type\"].unique()\n",
    "\n",
    "# Create a PDF file to save all volcano plots\n",
    "with PdfPages(f\"{resDir_figures}/volcano_plots_all_cell_types.pdf\") as pdf:\n",
    "    for cell_type in cell_types:\n",
    "        # Filter results for the current cell type\n",
    "        results_df = contrast[\"de_res_all\"][contrast[\"de_res_all\"][\"cell_type\"] == cell_type]\n",
    "\n",
    "        # Generate the volcano plot\n",
    "        plt.figure(figsize=(8, 4))\n",
    "        dc.plot_volcano_df(\n",
    "            results_df,\n",
    "            x='log2FoldChange',\n",
    "            y='padj',\n",
    "            top=20,\n",
    "            figsize=(8, 4)\n",
    "        )\n",
    "        \n",
    "        # Set title to indicate cell type\n",
    "        plt.title(f\"Volcano Plot for Cell Type: {cell_type}\")\n",
    "        \n",
    "        # Save the current figure to the PDF\n",
    "        pdf.savefig()\n",
    "        plt.close()  # Close the plot to avoid display issues in the next iteration\n",
    "\n",
    "print(\"PDF with all volcano plots saved as 'volcano_plots_all_cell_types.pdf'\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4e8a0ced-1656-4944-a650-fb3cccdbaf8f",
   "metadata": {},
   "source": [
    "## Pydeseq tumor vs normal FEMALE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "id": "c126cb3a-ff0f-4573-8abc-63bb8509e33e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "View of AnnData object with n_obs × n_vars = 696 × 17811\n",
       "    obs: 'sample', 'standardized_cell_type', 'uicc_stage', 'ever_smoker', 'donor_id', 'origin', 'dataset', 'tumor_stage', 'study', 'platform', 'suspension_type', 'assay_ontology_term_id', 'development_stage_ontology_term_id', 'disease_ontology_term_id', 'self_reported_ethnicity_ontology_term_id', 'is_primary_data', 'organism_ontology_term_id', 'sex_ontology_term_id', 'tissue_ontology_term_id', 'assay', 'disease', 'organism', 'sex', 'tissue', 'self_reported_ethnicity', 'development_stage', 'psbulk_n_cells', 'psbulk_counts'\n",
       "    var: 'is_highly_variable', 'mito', 'n_cells_by_counts', 'mean_counts', 'pct_dropout_by_counts', 'total_counts', 'feature_is_filtered', 'feature_name', 'feature_reference', 'feature_biotype'\n",
       "    layers: 'psbulk_props'"
      ]
     },
     "execution_count": 132,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pdata_tumor_normal_female"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "id": "22d9e8f8-abc9-46fe-85f0-480491aeb746",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "View of AnnData object with n_obs × n_vars = 518 × 17811\n",
       "    obs: 'sample', 'standardized_cell_type', 'uicc_stage', 'ever_smoker', 'donor_id', 'origin', 'dataset', 'tumor_stage', 'study', 'platform', 'suspension_type', 'assay_ontology_term_id', 'development_stage_ontology_term_id', 'disease_ontology_term_id', 'self_reported_ethnicity_ontology_term_id', 'is_primary_data', 'organism_ontology_term_id', 'sex_ontology_term_id', 'tissue_ontology_term_id', 'assay', 'disease', 'organism', 'sex', 'tissue', 'self_reported_ethnicity', 'development_stage', 'psbulk_n_cells', 'psbulk_counts'\n",
       "    var: 'is_highly_variable', 'mito', 'n_cells_by_counts', 'mean_counts', 'pct_dropout_by_counts', 'total_counts', 'feature_is_filtered', 'feature_name', 'feature_reference', 'feature_biotype'\n",
       "    layers: 'psbulk_props'"
      ]
     },
     "execution_count": 133,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pdata_tumor_normal_male"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "5ce29dbd-65cd-4bbb-a24d-552531d18388",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Guo_Zhang_2018_P1010_TTY_regulatory T cell         tumor-primary\n",
       "Guo_Zhang_2018_P1118_TTH_regulatory T cell         tumor-primary\n",
       "Guo_Zhang_2018_P1202_TTH_regulatory T cell         tumor-primary\n",
       "Guo_Zhang_2018_P1202_TTR_regulatory T cell         tumor-primary\n",
       "Guo_Zhang_2018_P1202_TTY_regulatory T cell         tumor-primary\n",
       "                                                      ...       \n",
       "UKIM-V-2_P5_tumor_primary_regulatory T cell        tumor-primary\n",
       "UKIM-V-2_P7_normal_adjacent_regulatory T cell    normal-adjacent\n",
       "UKIM-V-2_P7_tumor_primary_regulatory T cell        tumor-primary\n",
       "UKIM-V_P1_normal_adjacent_regulatory T cell      normal-adjacent\n",
       "UKIM-V_P1_tumor_primary_regulatory T cell          tumor-primary\n",
       "Name: origin, Length: 74, dtype: object"
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pdata_tumor_normal_female[pdata_tumor_normal_female.obs[cell_type_class] == \"regulatory T cell\"].obs.origin"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "61d2cdea-08a2-48d1-b7cc-2791a7210730",
   "metadata": {
    "lines_to_next_cell": 2,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "B cell\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Fitting size factors...\n",
      "... done in 0.03 seconds.\n",
      "\n",
      "Fitting dispersions...\n",
      "... done in 1.89 seconds.\n",
      "\n",
      "Fitting dispersion trend curve...\n",
      "... done in 0.39 seconds.\n",
      "\n",
      "Fitting MAP dispersions...\n",
      "... done in 2.31 seconds.\n",
      "\n",
      "Fitting LFCs...\n",
      "... done in 1.81 seconds.\n",
      "\n",
      "Calculating cook's distance...\n",
      "... done in 0.08 seconds.\n",
      "\n",
      "Replacing 85 outlier genes.\n",
      "\n",
      "Fitting dispersions...\n",
      "... done in 0.03 seconds.\n",
      "\n",
      "Fitting MAP dispersions...\n",
      "... done in 0.03 seconds.\n",
      "\n",
      "Fitting LFCs...\n",
      "... done in 0.03 seconds.\n",
      "\n",
      "Running Wald tests...\n",
      "... done in 10.53 seconds.\n",
      "\n",
      "Fitting MAP LFCs...\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Log2 fold change & Wald test p-value: origin tumor-primary vs normal-adjacent\n",
      "              baseMean  log2FoldChange     lfcSE      stat    pvalue      padj\n",
      "feature_name                                                                  \n",
      "A1BG          6.836024       -0.096811  0.290096 -0.333722  0.738589  0.998775\n",
      "A1BG-AS1      1.168325       -0.078309  0.423935 -0.184720  0.853449       NaN\n",
      "A2M           1.288579       -0.575284  0.495898 -1.160086  0.246014       NaN\n",
      "A2M-AS1       0.247093       -0.922256  0.751221 -1.227676  0.219569       NaN\n",
      "A2ML1         0.052073       -1.773977  1.897850 -0.934730  0.349927       NaN\n",
      "...                ...             ...       ...       ...       ...       ...\n",
      "ZXDC          2.439321       -0.560718  0.462610 -1.212075  0.225484  0.928499\n",
      "ZYG11A        0.217108       -0.436415  0.821729 -0.531093  0.595354       NaN\n",
      "ZYG11B        3.527944       -0.286148  0.383959 -0.745257  0.456117  0.988955\n",
      "ZYX           5.614568        0.189883  0.225259  0.842956  0.399253  0.988955\n",
      "ZZEF1         3.314924       -0.302941  0.315356 -0.960632  0.336737  0.972208\n",
      "\n",
      "[17811 rows x 6 columns]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/kvalem/.conda/envs/crc_atlas/lib/python3.11/site-packages/pydeseq2/utils.py:1260: RuntimeWarning: overflow encountered in exp\n",
      "  counts - (counts + size) / (1 + size * np.exp(-xbeta - offset))\n",
      "... done in 3.12 seconds.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Shrunk log2 fold change & Wald test p-value: origin tumor-primary vs normal-adjacent\n",
      "              baseMean  log2FoldChange     lfcSE      stat    pvalue      padj\n",
      "feature_name                                                                  \n",
      "A1BG          6.836024       -0.004149  0.095569 -0.333722  0.738589  0.998775\n",
      "A1BG-AS1      1.168325        0.001322  0.120595 -0.184720  0.853449       NaN\n",
      "A2M           1.288579       -0.008168  0.097222 -1.160086  0.246014       NaN\n",
      "A2M-AS1       0.247093       -0.000984  0.116783 -1.227676  0.219569       NaN\n",
      "A2ML1         0.052073       -0.001365  0.087317 -0.934730  0.349927       NaN\n",
      "...                ...             ...       ...       ...       ...       ...\n",
      "ZXDC          2.439321       -0.009504  0.093922 -1.212075  0.225484  0.928499\n",
      "ZYG11A        0.217108        0.003583  0.122013 -0.531093  0.595354       NaN\n",
      "ZYG11B        3.527944       -0.007244  0.095545 -0.745257  0.456117  0.988955\n",
      "ZYX           5.614568        0.013493  0.105674  0.842956  0.399253  0.988955\n",
      "ZZEF1         3.314924       -0.011278  0.100343 -0.960632  0.336737  0.972208\n",
      "\n",
      "[17811 rows x 6 columns]\n",
      "CD4+ T cell\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Fitting size factors...\n",
      "... done in 0.05 seconds.\n",
      "\n",
      "Fitting dispersions...\n",
      "... done in 3.13 seconds.\n",
      "\n",
      "Fitting dispersion trend curve...\n",
      "... done in 0.40 seconds.\n",
      "\n",
      "Fitting MAP dispersions...\n",
      "... done in 2.75 seconds.\n",
      "\n",
      "Fitting LFCs...\n",
      "... done in 2.21 seconds.\n",
      "\n",
      "Calculating cook's distance...\n",
      "... done in 0.13 seconds.\n",
      "\n",
      "Replacing 1979 outlier genes.\n",
      "\n",
      "Fitting dispersions...\n",
      "... done in 0.41 seconds.\n",
      "\n",
      "Fitting MAP dispersions...\n",
      "... done in 0.33 seconds.\n",
      "\n",
      "Fitting LFCs...\n",
      "... done in 0.30 seconds.\n",
      "\n",
      "Running Wald tests...\n",
      "... done in 10.57 seconds.\n",
      "\n",
      "Fitting MAP LFCs...\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Log2 fold change & Wald test p-value: origin tumor-primary vs normal-adjacent\n",
      "                baseMean  log2FoldChange     lfcSE      stat    pvalue  \\\n",
      "feature_name                                                             \n",
      "A1BG           68.592197       -0.112696  0.296928 -0.379540  0.704287   \n",
      "A1BG-AS1       13.471159       -0.319386  0.363249 -0.879248  0.379267   \n",
      "A2M            18.693596       -0.138240  0.366436 -0.377255  0.705984   \n",
      "A2M-AS1        16.792654        0.039457  0.408430  0.096608  0.923038   \n",
      "A2ML1           0.376706        0.285645  1.088145  0.262507  0.792931   \n",
      "...                  ...             ...       ...       ...       ...   \n",
      "ZXDC           27.036616        0.213716  0.325349  0.656881  0.511257   \n",
      "ZYG11A          0.215535       -0.558586  1.337815 -0.417536  0.676286   \n",
      "ZYG11B         41.940148        0.460023  0.320611  1.434833  0.151335   \n",
      "ZYX           241.478463        0.134089  0.217365  0.616886  0.537310   \n",
      "ZZEF1          64.229144        0.260197  0.224417  1.159438  0.246278   \n",
      "\n",
      "                  padj  \n",
      "feature_name            \n",
      "A1BG          0.960059  \n",
      "A1BG-AS1      0.873908  \n",
      "A2M           0.960371  \n",
      "A2M-AS1       0.992801  \n",
      "A2ML1              NaN  \n",
      "...                ...  \n",
      "ZXDC          0.918441  \n",
      "ZYG11A             NaN  \n",
      "ZYG11B        0.698006  \n",
      "ZYX           0.926737  \n",
      "ZZEF1         0.793194  \n",
      "\n",
      "[17811 rows x 6 columns]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/kvalem/.conda/envs/crc_atlas/lib/python3.11/site-packages/pydeseq2/utils.py:1260: RuntimeWarning: overflow encountered in exp\n",
      "  counts - (counts + size) / (1 + size * np.exp(-xbeta - offset))\n",
      "/home/kvalem/.conda/envs/crc_atlas/lib/python3.11/site-packages/pydeseq2/utils.py:1260: RuntimeWarning: overflow encountered in exp\n",
      "  counts - (counts + size) / (1 + size * np.exp(-xbeta - offset))\n",
      "/home/kvalem/.conda/envs/crc_atlas/lib/python3.11/site-packages/pydeseq2/utils.py:1260: RuntimeWarning: overflow encountered in exp\n",
      "  counts - (counts + size) / (1 + size * np.exp(-xbeta - offset))\n",
      "/home/kvalem/.conda/envs/crc_atlas/lib/python3.11/site-packages/pydeseq2/utils.py:1260: RuntimeWarning: overflow encountered in exp\n",
      "  counts - (counts + size) / (1 + size * np.exp(-xbeta - offset))\n",
      "... done in 6.05 seconds.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Shrunk log2 fold change & Wald test p-value: origin tumor-primary vs normal-adjacent\n",
      "                baseMean  log2FoldChange     lfcSE      stat    pvalue  \\\n",
      "feature_name                                                             \n",
      "A1BG           68.592197   -1.325718e-06  0.001485 -0.379540  0.704287   \n",
      "A1BG-AS1       13.471159   -2.685382e-06  0.001431 -0.879248  0.379267   \n",
      "A2M            18.693596   -7.674849e-07  0.001450 -0.377255  0.705984   \n",
      "A2M-AS1        16.792654    2.605234e-07  0.001519  0.096608  0.923038   \n",
      "A2ML1           0.376706    2.437946e-07  0.001558  0.262507  0.792931   \n",
      "...                  ...             ...       ...       ...       ...   \n",
      "ZXDC           27.036616    1.912286e-06  0.001580  0.656881  0.511257   \n",
      "ZYG11A          0.215535   -8.521301e-07  0.001244 -0.417536  0.676286   \n",
      "ZYG11B         41.940148    4.772592e-06  0.001665  1.434833  0.151335   \n",
      "ZYX           241.478463    5.043108e-06  0.001558  0.616886  0.537310   \n",
      "ZZEF1          64.229144    5.376799e-06  0.001594  1.159438  0.246278   \n",
      "\n",
      "                  padj  \n",
      "feature_name            \n",
      "A1BG          0.960059  \n",
      "A1BG-AS1      0.873908  \n",
      "A2M           0.960371  \n",
      "A2M-AS1       0.992801  \n",
      "A2ML1              NaN  \n",
      "...                ...  \n",
      "ZXDC          0.918441  \n",
      "ZYG11A             NaN  \n",
      "ZYG11B        0.698006  \n",
      "ZYX           0.926737  \n",
      "ZZEF1         0.793194  \n",
      "\n",
      "[17811 rows x 6 columns]\n",
      "CD8+ T cell\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Fitting size factors...\n",
      "... done in 0.05 seconds.\n",
      "\n",
      "Fitting dispersions...\n",
      "... done in 2.83 seconds.\n",
      "\n",
      "Fitting dispersion trend curve...\n",
      "... done in 0.40 seconds.\n",
      "\n",
      "Fitting MAP dispersions...\n",
      "... done in 2.67 seconds.\n",
      "\n",
      "Fitting LFCs...\n",
      "... done in 2.23 seconds.\n",
      "\n",
      "Calculating cook's distance...\n",
      "... done in 0.12 seconds.\n",
      "\n",
      "Replacing 1688 outlier genes.\n",
      "\n",
      "Fitting dispersions...\n",
      "... done in 0.33 seconds.\n",
      "\n",
      "Fitting MAP dispersions...\n",
      "... done in 0.28 seconds.\n",
      "\n",
      "Fitting LFCs...\n",
      "... done in 0.41 seconds.\n",
      "\n",
      "Running Wald tests...\n",
      "... done in 10.56 seconds.\n",
      "\n",
      "Fitting MAP LFCs...\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Log2 fold change & Wald test p-value: origin tumor-primary vs normal-adjacent\n",
      "                baseMean  log2FoldChange     lfcSE      stat    pvalue  \\\n",
      "feature_name                                                             \n",
      "A1BG           49.259383        0.225264  0.330502  0.681581  0.495504   \n",
      "A1BG-AS1        8.917060       -0.058168  0.367018 -0.158488  0.874072   \n",
      "A2M            17.648511       -0.041725  0.345686 -0.120703  0.903926   \n",
      "A2M-AS1        47.593598       -0.268557  0.371177 -0.723529  0.469355   \n",
      "A2ML1           0.283424       -0.135752  1.442236 -0.094126  0.925009   \n",
      "...                  ...             ...       ...       ...       ...   \n",
      "ZXDC           21.772692       -0.171237  0.319904 -0.535276  0.592459   \n",
      "ZYG11A          0.235857       -0.152158  1.613656 -0.094294  0.924876   \n",
      "ZYG11B         19.735799       -0.146523  0.317097 -0.462077  0.644026   \n",
      "ZYX           197.050642        0.048384  0.191843  0.252207  0.800881   \n",
      "ZZEF1          50.589753        0.005801  0.215541  0.026914  0.978528   \n",
      "\n",
      "                  padj  \n",
      "feature_name            \n",
      "A1BG          0.957154  \n",
      "A1BG-AS1      0.994603  \n",
      "A2M           0.995593  \n",
      "A2M-AS1       0.953376  \n",
      "A2ML1              NaN  \n",
      "...                ...  \n",
      "ZXDC          0.977855  \n",
      "ZYG11A             NaN  \n",
      "ZYG11B        0.983572  \n",
      "ZYX           0.992220  \n",
      "ZZEF1         0.998544  \n",
      "\n",
      "[17811 rows x 6 columns]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/kvalem/.conda/envs/crc_atlas/lib/python3.11/site-packages/pydeseq2/utils.py:1260: RuntimeWarning: overflow encountered in exp\n",
      "  counts - (counts + size) / (1 + size * np.exp(-xbeta - offset))\n",
      "/home/kvalem/.conda/envs/crc_atlas/lib/python3.11/site-packages/pydeseq2/utils.py:1260: RuntimeWarning: overflow encountered in exp\n",
      "  counts - (counts + size) / (1 + size * np.exp(-xbeta - offset))\n",
      "/home/kvalem/.conda/envs/crc_atlas/lib/python3.11/site-packages/pydeseq2/utils.py:1260: RuntimeWarning: overflow encountered in exp\n",
      "  counts - (counts + size) / (1 + size * np.exp(-xbeta - offset))\n",
      "/home/kvalem/.conda/envs/crc_atlas/lib/python3.11/site-packages/pydeseq2/utils.py:1260: RuntimeWarning: overflow encountered in exp\n",
      "  counts - (counts + size) / (1 + size * np.exp(-xbeta - offset))\n",
      "/home/kvalem/.conda/envs/crc_atlas/lib/python3.11/site-packages/pydeseq2/utils.py:1260: RuntimeWarning: overflow encountered in exp\n",
      "  counts - (counts + size) / (1 + size * np.exp(-xbeta - offset))\n",
      "... done in 6.20 seconds.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Shrunk log2 fold change & Wald test p-value: origin tumor-primary vs normal-adjacent\n",
      "                baseMean  log2FoldChange     lfcSE      stat    pvalue  \\\n",
      "feature_name                                                             \n",
      "A1BG           49.259383    1.967730e-06  0.001547  0.681581  0.495504   \n",
      "A1BG-AS1        8.917060   -5.853045e-07  0.001468 -0.158488  0.874072   \n",
      "A2M            17.648511   -4.318617e-07  0.001448 -0.120703  0.903926   \n",
      "A2M-AS1        47.593598   -2.011321e-06  0.001421 -0.723529  0.469355   \n",
      "A2ML1           0.283424   -2.820914e-07  0.001375 -0.094126  0.925009   \n",
      "...                  ...             ...       ...       ...       ...   \n",
      "ZXDC           21.772692   -1.845970e-06  0.001442 -0.535276  0.592459   \n",
      "ZYG11A          0.235857   -2.464692e-07  0.001346 -0.094294  0.924876   \n",
      "ZYG11B         19.735799   -1.635186e-06  0.001451 -0.462077  0.644026   \n",
      "ZYX           197.050642    5.255533e-06  0.001500  0.252207  0.800881   \n",
      "ZZEF1          50.589753    1.095858e-06  0.001488  0.026914  0.978528   \n",
      "\n",
      "                  padj  \n",
      "feature_name            \n",
      "A1BG          0.957154  \n",
      "A1BG-AS1      0.994603  \n",
      "A2M           0.995593  \n",
      "A2M-AS1       0.953376  \n",
      "A2ML1              NaN  \n",
      "...                ...  \n",
      "ZXDC          0.977855  \n",
      "ZYG11A             NaN  \n",
      "ZYG11B        0.983572  \n",
      "ZYX           0.992220  \n",
      "ZZEF1         0.998544  \n",
      "\n",
      "[17811 rows x 6 columns]\n",
      "dendritic cell\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Fitting size factors...\n",
      "... done in 0.04 seconds.\n",
      "\n",
      "Fitting dispersions...\n",
      "... done in 2.13 seconds.\n",
      "\n",
      "Fitting dispersion trend curve...\n",
      "... done in 0.40 seconds.\n",
      "\n",
      "Fitting MAP dispersions...\n",
      "... done in 2.41 seconds.\n",
      "\n",
      "Fitting LFCs...\n",
      "... done in 1.74 seconds.\n",
      "\n",
      "Calculating cook's distance...\n",
      "... done in 0.09 seconds.\n",
      "\n",
      "Replacing 126 outlier genes.\n",
      "\n",
      "Fitting dispersions...\n",
      "... done in 0.04 seconds.\n",
      "\n",
      "Fitting MAP dispersions...\n",
      "... done in 0.04 seconds.\n",
      "\n",
      "Fitting LFCs...\n",
      "... done in 0.04 seconds.\n",
      "\n",
      "Running Wald tests...\n",
      "... done in 10.68 seconds.\n",
      "\n",
      "Fitting MAP LFCs...\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Log2 fold change & Wald test p-value: origin tumor-primary vs normal-adjacent\n",
      "               baseMean  log2FoldChange     lfcSE      stat    pvalue  \\\n",
      "feature_name                                                            \n",
      "A1BG          25.745214       -0.136336  0.248643 -0.548318  0.583473   \n",
      "A1BG-AS1       2.712978        0.335790  0.266716  1.258979  0.208038   \n",
      "A2M           38.975712        0.018136  0.230811  0.078574  0.937372   \n",
      "A2M-AS1        0.986788       -0.151769  0.448897 -0.338092  0.735293   \n",
      "A2ML1          0.087913       -0.245416  1.640254 -0.149621  0.881064   \n",
      "...                 ...             ...       ...       ...       ...   \n",
      "ZXDC           4.706702       -0.314368  0.247365 -1.270865  0.203777   \n",
      "ZYG11A         0.195779       -0.499984  1.622292 -0.308196  0.757933   \n",
      "ZYG11B         7.493308       -0.175285  0.266476 -0.657791  0.510673   \n",
      "ZYX           73.714732        0.038158  0.137998  0.276514  0.782154   \n",
      "ZZEF1          7.039874       -0.290618  0.196075 -1.482175  0.138294   \n",
      "\n",
      "                  padj  \n",
      "feature_name            \n",
      "A1BG          0.893902  \n",
      "A1BG-AS1      0.661530  \n",
      "A2M           0.986109  \n",
      "A2M-AS1       0.938201  \n",
      "A2ML1              NaN  \n",
      "...                ...  \n",
      "ZXDC          0.657085  \n",
      "ZYG11A             NaN  \n",
      "ZYG11B        0.868413  \n",
      "ZYX           0.954615  \n",
      "ZZEF1         0.572029  \n",
      "\n",
      "[17811 rows x 6 columns]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "... done in 2.80 seconds.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Shrunk log2 fold change & Wald test p-value: origin tumor-primary vs normal-adjacent\n",
      "               baseMean  log2FoldChange     lfcSE      stat    pvalue  \\\n",
      "feature_name                                                            \n",
      "A1BG          25.745214       -0.025236  0.143650 -0.548318  0.583473   \n",
      "A1BG-AS1       2.712978        0.062333  0.171494  1.258979  0.208038   \n",
      "A2M           38.975712        0.003975  0.138517  0.078574  0.937372   \n",
      "A2M-AS1        0.986788       -0.009414  0.159900 -0.338092  0.735293   \n",
      "A2ML1          0.087913       -0.001084  0.168231 -0.149621  0.881064   \n",
      "...                 ...             ...       ...       ...       ...   \n",
      "ZXDC           4.706702       -0.066264  0.161961 -1.270865  0.203777   \n",
      "ZYG11A         0.195779       -0.006260  0.155871 -0.308196  0.757933   \n",
      "ZYG11B         7.493308       -0.029851  0.147666 -0.657791  0.510673   \n",
      "ZYX           73.714732        0.016240  0.108703  0.276514  0.782154   \n",
      "ZZEF1          7.039874       -0.095269  0.158718 -1.482175  0.138294   \n",
      "\n",
      "                  padj  \n",
      "feature_name            \n",
      "A1BG          0.893902  \n",
      "A1BG-AS1      0.661530  \n",
      "A2M           0.986109  \n",
      "A2M-AS1       0.938201  \n",
      "A2ML1              NaN  \n",
      "...                ...  \n",
      "ZXDC          0.657085  \n",
      "ZYG11A             NaN  \n",
      "ZYG11B        0.868413  \n",
      "ZYX           0.954615  \n",
      "ZZEF1         0.572029  \n",
      "\n",
      "[17811 rows x 6 columns]\n",
      "endothelial cell\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Fitting size factors...\n",
      "... done in 0.03 seconds.\n",
      "\n",
      "Fitting dispersions...\n",
      "... done in 2.00 seconds.\n",
      "\n",
      "Fitting dispersion trend curve...\n",
      "... done in 0.39 seconds.\n",
      "\n",
      "Fitting MAP dispersions...\n",
      "... done in 2.34 seconds.\n",
      "\n",
      "Fitting LFCs...\n",
      "... done in 1.81 seconds.\n",
      "\n",
      "Calculating cook's distance...\n",
      "... done in 0.06 seconds.\n",
      "\n",
      "Replacing 54 outlier genes.\n",
      "\n",
      "Fitting dispersions...\n",
      "... done in 0.02 seconds.\n",
      "\n",
      "Fitting MAP dispersions...\n",
      "... done in 0.02 seconds.\n",
      "\n",
      "Fitting LFCs...\n",
      "... done in 0.02 seconds.\n",
      "\n",
      "Running Wald tests...\n",
      "... done in 10.64 seconds.\n",
      "\n",
      "Fitting MAP LFCs...\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Log2 fold change & Wald test p-value: origin tumor-primary vs normal-adjacent\n",
      "                baseMean  log2FoldChange     lfcSE      stat    pvalue  \\\n",
      "feature_name                                                             \n",
      "A1BG            3.420936        0.910405  0.282033  3.228006  0.001247   \n",
      "A1BG-AS1        0.585804        0.199409  0.493449  0.404112  0.686130   \n",
      "A2M           212.601385        0.414325  0.181063  2.288285  0.022121   \n",
      "A2M-AS1         0.688039       -0.742509  0.565377 -1.313300  0.189082   \n",
      "A2ML1           0.087024        0.205179  1.737480  0.118090  0.905996   \n",
      "...                  ...             ...       ...       ...       ...   \n",
      "ZXDC            2.902518       -0.082311  0.376034 -0.218892  0.826734   \n",
      "ZYG11A          0.029248        0.272497  3.069963  0.088762  0.929271   \n",
      "ZYG11B          6.029228        0.025885  0.249445  0.103769  0.917353   \n",
      "ZYX            13.099102        0.405874  0.150850  2.690572  0.007133   \n",
      "ZZEF1           3.714210        0.006741  0.211402  0.031887  0.974562   \n",
      "\n",
      "                  padj  \n",
      "feature_name            \n",
      "A1BG          0.019209  \n",
      "A1BG-AS1           NaN  \n",
      "A2M           0.136250  \n",
      "A2M-AS1       0.477972  \n",
      "A2ML1              NaN  \n",
      "...                ...  \n",
      "ZXDC          0.927817  \n",
      "ZYG11A             NaN  \n",
      "ZYG11B        0.967970  \n",
      "ZYX           0.066149  \n",
      "ZZEF1         0.990853  \n",
      "\n",
      "[17811 rows x 6 columns]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "... done in 2.60 seconds.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Shrunk log2 fold change & Wald test p-value: origin tumor-primary vs normal-adjacent\n",
      "                baseMean  log2FoldChange     lfcSE      stat    pvalue  \\\n",
      "feature_name                                                             \n",
      "A1BG            3.420936        0.737634  0.286873  3.228006  0.001247   \n",
      "A1BG-AS1        0.585804        0.027669  0.261104  0.404112  0.686130   \n",
      "A2M           212.601385        0.321767  0.180599  2.288285  0.022121   \n",
      "A2M-AS1         0.688039       -0.158076  0.338386 -1.313300  0.189082   \n",
      "A2ML1           0.087024       -0.014676  0.278997  0.118090  0.905996   \n",
      "...                  ...             ...       ...       ...       ...   \n",
      "ZXDC            2.902518       -0.021069  0.237167 -0.218892  0.826734   \n",
      "ZYG11A          0.029248       -0.008557  0.250941  0.088762  0.929271   \n",
      "ZYG11B          6.029228        0.011692  0.193106  0.103769  0.917353   \n",
      "ZYX            13.099102        0.335234  0.150847  2.690572  0.007133   \n",
      "ZZEF1           3.714210        0.003635  0.173076  0.031887  0.974562   \n",
      "\n",
      "                  padj  \n",
      "feature_name            \n",
      "A1BG          0.019209  \n",
      "A1BG-AS1           NaN  \n",
      "A2M           0.136250  \n",
      "A2M-AS1       0.477972  \n",
      "A2ML1              NaN  \n",
      "...                ...  \n",
      "ZXDC          0.927817  \n",
      "ZYG11A             NaN  \n",
      "ZYG11B        0.967970  \n",
      "ZYX           0.066149  \n",
      "ZZEF1         0.990853  \n",
      "\n",
      "[17811 rows x 6 columns]\n",
      "epithelial cell\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Fitting size factors...\n",
      "... done in 0.03 seconds.\n",
      "\n",
      "Fitting dispersions...\n",
      "... done in 2.10 seconds.\n",
      "\n",
      "Fitting dispersion trend curve...\n",
      "... done in 0.42 seconds.\n",
      "\n",
      "Fitting MAP dispersions...\n",
      "... done in 2.43 seconds.\n",
      "\n",
      "Fitting LFCs...\n",
      "... done in 2.25 seconds.\n",
      "\n",
      "Calculating cook's distance...\n",
      "... done in 0.08 seconds.\n",
      "\n",
      "Replacing 320 outlier genes.\n",
      "\n",
      "Fitting dispersions...\n",
      "... done in 0.10 seconds.\n",
      "\n",
      "Fitting MAP dispersions...\n",
      "... done in 0.10 seconds.\n",
      "\n",
      "Fitting LFCs...\n",
      "... done in 0.12 seconds.\n",
      "\n",
      "Running Wald tests...\n",
      "... done in 10.52 seconds.\n",
      "\n",
      "Fitting MAP LFCs...\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Log2 fold change & Wald test p-value: origin tumor-primary vs normal-adjacent\n",
      "              baseMean  log2FoldChange     lfcSE      stat    pvalue      padj\n",
      "feature_name                                                                  \n",
      "A1BG          9.287246       -0.513872  0.423795 -1.212548  0.225303  0.409370\n",
      "A1BG-AS1      1.914802       -1.516623  0.406009 -3.735440  0.000187  0.002369\n",
      "A2M           4.109797        1.045371  0.580974  1.799342  0.071965  0.190632\n",
      "A2M-AS1       0.398711       -1.678088  0.623440 -2.691658  0.007110       NaN\n",
      "A2ML1         0.020607       -0.665224  3.023561 -0.220013  0.825861       NaN\n",
      "...                ...             ...       ...       ...       ...       ...\n",
      "ZXDC          4.202048       -0.243010  0.238332 -1.019626  0.307906  0.500928\n",
      "ZYG11A        0.132186        0.348567  1.178512  0.295769  0.767407       NaN\n",
      "ZYG11B        6.671836        0.285723  0.224307  1.273803  0.202733  0.380821\n",
      "ZYX           5.263825        0.621081  0.359226  1.728943  0.083819  0.211819\n",
      "ZZEF1         4.706567       -0.389099  0.241257 -1.612795  0.106789  0.249823\n",
      "\n",
      "[17811 rows x 6 columns]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "... done in 2.44 seconds.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Shrunk log2 fold change & Wald test p-value: origin tumor-primary vs normal-adjacent\n",
      "              baseMean  log2FoldChange     lfcSE      stat    pvalue      padj\n",
      "feature_name                                                                  \n",
      "A1BG          9.287246       -0.266714  0.371416 -1.212548  0.225303  0.409370\n",
      "A1BG-AS1      1.914802       -1.333762  0.427292 -3.735440  0.000187  0.002369\n",
      "A2M           4.109797        0.455060  0.566899  1.799342  0.071965  0.190632\n",
      "A2M-AS1       0.398711       -1.501786  0.730030 -2.691658  0.007110       NaN\n",
      "A2ML1         0.020607        0.026643  1.017134 -0.220013  0.825861       NaN\n",
      "...                ...             ...       ...       ...       ...       ...\n",
      "ZXDC          4.202048       -0.183487  0.225368 -1.019626  0.307906  0.500928\n",
      "ZYG11A        0.132186        0.240686  0.997138  0.295769  0.767407       NaN\n",
      "ZYG11B        6.671836        0.224880  0.215102  1.273803  0.202733  0.380821\n",
      "ZYX           5.263825        0.401044  0.349037  1.728943  0.083819  0.211819\n",
      "ZZEF1         4.706567       -0.300407  0.236986 -1.612795  0.106789  0.249823\n",
      "\n",
      "[17811 rows x 6 columns]\n",
      "fibroblast\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Fitting size factors...\n",
      "... done in 0.02 seconds.\n",
      "\n",
      "Fitting dispersions...\n",
      "... done in 1.64 seconds.\n",
      "\n",
      "Fitting dispersion trend curve...\n",
      "... done in 0.37 seconds.\n",
      "\n",
      "Fitting MAP dispersions...\n",
      "... done in 1.94 seconds.\n",
      "\n",
      "Fitting LFCs...\n",
      "... done in 1.72 seconds.\n",
      "\n",
      "Calculating cook's distance...\n",
      "... done in 0.03 seconds.\n",
      "\n",
      "Replacing 227 outlier genes.\n",
      "\n",
      "Fitting dispersions...\n",
      "... done in 0.08 seconds.\n",
      "\n",
      "Fitting MAP dispersions...\n",
      "... done in 0.08 seconds.\n",
      "\n",
      "Fitting LFCs...\n",
      "... done in 0.08 seconds.\n",
      "\n",
      "Running Wald tests...\n",
      "... done in 10.47 seconds.\n",
      "\n",
      "Fitting MAP LFCs...\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Log2 fold change & Wald test p-value: origin tumor-primary vs normal-adjacent\n",
      "                baseMean  log2FoldChange     lfcSE      stat    pvalue  \\\n",
      "feature_name                                                             \n",
      "A1BG            2.753374        1.726825  0.544518  3.171291  0.001518   \n",
      "A1BG-AS1        0.367909        0.801320  1.081224  0.741123  0.458619   \n",
      "A2M           332.107188       -0.180609  0.446152 -0.404816  0.685613   \n",
      "A2M-AS1         0.886699       -1.111197  0.833851 -1.332608  0.182660   \n",
      "A2ML1           0.044637       -0.825762  3.140544 -0.262936  0.792600   \n",
      "...                  ...             ...       ...       ...       ...   \n",
      "ZXDC            0.825521       -0.428782  0.820566 -0.522544  0.601291   \n",
      "ZYG11A          0.043281       -0.396306  3.140363 -0.126198  0.899575   \n",
      "ZYG11B          2.116193        0.109004  0.421844  0.258398  0.796100   \n",
      "ZYX            16.858977       -0.357000  0.247358 -1.443254  0.148949   \n",
      "ZZEF1           1.476396        0.257032  0.580560  0.442731  0.657960   \n",
      "\n",
      "                  padj  \n",
      "feature_name            \n",
      "A1BG          0.039161  \n",
      "A1BG-AS1           NaN  \n",
      "A2M           0.913492  \n",
      "A2M-AS1            NaN  \n",
      "A2ML1              NaN  \n",
      "...                ...  \n",
      "ZXDC               NaN  \n",
      "ZYG11A             NaN  \n",
      "ZYG11B        0.950999  \n",
      "ZYX           0.534569  \n",
      "ZZEF1         0.905399  \n",
      "\n",
      "[17811 rows x 6 columns]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/kvalem/.conda/envs/crc_atlas/lib/python3.11/site-packages/pydeseq2/utils.py:1260: RuntimeWarning: overflow encountered in exp\n",
      "  counts - (counts + size) / (1 + size * np.exp(-xbeta - offset))\n",
      "... done in 5.83 seconds.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Shrunk log2 fold change & Wald test p-value: origin tumor-primary vs normal-adjacent\n",
      "                baseMean  log2FoldChange     lfcSE      stat    pvalue  \\\n",
      "feature_name                                                             \n",
      "A1BG            2.753374    1.254283e+00  0.567963  3.171291  0.001518   \n",
      "A1BG-AS1        0.367909    1.218913e-06  0.001864  0.741123  0.458619   \n",
      "A2M           332.107188   -9.694307e-07  0.001422 -0.404816  0.685613   \n",
      "A2M-AS1         0.886699   -1.665328e-06  0.001387 -1.332608  0.182660   \n",
      "A2ML1           0.044637   -2.222531e-07  0.001163 -0.262936  0.792600   \n",
      "...                  ...             ...       ...       ...       ...   \n",
      "ZXDC            0.825521   -5.131180e-07  0.001479 -0.522544  0.601291   \n",
      "ZYG11A          0.043281    1.588732e-07  0.002076 -0.126198  0.899575   \n",
      "ZYG11B          2.116193    6.534478e-07  0.001639  0.258398  0.796100   \n",
      "ZYX            16.858977   -6.113224e-06  0.001443 -1.443254  0.148949   \n",
      "ZZEF1           1.476396    7.674500e-07  0.001629  0.442731  0.657960   \n",
      "\n",
      "                  padj  \n",
      "feature_name            \n",
      "A1BG          0.039161  \n",
      "A1BG-AS1           NaN  \n",
      "A2M           0.913492  \n",
      "A2M-AS1            NaN  \n",
      "A2ML1              NaN  \n",
      "...                ...  \n",
      "ZXDC               NaN  \n",
      "ZYG11A             NaN  \n",
      "ZYG11B        0.950999  \n",
      "ZYX           0.534569  \n",
      "ZZEF1         0.905399  \n",
      "\n",
      "[17811 rows x 6 columns]\n",
      "macrophage\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Fitting size factors...\n",
      "... done in 0.04 seconds.\n",
      "\n",
      "Fitting dispersions...\n",
      "... done in 2.28 seconds.\n",
      "\n",
      "Fitting dispersion trend curve...\n",
      "... done in 0.43 seconds.\n",
      "\n",
      "Fitting MAP dispersions...\n",
      "... done in 2.80 seconds.\n",
      "\n",
      "Fitting LFCs...\n",
      "... done in 2.00 seconds.\n",
      "\n",
      "Calculating cook's distance...\n",
      "... done in 0.10 seconds.\n",
      "\n",
      "Replacing 193 outlier genes.\n",
      "\n",
      "Fitting dispersions...\n",
      "... done in 0.07 seconds.\n",
      "\n",
      "Fitting MAP dispersions...\n",
      "... done in 0.06 seconds.\n",
      "\n",
      "Fitting LFCs...\n",
      "... done in 0.07 seconds.\n",
      "\n",
      "Running Wald tests...\n",
      "... done in 10.71 seconds.\n",
      "\n",
      "Fitting MAP LFCs...\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Log2 fold change & Wald test p-value: origin tumor-primary vs normal-adjacent\n",
      "                baseMean  log2FoldChange     lfcSE      stat        pvalue  \\\n",
      "feature_name                                                                 \n",
      "A1BG           61.624709        0.409183  0.251234  1.628691  1.033785e-01   \n",
      "A1BG-AS1       11.192793       -0.167666  0.155082 -1.081146  2.796321e-01   \n",
      "A2M           493.374677        1.781544  0.229750  7.754284  8.884356e-15   \n",
      "A2M-AS1         2.846238        0.051125  0.265131  0.192831  8.470915e-01   \n",
      "A2ML1           0.657493        0.858547  1.140514  0.752772  4.515867e-01   \n",
      "...                  ...             ...       ...       ...           ...   \n",
      "ZXDC           17.776871       -0.070301  0.164487 -0.427396  6.690910e-01   \n",
      "ZYG11A          0.181475        1.205039  1.073093  1.122959  2.614550e-01   \n",
      "ZYG11B         74.453040       -0.376119  0.156126 -2.409072  1.599314e-02   \n",
      "ZYX           373.453465        0.170851  0.138671  1.232060  2.179266e-01   \n",
      "ZZEF1          36.598425       -0.153528  0.111519 -1.376691  1.686080e-01   \n",
      "\n",
      "                      padj  \n",
      "feature_name                \n",
      "A1BG          2.418636e-01  \n",
      "A1BG-AS1      4.681705e-01  \n",
      "A2M           1.381122e-12  \n",
      "A2M-AS1       9.185844e-01  \n",
      "A2ML1         6.348803e-01  \n",
      "...                    ...  \n",
      "ZXDC          8.066803e-01  \n",
      "ZYG11A                 NaN  \n",
      "ZYG11B        6.301337e-02  \n",
      "ZYX           3.999610e-01  \n",
      "ZZEF1         3.364798e-01  \n",
      "\n",
      "[17811 rows x 6 columns]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "... done in 2.75 seconds.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Shrunk log2 fold change & Wald test p-value: origin tumor-primary vs normal-adjacent\n",
      "                baseMean  log2FoldChange     lfcSE      stat        pvalue  \\\n",
      "feature_name                                                                 \n",
      "A1BG           61.624709        0.305510  0.243987  1.628691  1.033785e-01   \n",
      "A1BG-AS1       11.192793       -0.145239  0.150885 -1.081146  2.796321e-01   \n",
      "A2M           493.374677        1.729215  0.231919  7.754284  8.884356e-15   \n",
      "A2M-AS1         2.846238        0.021143  0.233271  0.192831  8.470915e-01   \n",
      "A2ML1           0.657493       -0.021213  0.453969  0.752772  4.515867e-01   \n",
      "...                  ...             ...       ...       ...           ...   \n",
      "ZXDC           17.776871       -0.059363  0.154806 -0.427396  6.690910e-01   \n",
      "ZYG11A          0.181475        0.008360  0.483878  1.122959  2.614550e-01   \n",
      "ZYG11B         74.453040       -0.337140  0.152797 -2.409072  1.599314e-02   \n",
      "ZYX           373.453465        0.166083  0.135525  1.232060  2.179266e-01   \n",
      "ZZEF1          36.598425       -0.141723  0.110092 -1.376691  1.686080e-01   \n",
      "\n",
      "                      padj  \n",
      "feature_name                \n",
      "A1BG          2.418636e-01  \n",
      "A1BG-AS1      4.681705e-01  \n",
      "A2M           1.381122e-12  \n",
      "A2M-AS1       9.185844e-01  \n",
      "A2ML1         6.348803e-01  \n",
      "...                    ...  \n",
      "ZXDC          8.066803e-01  \n",
      "ZYG11A                 NaN  \n",
      "ZYG11B        6.301337e-02  \n",
      "ZYX           3.999610e-01  \n",
      "ZZEF1         3.364798e-01  \n",
      "\n",
      "[17811 rows x 6 columns]\n",
      "malignant cell\n",
      "Not running DEseq for: malignant cell : only present in: set()\n",
      "neutrophil\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/kvalem/.local/lib/python3.11/site-packages/anndata/_core/aligned_df.py:67: ImplicitModificationWarning: Transforming to str index.\n",
      "  warnings.warn(\"Transforming to str index.\", ImplicitModificationWarning)\n",
      "Fitting size factors...\n",
      "... done in 0.01 seconds.\n",
      "\n",
      "Fitting dispersions...\n",
      "... done in 1.30 seconds.\n",
      "\n",
      "Fitting dispersion trend curve...\n",
      "/home/kvalem/.conda/envs/crc_atlas/lib/python3.11/site-packages/pydeseq2/dds.py:716: UserWarning: The dispersion trend curve fitting did not converge. Switching to a mean-based dispersion trend.\n",
      "  self._fit_parametric_dispersion_trend(vst)\n",
      "... done in 0.31 seconds.\n",
      "\n",
      "Fitting MAP dispersions...\n",
      "... done in 1.91 seconds.\n",
      "\n",
      "Fitting LFCs...\n",
      "... done in 1.67 seconds.\n",
      "\n",
      "Calculating cook's distance...\n",
      "... done in 0.01 seconds.\n",
      "\n",
      "Replacing 165 outlier genes.\n",
      "\n",
      "Fitting dispersions...\n",
      "... done in 0.05 seconds.\n",
      "\n",
      "Fitting MAP dispersions...\n",
      "... done in 0.06 seconds.\n",
      "\n",
      "Fitting LFCs...\n",
      "... done in 0.05 seconds.\n",
      "\n",
      "Running Wald tests...\n",
      "... done in 10.64 seconds.\n",
      "\n",
      "Fitting MAP LFCs...\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Log2 fold change & Wald test p-value: origin tumor-primary vs normal-adjacent\n",
      "                baseMean  log2FoldChange     lfcSE      stat    pvalue  \\\n",
      "feature_name                                                             \n",
      "A1BG            0.049570        0.970603  2.468966  0.393121  0.694230   \n",
      "A1BG-AS1        0.064780        0.332821  2.526817  0.131716  0.895209   \n",
      "A2M             2.969021        0.220362  0.987247  0.223209  0.823373   \n",
      "A2M-AS1         0.458062       -1.152237  1.550626 -0.743079  0.457434   \n",
      "A2ML1           0.000000             NaN       NaN       NaN       NaN   \n",
      "...                  ...             ...       ...       ...       ...   \n",
      "ZXDC            6.761438       -0.736332  0.459147 -1.603695  0.108781   \n",
      "ZYG11A          0.000000             NaN       NaN       NaN       NaN   \n",
      "ZYG11B         10.061387       -0.433302  0.375258 -1.154676  0.248223   \n",
      "ZYX           139.065266        0.022885  0.204549  0.111881  0.910918   \n",
      "ZZEF1          10.449286       -0.870961  0.486076 -1.791822  0.073161   \n",
      "\n",
      "                  padj  \n",
      "feature_name            \n",
      "A1BG          0.999917  \n",
      "A1BG-AS1      0.999917  \n",
      "A2M           0.999917  \n",
      "A2M-AS1       0.999917  \n",
      "A2ML1              NaN  \n",
      "...                ...  \n",
      "ZXDC          0.999917  \n",
      "ZYG11A             NaN  \n",
      "ZYG11B        0.999917  \n",
      "ZYX           0.999917  \n",
      "ZZEF1         0.999917  \n",
      "\n",
      "[17811 rows x 6 columns]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/kvalem/.conda/envs/crc_atlas/lib/python3.11/site-packages/pydeseq2/utils.py:1260: RuntimeWarning: overflow encountered in exp\n",
      "  counts - (counts + size) / (1 + size * np.exp(-xbeta - offset))\n",
      "... done in 5.63 seconds.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Shrunk log2 fold change & Wald test p-value: origin tumor-primary vs normal-adjacent\n",
      "                baseMean  log2FoldChange     lfcSE      stat    pvalue  \\\n",
      "feature_name                                                             \n",
      "A1BG            0.049570    9.671307e-08  0.001417  0.393121  0.694230   \n",
      "A1BG-AS1        0.064780   -2.228125e-07  0.001158  0.131716  0.895209   \n",
      "A2M             2.969021    2.048603e-07  0.001419  0.223209  0.823373   \n",
      "A2M-AS1         0.458062   -7.880722e-07  0.001169 -0.743079  0.457434   \n",
      "A2ML1           0.000000             NaN       NaN       NaN       NaN   \n",
      "...                  ...             ...       ...       ...       ...   \n",
      "ZXDC            6.761438   -3.663442e-06  0.001255 -1.603695  0.108781   \n",
      "ZYG11A          0.000000             NaN       NaN       NaN       NaN   \n",
      "ZYG11B         10.061387   -3.271733e-06  0.001270 -1.154676  0.248223   \n",
      "ZYX           139.065266    5.142546e-07  0.001379  0.111881  0.910918   \n",
      "ZZEF1          10.449286   -3.792724e-06  0.001239 -1.791822  0.073161   \n",
      "\n",
      "                  padj  \n",
      "feature_name            \n",
      "A1BG          0.999917  \n",
      "A1BG-AS1      0.999917  \n",
      "A2M           0.999917  \n",
      "A2M-AS1       0.999917  \n",
      "A2ML1              NaN  \n",
      "...                ...  \n",
      "ZXDC          0.999917  \n",
      "ZYG11A             NaN  \n",
      "ZYG11B        0.999917  \n",
      "ZYX           0.999917  \n",
      "ZZEF1         0.999917  \n",
      "\n",
      "[17811 rows x 6 columns]\n",
      "plasma cell\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Fitting size factors...\n",
      "... done in 0.03 seconds.\n",
      "\n",
      "Fitting dispersions...\n",
      "... done in 2.03 seconds.\n",
      "\n",
      "Fitting dispersion trend curve...\n",
      "... done in 0.42 seconds.\n",
      "\n",
      "Fitting MAP dispersions...\n",
      "... done in 2.32 seconds.\n",
      "\n",
      "Fitting LFCs...\n",
      "... done in 2.34 seconds.\n",
      "\n",
      "Calculating cook's distance...\n",
      "... done in 0.06 seconds.\n",
      "\n",
      "Replacing 141 outlier genes.\n",
      "\n",
      "Fitting dispersions...\n",
      "... done in 0.05 seconds.\n",
      "\n",
      "Fitting MAP dispersions...\n",
      "... done in 0.05 seconds.\n",
      "\n",
      "Fitting LFCs...\n",
      "... done in 0.05 seconds.\n",
      "\n",
      "Running Wald tests...\n",
      "... done in 10.59 seconds.\n",
      "\n",
      "Fitting MAP LFCs...\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Log2 fold change & Wald test p-value: origin tumor-primary vs normal-adjacent\n",
      "               baseMean  log2FoldChange     lfcSE      stat    pvalue  \\\n",
      "feature_name                                                            \n",
      "A1BG          25.920354       -0.301475  0.222074 -1.357542  0.174609   \n",
      "A1BG-AS1       1.295659        0.112272  0.416550  0.269529  0.787523   \n",
      "A2M            9.845287       -0.090189  0.559427 -0.161217  0.871923   \n",
      "A2M-AS1        0.645594       -1.290793  0.972078 -1.327870  0.184221   \n",
      "A2ML1          0.015050       -1.425868  2.588165 -0.550919  0.581689   \n",
      "...                 ...             ...       ...       ...       ...   \n",
      "ZXDC           1.277678       -0.276857  0.451928 -0.612614  0.540132   \n",
      "ZYG11A         0.331280       -0.223771  0.653289 -0.342529  0.731953   \n",
      "ZYG11B         2.948444        0.453814  0.392684  1.155673  0.247815   \n",
      "ZYX            4.748772       -0.254933  0.356634 -0.714831  0.474714   \n",
      "ZZEF1          3.404421       -0.272120  0.319921 -0.850586  0.394999   \n",
      "\n",
      "                  padj  \n",
      "feature_name            \n",
      "A1BG          0.521796  \n",
      "A1BG-AS1           NaN  \n",
      "A2M           0.960932  \n",
      "A2M-AS1            NaN  \n",
      "A2ML1              NaN  \n",
      "...                ...  \n",
      "ZXDC               NaN  \n",
      "ZYG11A             NaN  \n",
      "ZYG11B             NaN  \n",
      "ZYX           0.767724  \n",
      "ZZEF1              NaN  \n",
      "\n",
      "[17811 rows x 6 columns]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "... done in 2.71 seconds.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Shrunk log2 fold change & Wald test p-value: origin tumor-primary vs normal-adjacent\n",
      "               baseMean  log2FoldChange     lfcSE      stat    pvalue  \\\n",
      "feature_name                                                            \n",
      "A1BG          25.920354       -0.096844  0.183137 -1.357542  0.174609   \n",
      "A1BG-AS1       1.295659        0.025356  0.239748  0.269529  0.787523   \n",
      "A2M            9.845287       -0.004355  0.218801 -0.161217  0.871923   \n",
      "A2M-AS1        0.645594       -0.010935  0.220984 -1.327870  0.184221   \n",
      "A2ML1          0.015050        0.000010  0.248723 -0.550919  0.581689   \n",
      "...                 ...             ...       ...       ...       ...   \n",
      "ZXDC           1.277678       -0.013319  0.226521 -0.612614  0.540132   \n",
      "ZYG11A         0.331280        0.017407  0.269614 -0.342529  0.731953   \n",
      "ZYG11B         2.948444        0.055479  0.236746  1.155673  0.247815   \n",
      "ZYX            4.748772       -0.031387  0.207016 -0.714831  0.474714   \n",
      "ZZEF1          3.404421       -0.044202  0.199162 -0.850586  0.394999   \n",
      "\n",
      "                  padj  \n",
      "feature_name            \n",
      "A1BG          0.521796  \n",
      "A1BG-AS1           NaN  \n",
      "A2M           0.960932  \n",
      "A2M-AS1            NaN  \n",
      "A2ML1              NaN  \n",
      "...                ...  \n",
      "ZXDC               NaN  \n",
      "ZYG11A             NaN  \n",
      "ZYG11B             NaN  \n",
      "ZYX           0.767724  \n",
      "ZZEF1              NaN  \n",
      "\n",
      "[17811 rows x 6 columns]\n",
      "regulatory T cell\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Fitting size factors...\n",
      "... done in 0.04 seconds.\n",
      "\n",
      "Fitting dispersions...\n",
      "... done in 2.48 seconds.\n",
      "\n",
      "Fitting dispersion trend curve...\n",
      "... done in 0.39 seconds.\n",
      "\n",
      "Fitting MAP dispersions...\n",
      "... done in 2.21 seconds.\n",
      "\n",
      "Fitting LFCs...\n",
      "... done in 2.05 seconds.\n",
      "\n",
      "Calculating cook's distance...\n",
      "... done in 0.08 seconds.\n",
      "\n",
      "Replacing 1459 outlier genes.\n",
      "\n",
      "Fitting dispersions...\n",
      "... done in 0.32 seconds.\n",
      "\n",
      "Fitting MAP dispersions...\n",
      "... done in 0.23 seconds.\n",
      "\n",
      "Fitting LFCs...\n",
      "... done in 0.21 seconds.\n",
      "\n",
      "Running Wald tests...\n",
      "... done in 10.88 seconds.\n",
      "\n",
      "Fitting MAP LFCs...\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Log2 fold change & Wald test p-value: origin tumor-primary vs normal-adjacent\n",
      "               baseMean  log2FoldChange     lfcSE      stat    pvalue  \\\n",
      "feature_name                                                            \n",
      "A1BG           8.394548       -0.023311  0.335508 -0.069478  0.944609   \n",
      "A1BG-AS1       1.646350        0.183794  0.459626  0.399877  0.689247   \n",
      "A2M            1.120256       -0.538052  0.577106 -0.932329  0.351166   \n",
      "A2M-AS1        0.964130       -0.977202  0.717689 -1.361596  0.173325   \n",
      "A2ML1          0.039965       -2.118367  3.044130 -0.695886  0.486500   \n",
      "...                 ...             ...       ...       ...       ...   \n",
      "ZXDC           2.175864       -0.483790  0.414537 -1.167061  0.243186   \n",
      "ZYG11A         0.057260       -2.055337  2.425106 -0.847525  0.396703   \n",
      "ZYG11B         4.984575       -0.067816  0.385280 -0.176016  0.860281   \n",
      "ZYX           19.014864       -0.347384  0.224488 -1.547451  0.121754   \n",
      "ZZEF1          6.800469       -0.189774  0.282040 -0.672861  0.501036   \n",
      "\n",
      "                  padj  \n",
      "feature_name            \n",
      "A1BG          0.978486  \n",
      "A1BG-AS1           NaN  \n",
      "A2M                NaN  \n",
      "A2M-AS1            NaN  \n",
      "A2ML1              NaN  \n",
      "...                ...  \n",
      "ZXDC          0.571328  \n",
      "ZYG11A             NaN  \n",
      "ZYG11B        0.949155  \n",
      "ZYX           0.427284  \n",
      "ZZEF1         0.781010  \n",
      "\n",
      "[17811 rows x 6 columns]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/kvalem/.conda/envs/crc_atlas/lib/python3.11/site-packages/pydeseq2/utils.py:1260: RuntimeWarning: overflow encountered in exp\n",
      "  counts - (counts + size) / (1 + size * np.exp(-xbeta - offset))\n",
      "/home/kvalem/.conda/envs/crc_atlas/lib/python3.11/site-packages/pydeseq2/utils.py:1260: RuntimeWarning: overflow encountered in exp\n",
      "  counts - (counts + size) / (1 + size * np.exp(-xbeta - offset))\n",
      "/home/kvalem/.conda/envs/crc_atlas/lib/python3.11/site-packages/pydeseq2/utils.py:1260: RuntimeWarning: overflow encountered in multiply\n",
      "  counts - (counts + size) / (1 + size * np.exp(-xbeta - offset))\n",
      "/home/kvalem/.conda/envs/crc_atlas/lib/python3.11/site-packages/pydeseq2/utils.py:1260: RuntimeWarning: overflow encountered in exp\n",
      "  counts - (counts + size) / (1 + size * np.exp(-xbeta - offset))\n",
      "... done in 6.10 seconds.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Shrunk log2 fold change & Wald test p-value: origin tumor-primary vs normal-adjacent\n",
      "               baseMean  log2FoldChange     lfcSE      stat    pvalue  \\\n",
      "feature_name                                                            \n",
      "A1BG           8.394548   -1.880110e-07  0.001673 -0.069478  0.944609   \n",
      "A1BG-AS1       1.646350    3.721156e-06  0.002116  0.399877  0.689247   \n",
      "A2M            1.120256   -5.311020e-07  0.001730 -0.932329  0.351166   \n",
      "A2M-AS1        0.964130   -1.028639e-06  0.001682 -1.361596  0.173325   \n",
      "A2ML1          0.039965    1.276819e-07  0.002271 -0.695886  0.486500   \n",
      "...                 ...             ...       ...       ...       ...   \n",
      "ZXDC           2.175864   -2.659679e-06  0.001786 -1.167061  0.243186   \n",
      "ZYG11A         0.057260   -6.464404e-08  0.001548 -0.847525  0.396703   \n",
      "ZYG11B         4.984575   -3.949415e-07  0.001750 -0.176016  0.860281   \n",
      "ZYX           19.014864   -7.261606e-06  0.001601 -1.547451  0.121754   \n",
      "ZZEF1          6.800469   -2.570082e-06  0.001728 -0.672861  0.501036   \n",
      "\n",
      "                  padj  \n",
      "feature_name            \n",
      "A1BG          0.978486  \n",
      "A1BG-AS1           NaN  \n",
      "A2M                NaN  \n",
      "A2M-AS1            NaN  \n",
      "A2ML1              NaN  \n",
      "...                ...  \n",
      "ZXDC          0.571328  \n",
      "ZYG11A             NaN  \n",
      "ZYG11B        0.949155  \n",
      "ZYX           0.427284  \n",
      "ZZEF1         0.781010  \n",
      "\n",
      "[17811 rows x 6 columns]\n"
     ]
    }
   ],
   "source": [
    "import pdb\n",
    "for contrast in contrasts:\n",
    "    de_res = {}\n",
    "\n",
    "    for ct in cell_types:\n",
    "        print(ct)\n",
    "        \n",
    "        pb_ct = pdata_tumor_normal_female[pdata_tumor_normal_female.obs[cell_type_class] == ct].copy()\n",
    "\n",
    "     \n",
    "        #pdb.set_trace()\n",
    "        if len(set(pb_ct.obs[\"origin\"]).intersection([contrast[\"reference\"], contrast[\"condition\"]])) < 1:\n",
    "             print(\n",
    "                 \"Not running DEseq for: \"\n",
    "                  + ct\n",
    "                  + \" : only present in: \"\n",
    "                  + str(set(pb_ct.obs[\"sex\"]).intersection([contrast[\"reference\"], contrast[\"condition\"]]))\n",
    "             )\n",
    "             continue\n",
    "        #pdb.set_trace()\n",
    "                     \n",
    "        dds = DeseqDataSet(\n",
    "           adata=pb_ct,\n",
    "           design_factors=[contrast[\"var\"]],\n",
    "           ref_level=[contrast[\"var\"], 'normal-adjacent'],\n",
    "           refit_cooks=True,\n",
    "           n_cpus=cpus,\n",
    "           )\n",
    "             \n",
    "        \n",
    "\n",
    "        # Compute LFCs\n",
    "        dds.deseq2()\n",
    "\n",
    " \n",
    "\n",
    "        \n",
    "\n",
    "        # Extract contrast\n",
    "        stat_res = DeseqStats(\n",
    "              dds,\n",
    "              contrast=[contrast[\"var\"],'tumor-primary', 'normal-adjacent'],\n",
    "              #n_cpus=cpus,\n",
    "          )\n",
    "   \n",
    "    \n",
    "\n",
    "        # Compute Wald test\n",
    "        stat_res.summary()\n",
    "\n",
    "        # Shrink LFCs\n",
    "\n",
    "        coeff = contrast[\"var\"] + \"_\" + contrast[\"name\"]\n",
    "        stat_res.lfc_shrink(coeff=coeff)\n",
    "\n",
    "        # Register cell type results\n",
    "        de_res[ct] = stat_res.results_df\n",
    "        de_res[ct][\"standardized_cell_type\"] = ct\n",
    "        de_res[ct][\"feature_name\"] = stat_res.results_df.index.values\n",
    "\n",
    "        de_res[ct].drop(columns=[\"feature_name\"]).to_csv(\n",
    "                Path(\n",
    "                    contrast[\"res_dir\"],\n",
    "                    ct.replace(\" \", \"_\") + \"_\" + contrast[\"name\"] + \"_deseq.tsv\",\n",
    "                ),\n",
    "                sep=\"\\t\",\n",
    "            )\n",
    "\n",
    "        # Register results for current contrast\n",
    "        contrast[\"de_res\"] = de_res\n",
    "        contrast[\"de_res_all\"] = pd.concat([df.assign(cell_type=ct) for ct, df in de_res.items()])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "id": "540889db-5756-4b37-addb-0a50cf021c43",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "contrast[\"de_res_all\"].to_csv(\"dea_female_tumor_vs_normal.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "e43c186f-8705-4881-a90e-8383b42b4ee1",
   "metadata": {},
   "outputs": [],
   "source": [
    "resDir = resDir_tables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "539d3087-179c-4eb3-a540-45666c954460",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Saved full B cell data to /data/projects/2023/LCBiome/nsclc_gender_atlas_tmp/out/012_LUAD/03012025/tables/LUAD_DE/Bcell_deg.csv\n",
      "Saved significant B cell data to /data/projects/2023/LCBiome/nsclc_gender_atlas_tmp/out/012_LUAD/03012025/tables/LUAD_DE/Bcell_sig_deg.csv\n",
      "Saved full CD4+ T cell data to /data/projects/2023/LCBiome/nsclc_gender_atlas_tmp/out/012_LUAD/03012025/tables/LUAD_DE/CD4+Tcell_deg.csv\n",
      "Saved significant CD4+ T cell data to /data/projects/2023/LCBiome/nsclc_gender_atlas_tmp/out/012_LUAD/03012025/tables/LUAD_DE/CD4+Tcell_sig_deg.csv\n",
      "Saved full CD8+ T cell data to /data/projects/2023/LCBiome/nsclc_gender_atlas_tmp/out/012_LUAD/03012025/tables/LUAD_DE/CD8+Tcell_deg.csv\n",
      "Saved significant CD8+ T cell data to /data/projects/2023/LCBiome/nsclc_gender_atlas_tmp/out/012_LUAD/03012025/tables/LUAD_DE/CD8+Tcell_sig_deg.csv\n",
      "Saved full dendritic cell data to /data/projects/2023/LCBiome/nsclc_gender_atlas_tmp/out/012_LUAD/03012025/tables/LUAD_DE/dendriticcell_deg.csv\n",
      "Saved significant dendritic cell data to /data/projects/2023/LCBiome/nsclc_gender_atlas_tmp/out/012_LUAD/03012025/tables/LUAD_DE/dendriticcell_sig_deg.csv\n",
      "Saved full endothelial cell data to /data/projects/2023/LCBiome/nsclc_gender_atlas_tmp/out/012_LUAD/03012025/tables/LUAD_DE/endothelialcell_deg.csv\n",
      "Saved significant endothelial cell data to /data/projects/2023/LCBiome/nsclc_gender_atlas_tmp/out/012_LUAD/03012025/tables/LUAD_DE/endothelialcell_sig_deg.csv\n",
      "Saved full epithelial cell data to /data/projects/2023/LCBiome/nsclc_gender_atlas_tmp/out/012_LUAD/03012025/tables/LUAD_DE/epithelialcell_deg.csv\n",
      "Saved significant epithelial cell data to /data/projects/2023/LCBiome/nsclc_gender_atlas_tmp/out/012_LUAD/03012025/tables/LUAD_DE/epithelialcell_sig_deg.csv\n",
      "Saved full fibroblast data to /data/projects/2023/LCBiome/nsclc_gender_atlas_tmp/out/012_LUAD/03012025/tables/LUAD_DE/fibroblast_deg.csv\n",
      "Saved significant fibroblast data to /data/projects/2023/LCBiome/nsclc_gender_atlas_tmp/out/012_LUAD/03012025/tables/LUAD_DE/fibroblast_sig_deg.csv\n",
      "Saved full macrophage data to /data/projects/2023/LCBiome/nsclc_gender_atlas_tmp/out/012_LUAD/03012025/tables/LUAD_DE/macrophage_deg.csv\n",
      "Saved significant macrophage data to /data/projects/2023/LCBiome/nsclc_gender_atlas_tmp/out/012_LUAD/03012025/tables/LUAD_DE/macrophage_sig_deg.csv\n",
      "Saved full neutrophil data to /data/projects/2023/LCBiome/nsclc_gender_atlas_tmp/out/012_LUAD/03012025/tables/LUAD_DE/neutrophil_deg.csv\n",
      "Saved significant neutrophil data to /data/projects/2023/LCBiome/nsclc_gender_atlas_tmp/out/012_LUAD/03012025/tables/LUAD_DE/neutrophil_sig_deg.csv\n",
      "Saved full plasma cell data to /data/projects/2023/LCBiome/nsclc_gender_atlas_tmp/out/012_LUAD/03012025/tables/LUAD_DE/plasmacell_deg.csv\n",
      "Saved significant plasma cell data to /data/projects/2023/LCBiome/nsclc_gender_atlas_tmp/out/012_LUAD/03012025/tables/LUAD_DE/plasmacell_sig_deg.csv\n",
      "Saved full regulatory T cell data to /data/projects/2023/LCBiome/nsclc_gender_atlas_tmp/out/012_LUAD/03012025/tables/LUAD_DE/regulatoryTcell_deg.csv\n",
      "Saved significant regulatory T cell data to /data/projects/2023/LCBiome/nsclc_gender_atlas_tmp/out/012_LUAD/03012025/tables/LUAD_DE/regulatoryTcell_sig_deg.csv\n"
     ]
    }
   ],
   "source": [
    "cell_types = contrast[\"de_res_all\"][\"cell_type\"].unique()\n",
    "\n",
    "# Loop through each cell type and save its data to CSV\n",
    "for cell_type in cell_types:\n",
    "    # Filter the DataFrame for the current cell type\n",
    "    ct_all_deg = contrast[\"de_res_all\"][contrast[\"de_res_all\"][\"cell_type\"] == cell_type]\n",
    "    \n",
    "    # Remove spaces from cell_type for the filename\n",
    "    cell_type_filename = cell_type.replace(\" \", \"\")\n",
    "    \n",
    "    # Save the full data for the cell type\n",
    "    filename_deg = f\"{cell_type_filename}_deg_80cells.csv\"\n",
    "    filepath_deg = os.path.join(resDir, filename_deg)\n",
    "    ct_all_deg.to_csv(filepath_deg, index=False)\n",
    "    print(f\"Saved full {cell_type} data to {filepath_deg}\")\n",
    "    \n",
    "    # Apply additional filtering for significant results\n",
    "    filtered_df = ct_all_deg[(ct_all_deg['padj'] < 0.1) & (ct_all_deg['log2FoldChange'].abs() > 1)]\n",
    "    \n",
    "    # Save the filtered significant data for the cell type\n",
    "    filename_sig_deg = f\"{cell_type_filename}_sig_deg.csv\"\n",
    "    filepath_sig_deg = os.path.join(resDir, filename_sig_deg)\n",
    "    filtered_df.to_csv(filepath_sig_deg, index=False)\n",
    "    print(f\"Saved significant {cell_type} data to {filepath_sig_deg}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "a8e0a65c-6563-43b7-b302-f818058a0547",
   "metadata": {
    "lines_to_next_cell": 2
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PDF with all volcano plots saved as 'volcano_plots_all_cell_types.pdf'\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<Figure size 800x400 with 0 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "<Figure size 800x400 with 0 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "<Figure size 800x400 with 0 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "<Figure size 800x400 with 0 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "<Figure size 800x400 with 0 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "<Figure size 800x400 with 0 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "<Figure size 800x400 with 0 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "<Figure size 800x400 with 0 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "<Figure size 800x400 with 0 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "<Figure size 800x400 with 0 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "<Figure size 800x400 with 0 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "from matplotlib.backends.backend_pdf import PdfPages\n",
    "\n",
    "# Assuming `contrast` and `dc` are already defined and set up in your environment\n",
    "cell_types = contrast[\"de_res_all\"][\"cell_type\"].unique()\n",
    "\n",
    "# Create a PDF file to save all volcano plots\n",
    "with PdfPages(f\"{resDir_figures}/volcano_plots_all_cell_types.pdf\") as pdf:\n",
    "    for cell_type in cell_types:\n",
    "        # Filter results for the current cell type\n",
    "        results_df = contrast[\"de_res_all\"][contrast[\"de_res_all\"][\"cell_type\"] == cell_type]\n",
    "\n",
    "        # Generate the volcano plot\n",
    "        plt.figure(figsize=(8, 4))\n",
    "        dc.plot_volcano_df(\n",
    "            results_df,\n",
    "            x='log2FoldChange',\n",
    "            y='padj',\n",
    "            top=20,\n",
    "            figsize=(8, 4)\n",
    "        )\n",
    "        \n",
    "        # Set title to indicate cell type\n",
    "        plt.title(f\"Volcano Plot for Cell Type: {cell_type}\")\n",
    "        \n",
    "        # Save the current figure to the PDF\n",
    "        pdf.savefig()\n",
    "        plt.close()  # Close the plot to avoid display issues in the next iteration\n",
    "\n",
    "print(\"PDF with all volcano plots saved as 'volcano_plots_all_cell_types.pdf'\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 135,
   "id": "e8f1aeba-9193-4992-ac90-eadab9bee2a1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>standardized_cell_type</th>\n",
       "      <th>uicc_stage</th>\n",
       "      <th>ever_smoker</th>\n",
       "      <th>donor_id</th>\n",
       "      <th>origin</th>\n",
       "      <th>dataset</th>\n",
       "      <th>tumor_stage</th>\n",
       "      <th>study</th>\n",
       "      <th>platform</th>\n",
       "      <th>...</th>\n",
       "      <th>tissue_ontology_term_id</th>\n",
       "      <th>assay</th>\n",
       "      <th>disease</th>\n",
       "      <th>organism</th>\n",
       "      <th>sex</th>\n",
       "      <th>tissue</th>\n",
       "      <th>self_reported_ethnicity</th>\n",
       "      <th>development_stage</th>\n",
       "      <th>psbulk_n_cells</th>\n",
       "      <th>psbulk_counts</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>He_Fan_2021_LUAD1_B cell</th>\n",
       "      <td>He_Fan_2021_LUAD1</td>\n",
       "      <td>B cell</td>\n",
       "      <td>I</td>\n",
       "      <td>no</td>\n",
       "      <td>He_Fan_2021_P1</td>\n",
       "      <td>tumor-primary</td>\n",
       "      <td>He_Fan_2021</td>\n",
       "      <td>early</td>\n",
       "      <td>He_Fan_2021</td>\n",
       "      <td>10x</td>\n",
       "      <td>...</td>\n",
       "      <td>UBERON:0002048</td>\n",
       "      <td>10x 3' v2</td>\n",
       "      <td>lung adenocarcinoma</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>lung</td>\n",
       "      <td>unknown</td>\n",
       "      <td>48-year-old human stage</td>\n",
       "      <td>89.0</td>\n",
       "      <td>118337.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>He_Fan_2021_LUAD3_B cell</th>\n",
       "      <td>He_Fan_2021_LUAD3</td>\n",
       "      <td>B cell</td>\n",
       "      <td>I</td>\n",
       "      <td>yes</td>\n",
       "      <td>He_Fan_2021_P3</td>\n",
       "      <td>tumor-primary</td>\n",
       "      <td>He_Fan_2021</td>\n",
       "      <td>early</td>\n",
       "      <td>He_Fan_2021</td>\n",
       "      <td>10x</td>\n",
       "      <td>...</td>\n",
       "      <td>UBERON:0002048</td>\n",
       "      <td>10x 3' v2</td>\n",
       "      <td>lung adenocarcinoma</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>lung</td>\n",
       "      <td>unknown</td>\n",
       "      <td>58-year-old human stage</td>\n",
       "      <td>45.0</td>\n",
       "      <td>44666.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>He_Fan_2021_LUAD5_B cell</th>\n",
       "      <td>He_Fan_2021_LUAD5</td>\n",
       "      <td>B cell</td>\n",
       "      <td>II</td>\n",
       "      <td>no</td>\n",
       "      <td>He_Fan_2021_P5</td>\n",
       "      <td>tumor-primary</td>\n",
       "      <td>He_Fan_2021</td>\n",
       "      <td>early</td>\n",
       "      <td>He_Fan_2021</td>\n",
       "      <td>10x</td>\n",
       "      <td>...</td>\n",
       "      <td>UBERON:0002048</td>\n",
       "      <td>10x 3' v2</td>\n",
       "      <td>lung adenocarcinoma</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>lung</td>\n",
       "      <td>unknown</td>\n",
       "      <td>56-year-old human stage</td>\n",
       "      <td>328.0</td>\n",
       "      <td>769382.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>He_Fan_2021_N1_B cell</th>\n",
       "      <td>He_Fan_2021_N1</td>\n",
       "      <td>B cell</td>\n",
       "      <td>I</td>\n",
       "      <td>no</td>\n",
       "      <td>He_Fan_2021_P1</td>\n",
       "      <td>normal-adjacent</td>\n",
       "      <td>He_Fan_2021</td>\n",
       "      <td>early</td>\n",
       "      <td>He_Fan_2021</td>\n",
       "      <td>10x</td>\n",
       "      <td>...</td>\n",
       "      <td>UBERON:0002048</td>\n",
       "      <td>10x 3' v2</td>\n",
       "      <td>lung adenocarcinoma</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>lung</td>\n",
       "      <td>unknown</td>\n",
       "      <td>48-year-old human stage</td>\n",
       "      <td>75.0</td>\n",
       "      <td>102672.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>He_Fan_2021_N3_B cell</th>\n",
       "      <td>He_Fan_2021_N3</td>\n",
       "      <td>B cell</td>\n",
       "      <td>I</td>\n",
       "      <td>yes</td>\n",
       "      <td>He_Fan_2021_P3</td>\n",
       "      <td>normal-adjacent</td>\n",
       "      <td>He_Fan_2021</td>\n",
       "      <td>early</td>\n",
       "      <td>He_Fan_2021</td>\n",
       "      <td>10x</td>\n",
       "      <td>...</td>\n",
       "      <td>UBERON:0002048</td>\n",
       "      <td>10x 3' v2</td>\n",
       "      <td>lung adenocarcinoma</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>lung</td>\n",
       "      <td>unknown</td>\n",
       "      <td>58-year-old human stage</td>\n",
       "      <td>21.0</td>\n",
       "      <td>26543.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>UKIM-V-2_P5_tumor_primary_regulatory T cell</th>\n",
       "      <td>UKIM-V-2_P5_tumor_primary</td>\n",
       "      <td>regulatory T cell</td>\n",
       "      <td>II</td>\n",
       "      <td>yes</td>\n",
       "      <td>UKIM-V-2_P5</td>\n",
       "      <td>tumor-primary</td>\n",
       "      <td>UKIM-V-2</td>\n",
       "      <td>early</td>\n",
       "      <td>UKIM-V</td>\n",
       "      <td>BD-Rhapsody</td>\n",
       "      <td>...</td>\n",
       "      <td>UBERON:0002048</td>\n",
       "      <td>BD Rhapsody Whole Transcriptome Analysis</td>\n",
       "      <td>lung adenocarcinoma</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>lung</td>\n",
       "      <td>unknown</td>\n",
       "      <td>62-year-old human stage</td>\n",
       "      <td>183.0</td>\n",
       "      <td>466299.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>UKIM-V-2_P7_normal_adjacent_regulatory T cell</th>\n",
       "      <td>UKIM-V-2_P7_normal_adjacent</td>\n",
       "      <td>regulatory T cell</td>\n",
       "      <td>III</td>\n",
       "      <td>yes</td>\n",
       "      <td>UKIM-V-2_P7</td>\n",
       "      <td>normal-adjacent</td>\n",
       "      <td>UKIM-V-2</td>\n",
       "      <td>advanced</td>\n",
       "      <td>UKIM-V</td>\n",
       "      <td>BD-Rhapsody</td>\n",
       "      <td>...</td>\n",
       "      <td>UBERON:0002048</td>\n",
       "      <td>BD Rhapsody Whole Transcriptome Analysis</td>\n",
       "      <td>lung adenocarcinoma</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>lung</td>\n",
       "      <td>unknown</td>\n",
       "      <td>69-year-old human stage</td>\n",
       "      <td>12.0</td>\n",
       "      <td>26681.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>UKIM-V-2_P7_tumor_primary_regulatory T cell</th>\n",
       "      <td>UKIM-V-2_P7_tumor_primary</td>\n",
       "      <td>regulatory T cell</td>\n",
       "      <td>III</td>\n",
       "      <td>yes</td>\n",
       "      <td>UKIM-V-2_P7</td>\n",
       "      <td>tumor-primary</td>\n",
       "      <td>UKIM-V-2</td>\n",
       "      <td>advanced</td>\n",
       "      <td>UKIM-V</td>\n",
       "      <td>BD-Rhapsody</td>\n",
       "      <td>...</td>\n",
       "      <td>UBERON:0002048</td>\n",
       "      <td>BD Rhapsody Whole Transcriptome Analysis</td>\n",
       "      <td>lung adenocarcinoma</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>lung</td>\n",
       "      <td>unknown</td>\n",
       "      <td>69-year-old human stage</td>\n",
       "      <td>90.0</td>\n",
       "      <td>188984.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>UKIM-V_P1_normal_adjacent_regulatory T cell</th>\n",
       "      <td>UKIM-V_P1_normal_adjacent</td>\n",
       "      <td>regulatory T cell</td>\n",
       "      <td>I</td>\n",
       "      <td>yes</td>\n",
       "      <td>UKIM-V_P1</td>\n",
       "      <td>normal-adjacent</td>\n",
       "      <td>UKIM-V</td>\n",
       "      <td>early</td>\n",
       "      <td>UKIM-V</td>\n",
       "      <td>BD-Rhapsody</td>\n",
       "      <td>...</td>\n",
       "      <td>UBERON:0002048</td>\n",
       "      <td>BD Rhapsody Whole Transcriptome Analysis</td>\n",
       "      <td>lung adenocarcinoma</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>lung</td>\n",
       "      <td>unknown</td>\n",
       "      <td>65-year-old human stage</td>\n",
       "      <td>34.0</td>\n",
       "      <td>548293.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>UKIM-V_P1_tumor_primary_regulatory T cell</th>\n",
       "      <td>UKIM-V_P1_tumor_primary</td>\n",
       "      <td>regulatory T cell</td>\n",
       "      <td>I</td>\n",
       "      <td>yes</td>\n",
       "      <td>UKIM-V_P1</td>\n",
       "      <td>tumor-primary</td>\n",
       "      <td>UKIM-V</td>\n",
       "      <td>early</td>\n",
       "      <td>UKIM-V</td>\n",
       "      <td>BD-Rhapsody</td>\n",
       "      <td>...</td>\n",
       "      <td>UBERON:0002048</td>\n",
       "      <td>BD Rhapsody Whole Transcriptome Analysis</td>\n",
       "      <td>lung adenocarcinoma</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>lung</td>\n",
       "      <td>unknown</td>\n",
       "      <td>65-year-old human stage</td>\n",
       "      <td>33.0</td>\n",
       "      <td>449844.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>696 rows × 28 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                    sample  \\\n",
       "He_Fan_2021_LUAD1_B cell                                 He_Fan_2021_LUAD1   \n",
       "He_Fan_2021_LUAD3_B cell                                 He_Fan_2021_LUAD3   \n",
       "He_Fan_2021_LUAD5_B cell                                 He_Fan_2021_LUAD5   \n",
       "He_Fan_2021_N1_B cell                                       He_Fan_2021_N1   \n",
       "He_Fan_2021_N3_B cell                                       He_Fan_2021_N3   \n",
       "...                                                                    ...   \n",
       "UKIM-V-2_P5_tumor_primary_regulatory T cell      UKIM-V-2_P5_tumor_primary   \n",
       "UKIM-V-2_P7_normal_adjacent_regulatory T cell  UKIM-V-2_P7_normal_adjacent   \n",
       "UKIM-V-2_P7_tumor_primary_regulatory T cell      UKIM-V-2_P7_tumor_primary   \n",
       "UKIM-V_P1_normal_adjacent_regulatory T cell      UKIM-V_P1_normal_adjacent   \n",
       "UKIM-V_P1_tumor_primary_regulatory T cell          UKIM-V_P1_tumor_primary   \n",
       "\n",
       "                                              standardized_cell_type  \\\n",
       "He_Fan_2021_LUAD1_B cell                                      B cell   \n",
       "He_Fan_2021_LUAD3_B cell                                      B cell   \n",
       "He_Fan_2021_LUAD5_B cell                                      B cell   \n",
       "He_Fan_2021_N1_B cell                                         B cell   \n",
       "He_Fan_2021_N3_B cell                                         B cell   \n",
       "...                                                              ...   \n",
       "UKIM-V-2_P5_tumor_primary_regulatory T cell        regulatory T cell   \n",
       "UKIM-V-2_P7_normal_adjacent_regulatory T cell      regulatory T cell   \n",
       "UKIM-V-2_P7_tumor_primary_regulatory T cell        regulatory T cell   \n",
       "UKIM-V_P1_normal_adjacent_regulatory T cell        regulatory T cell   \n",
       "UKIM-V_P1_tumor_primary_regulatory T cell          regulatory T cell   \n",
       "\n",
       "                                              uicc_stage ever_smoker  \\\n",
       "He_Fan_2021_LUAD1_B cell                               I          no   \n",
       "He_Fan_2021_LUAD3_B cell                               I         yes   \n",
       "He_Fan_2021_LUAD5_B cell                              II          no   \n",
       "He_Fan_2021_N1_B cell                                  I          no   \n",
       "He_Fan_2021_N3_B cell                                  I         yes   \n",
       "...                                                  ...         ...   \n",
       "UKIM-V-2_P5_tumor_primary_regulatory T cell           II         yes   \n",
       "UKIM-V-2_P7_normal_adjacent_regulatory T cell        III         yes   \n",
       "UKIM-V-2_P7_tumor_primary_regulatory T cell          III         yes   \n",
       "UKIM-V_P1_normal_adjacent_regulatory T cell            I         yes   \n",
       "UKIM-V_P1_tumor_primary_regulatory T cell              I         yes   \n",
       "\n",
       "                                                     donor_id  \\\n",
       "He_Fan_2021_LUAD1_B cell                       He_Fan_2021_P1   \n",
       "He_Fan_2021_LUAD3_B cell                       He_Fan_2021_P3   \n",
       "He_Fan_2021_LUAD5_B cell                       He_Fan_2021_P5   \n",
       "He_Fan_2021_N1_B cell                          He_Fan_2021_P1   \n",
       "He_Fan_2021_N3_B cell                          He_Fan_2021_P3   \n",
       "...                                                       ...   \n",
       "UKIM-V-2_P5_tumor_primary_regulatory T cell       UKIM-V-2_P5   \n",
       "UKIM-V-2_P7_normal_adjacent_regulatory T cell     UKIM-V-2_P7   \n",
       "UKIM-V-2_P7_tumor_primary_regulatory T cell       UKIM-V-2_P7   \n",
       "UKIM-V_P1_normal_adjacent_regulatory T cell         UKIM-V_P1   \n",
       "UKIM-V_P1_tumor_primary_regulatory T cell           UKIM-V_P1   \n",
       "\n",
       "                                                        origin      dataset  \\\n",
       "He_Fan_2021_LUAD1_B cell                         tumor-primary  He_Fan_2021   \n",
       "He_Fan_2021_LUAD3_B cell                         tumor-primary  He_Fan_2021   \n",
       "He_Fan_2021_LUAD5_B cell                         tumor-primary  He_Fan_2021   \n",
       "He_Fan_2021_N1_B cell                          normal-adjacent  He_Fan_2021   \n",
       "He_Fan_2021_N3_B cell                          normal-adjacent  He_Fan_2021   \n",
       "...                                                        ...          ...   \n",
       "UKIM-V-2_P5_tumor_primary_regulatory T cell      tumor-primary     UKIM-V-2   \n",
       "UKIM-V-2_P7_normal_adjacent_regulatory T cell  normal-adjacent     UKIM-V-2   \n",
       "UKIM-V-2_P7_tumor_primary_regulatory T cell      tumor-primary     UKIM-V-2   \n",
       "UKIM-V_P1_normal_adjacent_regulatory T cell    normal-adjacent       UKIM-V   \n",
       "UKIM-V_P1_tumor_primary_regulatory T cell        tumor-primary       UKIM-V   \n",
       "\n",
       "                                              tumor_stage        study  \\\n",
       "He_Fan_2021_LUAD1_B cell                            early  He_Fan_2021   \n",
       "He_Fan_2021_LUAD3_B cell                            early  He_Fan_2021   \n",
       "He_Fan_2021_LUAD5_B cell                            early  He_Fan_2021   \n",
       "He_Fan_2021_N1_B cell                               early  He_Fan_2021   \n",
       "He_Fan_2021_N3_B cell                               early  He_Fan_2021   \n",
       "...                                                   ...          ...   \n",
       "UKIM-V-2_P5_tumor_primary_regulatory T cell         early       UKIM-V   \n",
       "UKIM-V-2_P7_normal_adjacent_regulatory T cell    advanced       UKIM-V   \n",
       "UKIM-V-2_P7_tumor_primary_regulatory T cell      advanced       UKIM-V   \n",
       "UKIM-V_P1_normal_adjacent_regulatory T cell         early       UKIM-V   \n",
       "UKIM-V_P1_tumor_primary_regulatory T cell           early       UKIM-V   \n",
       "\n",
       "                                                  platform  ...  \\\n",
       "He_Fan_2021_LUAD1_B cell                               10x  ...   \n",
       "He_Fan_2021_LUAD3_B cell                               10x  ...   \n",
       "He_Fan_2021_LUAD5_B cell                               10x  ...   \n",
       "He_Fan_2021_N1_B cell                                  10x  ...   \n",
       "He_Fan_2021_N3_B cell                                  10x  ...   \n",
       "...                                                    ...  ...   \n",
       "UKIM-V-2_P5_tumor_primary_regulatory T cell    BD-Rhapsody  ...   \n",
       "UKIM-V-2_P7_normal_adjacent_regulatory T cell  BD-Rhapsody  ...   \n",
       "UKIM-V-2_P7_tumor_primary_regulatory T cell    BD-Rhapsody  ...   \n",
       "UKIM-V_P1_normal_adjacent_regulatory T cell    BD-Rhapsody  ...   \n",
       "UKIM-V_P1_tumor_primary_regulatory T cell      BD-Rhapsody  ...   \n",
       "\n",
       "                                              tissue_ontology_term_id  \\\n",
       "He_Fan_2021_LUAD1_B cell                               UBERON:0002048   \n",
       "He_Fan_2021_LUAD3_B cell                               UBERON:0002048   \n",
       "He_Fan_2021_LUAD5_B cell                               UBERON:0002048   \n",
       "He_Fan_2021_N1_B cell                                  UBERON:0002048   \n",
       "He_Fan_2021_N3_B cell                                  UBERON:0002048   \n",
       "...                                                               ...   \n",
       "UKIM-V-2_P5_tumor_primary_regulatory T cell            UBERON:0002048   \n",
       "UKIM-V-2_P7_normal_adjacent_regulatory T cell          UBERON:0002048   \n",
       "UKIM-V-2_P7_tumor_primary_regulatory T cell            UBERON:0002048   \n",
       "UKIM-V_P1_normal_adjacent_regulatory T cell            UBERON:0002048   \n",
       "UKIM-V_P1_tumor_primary_regulatory T cell              UBERON:0002048   \n",
       "\n",
       "                                                                                  assay  \\\n",
       "He_Fan_2021_LUAD1_B cell                                                      10x 3' v2   \n",
       "He_Fan_2021_LUAD3_B cell                                                      10x 3' v2   \n",
       "He_Fan_2021_LUAD5_B cell                                                      10x 3' v2   \n",
       "He_Fan_2021_N1_B cell                                                         10x 3' v2   \n",
       "He_Fan_2021_N3_B cell                                                         10x 3' v2   \n",
       "...                                                                                 ...   \n",
       "UKIM-V-2_P5_tumor_primary_regulatory T cell    BD Rhapsody Whole Transcriptome Analysis   \n",
       "UKIM-V-2_P7_normal_adjacent_regulatory T cell  BD Rhapsody Whole Transcriptome Analysis   \n",
       "UKIM-V-2_P7_tumor_primary_regulatory T cell    BD Rhapsody Whole Transcriptome Analysis   \n",
       "UKIM-V_P1_normal_adjacent_regulatory T cell    BD Rhapsody Whole Transcriptome Analysis   \n",
       "UKIM-V_P1_tumor_primary_regulatory T cell      BD Rhapsody Whole Transcriptome Analysis   \n",
       "\n",
       "                                                           disease  \\\n",
       "He_Fan_2021_LUAD1_B cell                       lung adenocarcinoma   \n",
       "He_Fan_2021_LUAD3_B cell                       lung adenocarcinoma   \n",
       "He_Fan_2021_LUAD5_B cell                       lung adenocarcinoma   \n",
       "He_Fan_2021_N1_B cell                          lung adenocarcinoma   \n",
       "He_Fan_2021_N3_B cell                          lung adenocarcinoma   \n",
       "...                                                            ...   \n",
       "UKIM-V-2_P5_tumor_primary_regulatory T cell    lung adenocarcinoma   \n",
       "UKIM-V-2_P7_normal_adjacent_regulatory T cell  lung adenocarcinoma   \n",
       "UKIM-V-2_P7_tumor_primary_regulatory T cell    lung adenocarcinoma   \n",
       "UKIM-V_P1_normal_adjacent_regulatory T cell    lung adenocarcinoma   \n",
       "UKIM-V_P1_tumor_primary_regulatory T cell      lung adenocarcinoma   \n",
       "\n",
       "                                                   organism     sex tissue  \\\n",
       "He_Fan_2021_LUAD1_B cell                       Homo sapiens  female   lung   \n",
       "He_Fan_2021_LUAD3_B cell                       Homo sapiens  female   lung   \n",
       "He_Fan_2021_LUAD5_B cell                       Homo sapiens  female   lung   \n",
       "He_Fan_2021_N1_B cell                          Homo sapiens  female   lung   \n",
       "He_Fan_2021_N3_B cell                          Homo sapiens  female   lung   \n",
       "...                                                     ...     ...    ...   \n",
       "UKIM-V-2_P5_tumor_primary_regulatory T cell    Homo sapiens  female   lung   \n",
       "UKIM-V-2_P7_normal_adjacent_regulatory T cell  Homo sapiens  female   lung   \n",
       "UKIM-V-2_P7_tumor_primary_regulatory T cell    Homo sapiens  female   lung   \n",
       "UKIM-V_P1_normal_adjacent_regulatory T cell    Homo sapiens  female   lung   \n",
       "UKIM-V_P1_tumor_primary_regulatory T cell      Homo sapiens  female   lung   \n",
       "\n",
       "                                              self_reported_ethnicity  \\\n",
       "He_Fan_2021_LUAD1_B cell                                      unknown   \n",
       "He_Fan_2021_LUAD3_B cell                                      unknown   \n",
       "He_Fan_2021_LUAD5_B cell                                      unknown   \n",
       "He_Fan_2021_N1_B cell                                         unknown   \n",
       "He_Fan_2021_N3_B cell                                         unknown   \n",
       "...                                                               ...   \n",
       "UKIM-V-2_P5_tumor_primary_regulatory T cell                   unknown   \n",
       "UKIM-V-2_P7_normal_adjacent_regulatory T cell                 unknown   \n",
       "UKIM-V-2_P7_tumor_primary_regulatory T cell                   unknown   \n",
       "UKIM-V_P1_normal_adjacent_regulatory T cell                   unknown   \n",
       "UKIM-V_P1_tumor_primary_regulatory T cell                     unknown   \n",
       "\n",
       "                                                     development_stage  \\\n",
       "He_Fan_2021_LUAD1_B cell                       48-year-old human stage   \n",
       "He_Fan_2021_LUAD3_B cell                       58-year-old human stage   \n",
       "He_Fan_2021_LUAD5_B cell                       56-year-old human stage   \n",
       "He_Fan_2021_N1_B cell                          48-year-old human stage   \n",
       "He_Fan_2021_N3_B cell                          58-year-old human stage   \n",
       "...                                                                ...   \n",
       "UKIM-V-2_P5_tumor_primary_regulatory T cell    62-year-old human stage   \n",
       "UKIM-V-2_P7_normal_adjacent_regulatory T cell  69-year-old human stage   \n",
       "UKIM-V-2_P7_tumor_primary_regulatory T cell    69-year-old human stage   \n",
       "UKIM-V_P1_normal_adjacent_regulatory T cell    65-year-old human stage   \n",
       "UKIM-V_P1_tumor_primary_regulatory T cell      65-year-old human stage   \n",
       "\n",
       "                                              psbulk_n_cells psbulk_counts  \n",
       "He_Fan_2021_LUAD1_B cell                                89.0      118337.0  \n",
       "He_Fan_2021_LUAD3_B cell                                45.0       44666.0  \n",
       "He_Fan_2021_LUAD5_B cell                               328.0      769382.0  \n",
       "He_Fan_2021_N1_B cell                                   75.0      102672.0  \n",
       "He_Fan_2021_N3_B cell                                   21.0       26543.0  \n",
       "...                                                      ...           ...  \n",
       "UKIM-V-2_P5_tumor_primary_regulatory T cell            183.0      466299.0  \n",
       "UKIM-V-2_P7_normal_adjacent_regulatory T cell           12.0       26681.0  \n",
       "UKIM-V-2_P7_tumor_primary_regulatory T cell             90.0      188984.0  \n",
       "UKIM-V_P1_normal_adjacent_regulatory T cell             34.0      548293.0  \n",
       "UKIM-V_P1_tumor_primary_regulatory T cell               33.0      449844.0  \n",
       "\n",
       "[696 rows x 28 columns]"
      ]
     },
     "execution_count": 135,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pdata_tumor_normal_female.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 139,
   "id": "f4b5747e-dd12-406a-980d-2814711b2b45",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Count matrix and samplesheet saved successfully.\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "\n",
    "def extract_b_cell_data(adata, gender):\n",
    "    \"\"\"\n",
    "    Extracts B cell data from an AnnData object and returns count matrix and metadata.\n",
    "    \n",
    "    Parameters:\n",
    "        adata (AnnData): The AnnData object\n",
    "        gender (str): Gender label for samples (\"female\" or \"male\")\n",
    "\n",
    "    Returns:\n",
    "        count_matrix (pd.DataFrame): Gene expression matrix\n",
    "        samplesheet (pd.DataFrame): Metadata for each sample\n",
    "    \"\"\"\n",
    "    # Subset to only B cells\n",
    "    b_cell_adata = adata[adata.obs[\"standardized_cell_type\"] == \"B cell\"].copy()\n",
    "\n",
    "    # Extract count matrix (assuming raw counts are stored in .X or .layers[\"counts\"])\n",
    "    count_matrix = pd.DataFrame(\n",
    "        b_cell_adata.X.A if isinstance(b_cell_adata.X, np.matrix) else b_cell_adata.X,\n",
    "        index=b_cell_adata.obs_names,\n",
    "        columns=b_cell_adata.var_names\n",
    "    )\n",
    "    \n",
    "    # Extract metadata for the samplesheet\n",
    "    samplesheet = b_cell_adata.obs.copy()\n",
    "    samplesheet[\"gender\"] = gender  # Add gender column\n",
    "    \n",
    "    return count_matrix, samplesheet\n",
    "\n",
    "# Process female and male datasets\n",
    "count_matrix_female, samplesheet_female = extract_b_cell_data(pdata_tumor_normal_female, \"female\")\n",
    "count_matrix_male, samplesheet_male = extract_b_cell_data(pdata_tumor_normal_male, \"male\")\n",
    "\n",
    "# Combine results from both genders\n",
    "final_count_matrix = pd.concat([count_matrix_female, count_matrix_male], axis=0)\n",
    "final_samplesheet = pd.concat([samplesheet_female, samplesheet_male], axis=0)\n",
    "\n",
    "# Save to CSV\n",
    "final_count_matrix.T.to_csv(\"B_cell_count_matrix.csv\")\n",
    "final_samplesheet.to_csv(\"B_cell_samplesheet.csv\")\n",
    "\n",
    "print(\"Count matrix and samplesheet saved successfully.\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0337d7fd-c8f7-4b86-854f-21b6768b14df",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "crc_atlas",
   "language": "python",
   "name": "crc_atlas"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
